<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Contributing to seq2science &mdash; seq2science  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Frequently Asked Questions (FAQ)" href="content/faq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="content/gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/faq.html">Frequently Asked Questions (FAQ)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Contributing to seq2science</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#i-don-t-want-to-read-this-whole-thing-i-just-have-a-question">I don’t want to read this whole thing i just have a question</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reporting-bugs">Reporting bugs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#code-documentation">Code/documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#gitflow">Gitflow</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#project-design">Project design</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#documentation">Documentation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tests">tests</a></li>
<li class="toctree-l3"><a class="reference internal" href="#source-code">source code</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#envs">envs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#imgs">imgs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#rules">rules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#schemas">schemas</a></li>
<li class="toctree-l4"><a class="reference internal" href="#scripts">scripts</a></li>
<li class="toctree-l4"><a class="reference internal" href="#workflows">workflows</a></li>
<li class="toctree-l4"><a class="reference internal" href="#util-py">util.py</a></li>
<li class="toctree-l4"><a class="reference internal" href="#important-global-variables">important global variables</a></li>
<li class="toctree-l4"><a class="reference internal" href="#global-wildcard-constraints">Global wildcard constraints</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#example-pr">Example PR</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Contributing to seq2science</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/CONTRIBUTING.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="contributing-to-seq2science">
<h1>Contributing to seq2science<a class="headerlink" href="#contributing-to-seq2science" title="Link to this heading"></a></h1>
<p><strong>Thanks for taking the time to contribute!!</strong></p>
<p>The following is a set of guidelines and hints how to get started with contributing to seq2science.</p>
<section id="i-don-t-want-to-read-this-whole-thing-i-just-have-a-question">
<h2>I don’t want to read this whole thing i just have a question<a class="headerlink" href="#i-don-t-want-to-read-this-whole-thing-i-just-have-a-question" title="Link to this heading"></a></h2>
<p>Questions can be asked on the <a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/issues">github issues page</a>.
Make sure to check if the question has been asked before in old issues and check the <a class="reference external" href="content/faq.html">Frequently Asked Section (FAQ)</a> in the docs if it’s there.</p>
<section id="reporting-bugs">
<h3>Reporting bugs<a class="headerlink" href="#reporting-bugs" title="Link to this heading"></a></h3>
<p>Bug reports can be made by making an issue on the <a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/issues">github issues page</a>.</p>
</section>
</section>
<section id="code-documentation">
<h2>Code/documentation<a class="headerlink" href="#code-documentation" title="Link to this heading"></a></h2>
<p>Outside of bugs/docs fixes, before you make a code contribution make sure to check in with us (the developers) if we are willing to support the contribution by making an issue first.</p>
<p>For an example on how to add to the documentation, see the example PR <a class="reference external" href="#example-pr">below</a>.</p>
<section id="gitflow">
<h3>Gitflow<a class="headerlink" href="#gitflow" title="Link to this heading"></a></h3>
<p>We (try to) use the <a class="reference external" href="https://nvie.com/posts/a-successful-git-branching-model/">gitflow</a> model to organise our parallel development.
When making a PR, make sure to start from the develop branch and make a PR to develop branch.</p>
</section>
</section>
<section id="project-design">
<h2>Project design<a class="headerlink" href="#project-design" title="Link to this heading"></a></h2>
<p>Seq2science has grown to be a rather large project, with a complicated code structure.
Before contributing to seq2science it is important to be up-to-date with the file and logic structure.</p>
<section id="documentation">
<h3>Documentation<a class="headerlink" href="#documentation" title="Link to this heading"></a></h3>
<p>The documentation can always be improved, and contributions are highly appreciated!
Apart from the auto-generated part of the docs, all the stuff relevant to the documentation is in the <code class="docutils literal notranslate"><span class="pre">docs</span></code> folder in the root.
The markdown files that make up the documentation are in <code class="docutils literal notranslate"><span class="pre">docs/content</span></code>.</p>
</section>
<section id="tests">
<h3>tests<a class="headerlink" href="#tests" title="Link to this heading"></a></h3>
<p>Seq2science runs tests on a Jenkins server after each push.
Everything related to the tests can be found in the <code class="docutils literal notranslate"><span class="pre">tests</span></code> folder in the root.</p>
</section>
<section id="source-code">
<h3>source code<a class="headerlink" href="#source-code" title="Link to this heading"></a></h3>
<p>The “source code” is split across multiple folders in the <code class="docutils literal notranslate"><span class="pre">seq2science</span></code> folder.</p>
<section id="envs">
<h4>envs<a class="headerlink" href="#envs" title="Link to this heading"></a></h4>
<p>In the <code class="docutils literal notranslate"><span class="pre">seq2science/envs</span></code> folder all the environments for each rule is stored.
When adding new rules make sure to add a new environment here.</p>
</section>
<section id="imgs">
<h4>imgs<a class="headerlink" href="#imgs" title="Link to this heading"></a></h4>
<p>These images are being shipped with seq2science.
At the time of writing the only image that is in this folder is a Radboud University logo that is added to the MultiQC report.</p>
</section>
<section id="rules">
<h4>rules<a class="headerlink" href="#rules" title="Link to this heading"></a></h4>
<p>In <code class="docutils literal notranslate"><span class="pre">seq2science/rules</span></code> is where the magic happens.
Practically all rules and logic is in this folder. Rules are split across ‘logical’ topics, such as alignment, peak calling or downloading.
A workflow can thus easily include a couple of these files and link different topics.</p>
<p>All the files that start with <code class="docutils literal notranslate"><span class="pre">configuration_</span></code> actually do not contain any rules.
These files are used to parse the samples.tsv and config.yaml.
These files are unfortunately pretty complex, but any time something big is changed, you probably have to change some stuff here.</p>
</section>
<section id="schemas">
<h4>schemas<a class="headerlink" href="#schemas" title="Link to this heading"></a></h4>
<p>This folder contains multple schemas.
It contains schemas to parse and extend the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> at the start of the run, but it also contains e.g. the schema for the MultiQC report.</p>
</section>
<section id="scripts">
<h4>scripts<a class="headerlink" href="#scripts" title="Link to this heading"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">seq2science/scripts</span></code> contains all the scripts that rules use to.</p>
</section>
<section id="workflows">
<h4>workflows<a class="headerlink" href="#workflows" title="Link to this heading"></a></h4>
<p>In <code class="docutils literal notranslate"><span class="pre">seq2science/workflows</span></code> are the different workflows.
Each workflow consists of a Snakefile, which defines what the total output is, which schemas are needed, and which rules.
In this folder is also an example samples.tsv and config.yaml, which is copied when using <code class="docutils literal notranslate"><span class="pre">seq2science</span> <span class="pre">init</span></code>.</p>
</section>
<section id="util-py">
<h4>util.py<a class="headerlink" href="#util-py" title="Link to this heading"></a></h4>
<p>Useful utility functions are deposited in this file.</p>
</section>
<section id="important-global-variables">
<h4>important global variables<a class="headerlink" href="#important-global-variables" title="Link to this heading"></a></h4>
<p>Global variables are variables that can be read from anywhere in the workflow.
Both Snakemake and Seq2science uses global variables to keep track of steps and samples.
Most global variables are declared in UPPER_CASE in <code class="docutils literal notranslate"><span class="pre">seq2science/rules/configuration_generic.smk</span></code>.</p>
<ul class="simple">
<li><p><strong>config</strong> is a dictionary with all the configuration stuff.
Everything that you write in the config.yaml is in here.
The config is parsed by the schemas (specified by the Snakefile) in <code class="docutils literal notranslate"><span class="pre">seq2science/schemas</span></code>, and in <code class="docutils literal notranslate"><span class="pre">seq2science/rules/configuration_</span></code>.</p></li>
<li><p><strong>samples</strong> is the samples.tsv as a pandas dataframe, after parsing done in <code class="docutils literal notranslate"><span class="pre">seq2science/rules/configuration_</span></code>.
Unused columns are removed, and used columns with empty spaces are filled.</p>
<ul>
<li><p><strong>ALL_SAMPLES</strong> is a list with all sample names from samples.</p></li>
</ul>
</li>
<li><p><strong>SAMPLEDICT</strong> is a dictionary containing information per sample that is not present in the samples.tsv file.
In here is stored whether samples are SINGLE-end or PAIRED-end, and their corresponding SRR numbers.
There are several sub-dictionaries of SAMPLEDICT:</p>
<ul>
<li><p><strong>ENA_SINGLE_END</strong>, <strong>ENA_PAIRED_END</strong>, <strong>SRA_SINGLE_END</strong>, <strong>SRA_PAIRED_END</strong> contain subsets of the samples,</p></li>
<li><p><strong>RUN2DOWNLOAD</strong> contains the download link per sample run.</p></li>
</ul>
</li>
<li><p><strong>breps</strong> &amp; <strong>treps</strong>, are dataframes with the technical and biological replicates merged (respectively).
If you want to iterate over either of the technical replicates or biological replicates, or check what a sample belongs to, it is probably easiest to use these.
There are several accompanying globals:</p>
<ul>
<li><p><strong>MERGED_TREPS</strong> is a list with technical replicates that contain &gt;1 sample (will be merged during the workflow).</p></li>
<li><p><strong>TREPS_FROM_BREP</strong> and <strong>BREP_FROM_TREPS</strong> are dictionaries that track which replicate contains which sub-replicates.</p></li>
</ul>
</li>
<li><p><strong>ALL_ASSEMBLIES</strong> is a list of assemblies used in the workflow.</p>
<ul>
<li><p><strong>PROVIDERS</strong> is a dictionary to keep track which assembly can be downloaded where.</p></li>
<li><p><strong>HAS_ANNOTATION</strong> is a dictionary to keep track if an assembly has a gene annotation (or has one available for download)</p></li>
<li><p><strong>CUSTOM_ASSEMBLIES</strong> is a dictionary to convert assembly names to custom assembly names, if you use custom assemblies.</p></li>
<li><p><strong>ORI_ASSEMBLIES</strong> is a dictionary to convert custom assembly names to assembly names, if you use custom assemblies.</p></li>
<li><p><strong>UCSC_ASSEMBLIES</strong> is a dictionary that denotes whether an assembly is already available on the UCSC trackhubs.</p></li>
</ul>
</li>
<li><p><strong>WORKFLOW</strong> is a string with the current workflow, as spelled in <code class="docutils literal notranslate"><span class="pre">seq2science/workflows</span></code> (e.g. “chip_seq”).</p></li>
<li><p><strong>FINAL_BAM</strong> is a variable path to the (filtered and) coordinate sorted bam files with wildcards <code class="docutils literal notranslate"><span class="pre">{assembly}</span></code> and <code class="docutils literal notranslate"><span class="pre">{sample}</span></code>.</p>
<ul>
<li><p><strong>FINAL_BAI</strong> is the variable path to the matching bam index.
We often use this to apply workflow-specific logic.</p></li>
</ul>
</li>
<li><p><strong>SEQUENCING_PROTOCOL</strong> contains the current workflow, spelled out properly (e.g. “ChIP-seq”).</p></li>
</ul>
</section>
<section id="global-wildcard-constraints">
<h4>Global wildcard constraints<a class="headerlink" href="#global-wildcard-constraints" title="Link to this heading"></a></h4>
<p>Seq2science sets global wildcard constraints in <code class="docutils literal notranslate"><span class="pre">seq2science/rules/configuration_generic.smk</span></code>.
As wildcards are essentially greedy regex (<code class="docutils literal notranslate"><span class="pre">.*</span></code>), these constraints make sure we adhere to the naming convention of samples and assembly.</p>
<p>For example, one default wildcard constraint in seq2science is that the wildcard <code class="docutils literal notranslate"><span class="pre">{replicate}</span></code> can only match with
the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column of the samples.tsv.</p>
</section>
</section>
<section id="example-pr">
<h3>Example PR<a class="headerlink" href="#example-pr" title="Link to this heading"></a></h3>
<p>If something in the documentation is wrong or unclear, please feel free to change it, no matter how small.
Veteran FOSS (Free and Open Source Software) contributors already know how the process of suggesting changes to a project work (called a PR/Pull Request).
However for people new to FOSS this relatively easy process can be a daunting task.
Here we give an example of how to make changes to the documentation and propose them to us (the maintainers).</p>
<ol class="arabic">
<li><p>Make a github account</p></li>
<li><p>Make a “fork” of the repository</p>
<a class="reference external image-reference" href="resources/fork.jpg"><img alt="fork" src="_images/fork.jpg" /></a>
</li>
<li><p>Now navigate on your “fork” to the documentation you want to change, for instance you could go to<code class="docutils literal notranslate"><span class="pre">docs/content/FAQ.md</span></code> to add a new question (and answer!).</p></li>
<li><p>make some changes</p>
<a class="reference external image-reference" href="resources/edit.jpg"><img alt="fork" src="_images/edit.jpg" /></a>
</li>
<li><p>“commit” your changes to the <code class="docutils literal notranslate"><span class="pre">master</span></code> branch`.</p></li>
<li><p>Make a Pull Request (PR).</p>
<ul>
<li><a class="reference external image-reference" href="resources/PR1.jpg"><img alt="fork" src="_images/PR1.jpg" /></a>
</li>
<li><a class="reference external image-reference" href="resources/PR2.jpg"><img alt="fork" src="_images/PR2.jpg" /></a>
</li>
<li><a class="reference external image-reference" href="resources/PR3.jpg"><img alt="fork" src="_images/PR3.jpg" /></a>
</li>
<li><a class="reference external image-reference" href="resources/PR4.jpg"><img alt="fork" src="_images/PR4.jpg" /></a>
</li>
</ul>
</li>
<li><p>Wait for us to accept it…</p></li>
<li><p>See the changed docs online!</p></li>
</ol>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="content/faq.html" class="btn btn-neutral float-left" title="Frequently Asked Questions (FAQ)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>