# do onstart/onexit things
SAMPLE_SCHEMAS = ['sample', 'assembly', 'technical_replicates']
CONFIG_SCHEMAS = ['general', 'download', 'alignment_general', 'alignment_specific', 'scatac']
include: "../../rules/configuration.smk"


# load all the relevant rules
include: f"{config['rule_dir']}/get_fastq.smk"
include: f"{config['rule_dir']}/trimming.smk"
include: f"{config['rule_dir']}/merge_replicates.smk"
include: f"{config['rule_dir']}/get_genome.smk"
include: f"{config['rule_dir']}/blacklist.smk"
include: f"{config['rule_dir']}/alignment.smk"
include: f"{config['rule_dir']}/bam_cleaning.smk"
include: f"{config['rule_dir']}/qc.smk"
include: f"{config['rule_dir']}/scATAC.smk"


# set the quality_control functions
QUALITY_CONTROL = [get_trimming_qc, get_alignment_qc]


rule seq2science:
    """
    """
    input:
        expand(
            (["{qc_dir}/multiqc_{assemblies}.html"] if config["create_qc_report"] else []) +
            ([f"{{result_dir}}/snap/{CUSTOM_ASSEMBLIES[treps.loc[trep, 'assembly']]}-{trep}.binned.snap" for trep in treps.index]),
            assemblies=ALL_ASSEMBLIES, **config)
