<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Per rule explanation &mdash; seq2science  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="All configurable options" href="schemas.html" />
    <link rel="prev" title="Command line interface" href="cli.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="extensive_docs.html">Extensive docs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Per rule explanation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bam2cram">bam2cram</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bam-bigwig">bam_bigwig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedgraph-bigwig">bedgraph_bigwig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedgraphish-to-bedgraph">bedgraphish_to_bedgraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedtools-slop">bedtools_slop</a></li>
<li class="toctree-l3"><a class="reference internal" href="#blind-clustering">blind_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bowtie2-align">bowtie2_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bowtie2-index">bowtie2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-index">bwa_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-mem">bwa_mem</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-mem2">bwa_mem2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-mem2-index">bwa_mem2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#call-peak-genrich">call_peak_genrich</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chipseeker">chipseeker</a></li>
<li class="toctree-l3"><a class="reference internal" href="#combine-peaks">combine_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#combine-qc-files">combine_qc_files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#complement-blacklist">complement_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computematrix-gene">computeMatrix_gene</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computematrix-peak">computeMatrix_peak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#count-matrix">count_matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#count-matrix-dexseq">count_matrix_DEXseq</a></li>
<li class="toctree-l3"><a class="reference internal" href="#coverage-table">coverage_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cp-unsieved2sieved">cp_unsieved2sieved</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-snap-object">create_SNAP_object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-bins-snap-object">create_bins_SNAP_object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cytoband">cytoband</a></li>
<li class="toctree-l3"><a class="reference internal" href="#decoy-transcripts">decoy_transcripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#deseq2">deseq2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dexseq-count">dexseq_count</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dupradar">dupRadar</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dupradar-combine">dupRadar_combine</a></li>
<li class="toctree-l3"><a class="reference internal" href="#edger-normalization">edgeR_normalization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ena2fastq-pe">ena2fastq_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ena2fastq-se">ena2fastq_SE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extend-genome">extend_genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extend-genome-annotation">extend_genome_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extend-genome-blacklist">extend_genome_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastp-pe">fastp_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastp-se">fastp_SE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastp-qc-pe">fastp_qc_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastp-qc-se">fastp_qc_SE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastq-pair">fastq_pair</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastqc">fastqc</a></li>
<li class="toctree-l3"><a class="reference internal" href="#featurecounts">featureCounts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">featurecounts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gcpercent">gcPercent</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene-id2name">gene_id2name</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genrich-pileup">genrich_pileup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-genome">get_genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-genome-annotation">get_genome_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-genome-blacklist">get_genome_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-genome-support-files">get_genome_support_files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-transcripts">get_transcripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hisat2-align">hisat2_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hisat2-index">hisat2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hisat2-splice-aware-index">hisat2_splice_aware_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hmmratac">hmmratac</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hmmratac-genome-info">hmmratac_genome_info</a></li>
<li class="toctree-l3"><a class="reference internal" href="#idr">idr</a></li>
<li class="toctree-l3"><a class="reference internal" href="#infer-strandedness">infer_strandedness</a></li>
<li class="toctree-l3"><a class="reference internal" href="#insert-size-metrics">insert_size_metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kallistobus-count">kallistobus_count</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kallistobus-ref">kallistobus_ref</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kallistobus-ref-kite">kallistobus_ref_kite</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kb-seurat-pp">kb_seurat_pp</a></li>
<li class="toctree-l3"><a class="reference internal" href="#keep-mates">keep_mates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#linked-txome">linked_txome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#log-normalization">log_normalization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs2-callpeak">macs2_callpeak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs-bdgcmp">macs_bdgcmp</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs-cmbreps">macs_cmbreps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mark-duplicates">mark_duplicates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mean-center">mean_center</a></li>
<li class="toctree-l3"><a class="reference internal" href="#merge-replicates">merge_replicates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#minimap2-align">minimap2_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#minimap2-index">minimap2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mt-nuc-ratio-calculator">mt_nuc_ratio_calculator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multibamsummary">multiBamSummary</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-explain">multiqc_explain</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-filter-buttons">multiqc_filter_buttons</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-samplesconfig">multiqc_samplesconfig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-schema">multiqc_schema</a></li>
<li class="toctree-l3"><a class="reference internal" href="#narrowpeak-summit">narrowpeak_summit</a></li>
<li class="toctree-l3"><a class="reference internal" href="#onehot-peaks">onehot_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#peak-bigpeak">peak_bigpeak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotcorrelation">plotCorrelation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotfingerprint">plotFingerprint</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotheatmap-peak">plotHeatmap_peak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotpca">plotPCA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotprofile-gene">plotProfile_gene</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prepare-dexseq-annotation">prepare_DEXseq_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#quantile-normalization">quantile_normalization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#random-subset-peaks">random_subset_peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rename-sample">rename_sample</a></li>
<li class="toctree-l3"><a class="reference internal" href="#run2sra">run2sra</a></li>
<li class="toctree-l3"><a class="reference internal" href="#runs2sample">runs2sample</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-decoy-aware-index">salmon_decoy_aware_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-index">salmon_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-quant">salmon_quant</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sam2bam">sam2bam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sambamba-sort">sambamba_sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-index">samtools_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-index-cram">samtools_index_cram</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-presort">samtools_presort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-sort">samtools_sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-stats">samtools_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="#setup-blacklist">setup_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sieve-bam">sieve_bam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#twobit">twobit</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="schemas.html">All configurable options</a></li>
<li class="toctree-l2"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
<li class="toctree-l2"><a class="reference internal" href="../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="extensive_docs.html">Extensive docs</a> &raquo;</li>
      <li>Per rule explanation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/all_rules.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="per-rule-explanation">
<h1>Per rule explanation<a class="headerlink" href="#per-rule-explanation" title="Permalink to this headline">¶</a></h1>
<p>This is an automatically generated list of all supported rules, their docstrings, and command. At the start of each workflow run a list is printed of which rules will be run. And while the workflow is running it prints which rules are being started and finished. This page is here to give an explanation to the user about what each rule does, and for developers to find what is, and isn’t yet supported.</p>
<section id="bam2cram">
<h2>bam2cram<a class="headerlink" href="#bam2cram" title="Permalink to this headline">¶</a></h2>
<p>Convert a bam file to the more compressed cram format.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools view -T {input.assembly} -C {input.bam} -@ {threads} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="bam-bigwig">
<h2>bam_bigwig<a class="headerlink" href="#bam-bigwig" title="Permalink to this headline">¶</a></h2>
<p>Convert a bam file into a bigwig file.
Can output strand specific bam</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bamCoverage --bam {input.bam} --outFileName {output} {params.strand_filter} --numberOfProcessors {threads} {params.flags} --verbose &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="bedgraph-bigwig">
<h2>bedgraph_bigwig<a class="headerlink" href="#bedgraph-bigwig" title="Permalink to this headline">¶</a></h2>
<p>Convert a bedgraph file into a bigwig.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>awk -v OFS=&#39;\\t&#39; &#39;{{print $1, $2, $3, $4}}&#39; {input.bedgraph} | sed &#39;/experimental/d&#39; |
bedSort /dev/stdin {output.tmp} &gt; {log} 2&gt;&amp;1
bedGraphToBigWig {output.tmp} {input.genome_size} {output.out} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="bedgraphish-to-bedgraph">
<h2>bedgraphish_to_bedgraph<a class="headerlink" href="#bedgraphish-to-bedgraph" title="Permalink to this headline">¶</a></h2>
<p>Convert the bedgraph-ish file generated by genrich into a true bedgraph file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>splits=$(grep -Pno &quot;([^\/]*)(?=\.bam)&quot; {input})
splits=($splits)
lncnt=$(wc -l {input} | echo &quot;$(grep -Po &quot;.*(?=\ )&quot;)&quot;:)
splits+=($lncnt)

counter=1
for split in &quot;${{splits[@]::${{\#splits[@]}}-1}}&quot;;
do
    filename=$(grep -Po &quot;(?&lt;=:).*&quot; &lt;&lt;&lt; $split);
    if [[ $filename =~ {wildcards.sample} ]]; then
        startnr=$(grep -Po &quot;.*(?=\:)&quot; &lt;&lt;&lt; $split);
        endnr=$(grep -Po &quot;.*(?=\:)&quot; &lt;&lt;&lt; ${{splits[counter]}});

        lines=&quot;NR&gt;=startnr &amp;&amp; NR&lt;=endnr {{ print \$1, \$2, \$3, \$4 }}&quot;
        lines=${{lines/startnr/$((startnr + 2))}}
        lines=${{lines/endnr/$((endnr - 1))}}

        awk &quot;$lines&quot; {input} &gt; {output}
    fi
    ((counter++))
done
</pre></div>
</div>
</section>
<section id="bedtools-slop">
<h2>bedtools_slop<a class="headerlink" href="#bedtools-slop" title="Permalink to this headline">¶</a></h2>
<p>After combine_peaks we end up with just a bed file of summits. We extend all peaks
to a total width of 200, for a fair comparison between peaks.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bedtools slop -i {input.bedfile} -g {input.sizes} -b {params.slop} | uniq &gt; {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="blind-clustering">
<h2>blind_clustering<a class="headerlink" href="#blind-clustering" title="Permalink to this headline">¶</a></h2>
<p>Create a sample distance cluster heatmap,
and various correlation cluster heatmaps.</p>
<p>A (blind) variance stabilizing transformation is applied to the counts,
to reduce the impact of genes with low expression levels.</p>
</section>
<section id="bowtie2-align">
<h2>bowtie2_align<a class="headerlink" href="#bowtie2-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with bowtie2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bowtie2 {params.params} --threads {threads} -x {input.index}/{wildcards.assembly} {params.input} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</section>
<section id="bowtie2-index">
<h2>bowtie2_index<a class="headerlink" href="#bowtie2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for bowtie2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

bowtie2-build {params} --threads {threads} {input} {output}/{wildcards.assembly} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="bwa-index">
<h2>bwa_index<a class="headerlink" href="#bwa-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for bwa (mem). This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

bwa index -p {params.prefix} {params.params} {input} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="bwa-mem">
<h2>bwa_mem<a class="headerlink" href="#bwa-mem" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with bwa-mem, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bwa mem {params.params} -t {threads} {params.index_dir} {input.reads} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</section>
<section id="bwa-mem2">
<h2>bwa_mem2<a class="headerlink" href="#bwa-mem2" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with bwa-mem2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bwa-mem2 mem {params.params} -t {threads} {params.index_dir} {input.reads} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</section>
<section id="bwa-mem2-index">
<h2>bwa_mem2_index<a class="headerlink" href="#bwa-mem2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for bwa-mem2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

bwa-mem2 index -p {params.prefix} {input} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="call-peak-genrich">
<h2>call_peak_genrich<a class="headerlink" href="#call-peak-genrich" title="Permalink to this headline">¶</a></h2>
<p>Call peaks with genrich based on the pileup.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>Genrich -P -f {input.log} -o {output.narrowpeak} {params} -v &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="chipseeker">
<h2>chipseeker<a class="headerlink" href="#chipseeker" title="Permalink to this headline">¶</a></h2>
<p>Generate a multiqc header file with contact info and date of multiqc generation.</p>
</section>
<section id="combine-peaks">
<h2>combine_peaks<a class="headerlink" href="#combine-peaks" title="Permalink to this headline">¶</a></h2>
<p>Uses gimmemotifs’ combine_peaks to “combine” peaks. This finds all peaks close
together and takes the most significant one as the true peak.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>combine_peaks -i {input.summitfiles} -g {input.sizes} \
--window {params.windowsize} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="combine-qc-files">
<h2>combine_qc_files<a class="headerlink" href="#combine-qc-files" title="Permalink to this headline">¶</a></h2>
<p>Aggregate all the quality control metrics for every sample into a single
multiqc report.</p>
<p>The input can get very long (causing problems with the shell), so we have
to write the input to a file, and then we can use that file as input to
multiqc.</p>
</section>
<section id="complement-blacklist">
<h2>complement_blacklist<a class="headerlink" href="#complement-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Take the complement of the blacklist. We need this complement to tell samtools
to only keep reads that are in the complement of the blacklist.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sortBed -faidx {input.sizes} -i {input.blacklist} 2&gt;&gt; {log} |
complementBed -i stdin -g {input.sizes} &gt; {output} 2&gt;&gt; {log}
</pre></div>
</div>
</section>
<section id="computematrix-gene">
<h2>computeMatrix_gene<a class="headerlink" href="#computematrix-gene" title="Permalink to this headline">¶</a></h2>
<p>Pre-compute correlations between bams using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>computeMatrix scale-regions -S {input.bw} {params.labels} -R {params.annotation} \
-p {threads} -b 2000 -a 500 -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="computematrix-peak">
<h2>computeMatrix_peak<a class="headerlink" href="#computematrix-peak" title="Permalink to this headline">¶</a></h2>
<p>Pre-compute correlations between bams using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>computeMatrix scale-regions -S {input.bw} {params.labels} -R {input.peaks} \
-p {threads} --binSize 5 -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="count-matrix">
<h2>count_matrix<a class="headerlink" href="#count-matrix" title="Permalink to this headline">¶</a></h2>
<p>Combine count tables into one count matrix per assembly</p>
</section>
<section id="count-matrix-dexseq">
<h2>count_matrix_DEXseq<a class="headerlink" href="#count-matrix-dexseq" title="Permalink to this headline">¶</a></h2>
<p>Combine DEXSeq counts into one count matrix per assembly
for use in function <code class="docutils literal notranslate"><span class="pre">DEXSeqDataSet()</span></code></p>
</section>
<section id="coverage-table">
<h2>coverage_table<a class="headerlink" href="#coverage-table" title="Permalink to this headline">¶</a></h2>
<p>Use gimmemotif’s coverage_table to generate a count table with the nr of reads
under each peak per sample.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>echo &quot;\# The number of reads under each peak&quot; &gt; {output}
coverage_table -p {input.peaks} -d {input.replicates} 2&gt; {log} | grep -vE &quot;^\#&quot; 2&gt;&gt; {log} |
awk &#39;BEGIN {{ FS = &quot;@&quot; }} NR==1{{gsub(&quot;{wildcards.assembly}-|.samtools-coordinate&quot;,&quot;&quot;,$0)}}; \
{{print $0}}&#39; &gt;&gt; {output}
</pre></div>
</div>
</section>
<section id="cp-unsieved2sieved">
<h2>cp_unsieved2sieved<a class="headerlink" href="#cp-unsieved2sieved" title="Permalink to this headline">¶</a></h2>
<p>Copy a bam file if no sieving is necessary.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cp {input} {output}
</pre></div>
</div>
</section>
<section id="create-snap-object">
<h2>create_SNAP_object<a class="headerlink" href="#create-snap-object" title="Permalink to this headline">¶</a></h2>
<p>Create a snapobject for each BAM file.</p>
<p>These snapobjects can be merged later using snaptools in R.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>snaptools snap-pre --input-file={input.bams} --output-snap={output} --genome-name={wildcards.assembly} \
--genome-size={input.genome_size} {params.params} {params.chrm} {params.mapq} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="create-bins-snap-object">
<h2>create_bins_SNAP_object<a class="headerlink" href="#create-bins-snap-object" title="Permalink to this headline">¶</a></h2>
<p>Add a Binned genome matrix to the SNAPobject, after which it is renamed and moved
to the Snapfiles folder for downstream analysis in R using Snaptools</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>snaptools snap-add-bmat --snap-file={input} {params} &gt; {log} 2&gt;&amp;1
echo &quot;bmat added, moving file&quot; &gt;&gt; {log} 2&gt;&amp;1
mv {input} {output}    &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="cytoband">
<h2>cytoband<a class="headerlink" href="#cytoband" title="Permalink to this headline">¶</a></h2>
<p>Generate a cytoband track for each assembly</p>
<p>source: <a class="reference external" href="http://genomewiki.ucsc.edu/index.php/Assembly_Hubs#Cytoband_Track">http://genomewiki.ucsc.edu/index.php/Assembly_Hubs#Cytoband_Track</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cat {input.sizes} |
bedSort /dev/stdin /dev/stdout |
awk &#39;{{print $1,0,$2,$1,&quot;gneg&quot;}}&#39; &gt; {output.cytoband_bd}

bedToBigBed -type=bed4 {output.cytoband_bd} -as={params.schema} \
{input.sizes} {output.cytoband_bb} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="decoy-transcripts">
<h2>decoy_transcripts<a class="headerlink" href="#decoy-transcripts" title="Permalink to this headline">¶</a></h2>
<p>Generate decoy_transcripts.txt for Salmon indexing</p>
<p>script source: <a class="reference external" href="https://github.com/COMBINE-lab/SalmonTools">https://github.com/COMBINE-lab/SalmonTools</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>Generate a decoy aware transcriptome index for Salmon.
</pre></div>
</div>
</section>
<section id="deseq2">
<h2>deseq2<a class="headerlink" href="#deseq2" title="Permalink to this headline">¶</a></h2>
<p>Differential gene expression analysis with DESeq2.</p>
</section>
<section id="dexseq-count">
<h2>dexseq_count<a class="headerlink" href="#dexseq-count" title="Permalink to this headline">¶</a></h2>
<p>count exon usage</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>current_conda_env=$(conda env list | grep \* | cut -d &quot;*&quot; -f2-)
DEXseq_path=${{current_conda_env}}/lib/R/library/DEXSeq/python_scripts

python ${{DEXseq_path}}/dexseq_count.py -f bam -r pos {params.endedness} -s {params.strandedness} {input.gff} {input.bam} {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="dupradar">
<h2>dupRadar<a class="headerlink" href="#dupradar" title="Permalink to this headline">¶</a></h2>
<p>visualize fraction of artifactual reads to normal read duplication
(due to natural over-sequencing of highly expressed genes).</p>
</section>
<section id="dupradar-combine">
<h2>dupRadar_combine<a class="headerlink" href="#dupradar-combine" title="Permalink to this headline">¶</a></h2>
<p>Combine the individual images (so we can group them nicely in the MultiQC).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>convert {params.good_example} {params.bad_example} {input} -append {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="edger-normalization">
<h2>edgeR_normalization<a class="headerlink" href="#edger-normalization" title="Permalink to this headline">¶</a></h2>
<p>edgeR supports three different types of normalization: TMM, RLE, and upperquartile.</p>
<p>TMM: is the weighted trimmed mean of M-values proposed by Robinson and Oshlack (2010).
RLE: is the scaling factor method proposed by Anders and Huber (2010). DEseq2 normalisation</p>
<blockquote>
<div><p>is based on this.</p>
</div></blockquote>
<p>upperquartile: is the upper-quartile normalization method of Bullard et al (2010).</p>
<p>In addition to</p>
</section>
<section id="ena2fastq-pe">
<h2>ena2fastq_PE<a class="headerlink" href="#ena2fastq-pe" title="Permalink to this headline">¶</a></h2>
<p>Download paired-end fastq files directly from the ENA.</p>
</section>
<section id="ena2fastq-se">
<h2>ena2fastq_SE<a class="headerlink" href="#ena2fastq-se" title="Permalink to this headline">¶</a></h2>
<p>Download single-end fastq files directly from the ENA.</p>
</section>
<section id="extend-genome">
<h2>extend_genome<a class="headerlink" href="#extend-genome" title="Permalink to this headline">¶</a></h2>
<p>Append given file(s) to genome</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# extend the genome.fa
cp {input.genome} {output.genome}

for FILE in {input.extension}; do
    cat $FILE &gt;&gt; {output.genome}
done
</pre></div>
</div>
</section>
<section id="extend-genome-annotation">
<h2>extend_genome_annotation<a class="headerlink" href="#extend-genome-annotation" title="Permalink to this headline">¶</a></h2>
<p>Append given file(s) to genome annotation</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# extend the genome.annotation.gtf
cp {input.gtf} {output.gtf}

for FILE in {input.extension}; do
    cat $FILE &gt;&gt; {output.gtf}
done

\# generate an extended genome.annotation.bed
gtfToGenePred {output.gtf} {output.gp}
genePredToBed {output.gp} {output.bed}
</pre></div>
</div>
</section>
<section id="extend-genome-blacklist">
<h2>extend_genome_blacklist<a class="headerlink" href="#extend-genome-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Copy blacklist to the custom genome directory</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cp {input} {output}
</pre></div>
</div>
</section>
<section id="fastp-pe">
<h2>fastp_PE<a class="headerlink" href="#fastp-pe" title="Permalink to this headline">¶</a></h2>
<p>Automated adapter detection, adapter trimming, and quality trimming through fastp (paired-end).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastp -w {params.threads} --in1 {input[0]} --in2 {input[1]} \
--out1 {output.r1} --out2 {output.r2} -h {output.qc_html} -j {output.qc_json} \
{params.config} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="fastp-se">
<h2>fastp_SE<a class="headerlink" href="#fastp-se" title="Permalink to this headline">¶</a></h2>
<p>Automated adapter detection, adapter trimming, and quality trimming through fastp (single-end).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastp -w {params.threads} --in1 {input} --out1 {output.se} -h {output.qc_html} -j {output.qc_json} \
{params.config} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="fastp-qc-pe">
<h2>fastp_qc_PE<a class="headerlink" href="#fastp-qc-pe" title="Permalink to this headline">¶</a></h2>
<p>Get quality scores for (technical) replicates</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastp -w {threads} --in1 {input[0]} --in2 {input[1]} \
-h {output.qc_html} -j {output.qc_json} \
{params.config} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="fastp-qc-se">
<h2>fastp_qc_SE<a class="headerlink" href="#fastp-qc-se" title="Permalink to this headline">¶</a></h2>
<p>Get quality scores for (technical) replicates</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastp -w {threads} --in1 {input} -h {output.qc_html} -j {output.qc_json} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="fastq-pair">
<h2>fastq_pair<a class="headerlink" href="#fastq-pair" title="Permalink to this headline">¶</a></h2>
<p>fastq_pair re-writes paired-end fastq files to ensure that each read has a mate and
dsicards singleton reads. This step is required after scRNA trimming since we only trim the fastq
containing reads and not the barcode fastq.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>gunzip -c {input.r1} &gt; {output.intermediates1[0]} 2&gt; {log}
gunzip -c {input.r2} &gt; {output.intermediates1[1]} 2&gt;&gt; {log}
if [ {params.tused} == true ]
then
  opts=&quot;{params.options}&quot;
else
  echo &quot;\nsetting parameter t with the number of reads in the fastq\n&quot; &gt;&gt; {log}
  opts=&quot;-p -t &quot;$(wc -l {input.r1} | grep -Po &#39;^\d+&#39; | awk &#39;{{print int($1/4)}}&#39;)
fi
fastq_pair $opts {output.intermediates1} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="fastqc">
<h2>fastqc<a class="headerlink" href="#fastqc" title="Permalink to this headline">¶</a></h2>
<p>Generate quality control report for fastq files.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastqc {input} -O {params} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="featurecounts">
<h2>featureCounts<a class="headerlink" href="#featurecounts" title="Permalink to this headline">¶</a></h2>
<p>Use featureCounts to generate the fraction reads in peaks score
(frips/assigned reads). See: <a class="reference external" href="https://www.biostars.org/p/337872/">https://www.biostars.org/p/337872/</a> and
<a class="reference external" href="https://www.biostars.org/p/228636/">https://www.biostars.org/p/228636/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# Make a custom &quot;SAF&quot; file which featureCounts needs:
awk &#39;BEGIN{{FS=OFS=&quot;\t&quot;; print &quot;GeneID\tChr\tStart\tEnd\tStrand&quot;}}{{print $4, $1, $2+1, $3, &quot;.&quot;}}&#39; \
{input.peak} 1&gt; {output.tmp_saf} 2&gt; {log}

\# run featureCounts
featureCounts -T {threads} -p -a {output.tmp_saf} -F SAF -o {output.real_out} {input.bam} &gt; {log} 2&gt;&amp;1

\# move the summary to qc directory
mv $(dirname {output.real_out})/$(basename {output.summary}) {output.summary}
</pre></div>
</div>
</section>
<section id="id1">
<h2>featurecounts<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>summarize reads to gene level. Outputs a counts table per bam file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>featureCounts -a {input.gtf} {input.bam} {params.endedness} -s {params.strandedness} {params.user_flags} -T {threads} -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="gcpercent">
<h2>gcPercent<a class="headerlink" href="#gcpercent" title="Permalink to this headline">¶</a></h2>
<p>Generate a gc content track</p>
<p>source: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/gc5Base/">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/gc5Base/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>        hgGcPercent -wigOut -doGaps -file=stdout -win=5 -verbose=0 {wildcards.assembly} {input.twobit} \
| gzip &gt; {output.gcpcvar}

        wigToBigWig {output.gcpcvar} {input.sizes} {output.gcpc} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="gene-id2name">
<h2>gene_id2name<a class="headerlink" href="#gene-id2name" title="Permalink to this headline">¶</a></h2>
<p>Parse the gtf file to generate a gene_id to gene_name conversion table.</p>
</section>
<section id="genrich-pileup">
<h2>genrich_pileup<a class="headerlink" href="#genrich-pileup" title="Permalink to this headline">¶</a></h2>
<p>Generate the pileup. We do this separately from peak-calling since these two
processes have a very different computational footprint.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>input=$(echo {input.reps} | tr &#39; &#39; &#39;,&#39;)
Genrich -X -t $input -f {output.log} {params.control} -k {output.bedgraphish} \
{params.params} -v &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="get-genome">
<h2>get_genome<a class="headerlink" href="#get-genome" title="Permalink to this headline">¶</a></h2>
<p>Download a genome through genomepy.</p>
</section>
<section id="get-genome-annotation">
<h2>get_genome_annotation<a class="headerlink" href="#get-genome-annotation" title="Permalink to this headline">¶</a></h2>
<p>Download a gene annotation through genomepy.</p>
</section>
<section id="get-genome-blacklist">
<h2>get_genome_blacklist<a class="headerlink" href="#get-genome-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Download a genome blacklist for a genome, if it exists, through genomepy.</p>
</section>
<section id="get-genome-support-files">
<h2>get_genome_support_files<a class="headerlink" href="#get-genome-support-files" title="Permalink to this headline">¶</a></h2>
<p>Generate supporting files for a genome.</p>
</section>
<section id="get-transcripts">
<h2>get_transcripts<a class="headerlink" href="#get-transcripts" title="Permalink to this headline">¶</a></h2>
<p>Generate transcripts.fasta using gffread.</p>
<p>Requires genome.fa and annotation.gtf (with matching chromosome/scaffold names)</p>
</section>
<section id="hisat2-align">
<h2>hisat2_align<a class="headerlink" href="#hisat2-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with hisat2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>hisat2 {params.params} --threads {threads} -x {input.index}/part {params.input} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</section>
<section id="hisat2-index">
<h2>hisat2_index<a class="headerlink" href="#hisat2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for hisat2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

hisat2-build {params} -p {threads} {input} {output}/part &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="hisat2-splice-aware-index">
<h2>hisat2_splice_aware_index<a class="headerlink" href="#hisat2-splice-aware-index" title="Permalink to this headline">¶</a></h2>
<p>Make an exon-junction and splice aware index for hisat2.
This index is required for alignment and quantification of RNA-seq data.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

hp=$(which hisat2)
python3 ${{hp}}_extract_splice_sites.py {input.gtf} &gt; {output}/splice_sites.tsv 2&gt;&gt; {log}
python3 ${{hp}}_extract_exons.py {input.gtf} &gt; {output}/exons.tsv 2&gt;&gt; {log}

hisat2-build {params} -p {threads} --ss {output}/splice_sites.tsv --exon {output}/exons.tsv \
{input.fasta} {output}/part &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="hmmratac">
<h2>hmmratac<a class="headerlink" href="#hmmratac" title="Permalink to this headline">¶</a></h2>
<p>Call gappedpeaks with HMMRATAC.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>HMMRATAC --bedgraph true -o {params.basename} {params.hmmratac_params} -Xmx22G -b {input.bam} -i {input.bam_index} -g {input.genome_size} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="hmmratac-genome-info">
<h2>hmmratac_genome_info<a class="headerlink" href="#hmmratac-genome-info" title="Permalink to this headline">¶</a></h2>
<p>Generate the ‘genome info’ that hmmratac requires for peak calling.
<a class="reference external" href="https://github.com/LiuLabUB/HMMRATAC/issues/17">https://github.com/LiuLabUB/HMMRATAC/issues/17</a></p>
<p>TODO: isnt this just .fa.sizes?</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools view -H {input} 2&gt;  {log} | grep SQ 2&gt;&gt; {log} | cut -f 2-3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2   2&gt;&gt; {log} | cut -f 1        &gt; {output.tmp1} 2&gt; {log}
samtools view -H {input} 2&gt;&gt; {log} | grep SQ 2&gt;&gt; {log} | cut -f 2-3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2,3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2 &gt; {output.tmp2} 2&gt; {log}
paste {output.tmp1} {output.tmp2} &gt; {output.out} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="idr">
<h2>idr<a class="headerlink" href="#idr" title="Permalink to this headline">¶</a></h2>
<p>Combine replicates based on the irreproducible discovery rate (IDR). Can only handle two replicates.
For more than two replicates use fisher’s method.</p>
<p>When combining narrowpeak files with IDR, the q-score and summit are set to -1 (means not set). However some
downstream tools require these to be set. So we set the q-score to zero, and place the summit of the peak in
the middle of the peak.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                if [ &quot;{params.nr_reps}&quot; == &quot;1&quot; ]; then
cp {input} {output.true}
touch {output.temp}
                else
idr --samples {input} {params.rank} --output-file {output.temp} &gt; {log} 2&gt;&amp;1
if [ &quot;{wildcards.ftype}&quot; == &quot;narrowPeak&quot; ]; then
    awk &#39;IFS=&quot;\t&quot;,OFS=&quot;\t&quot; {{$9=0; $10=int(($3 - $2) / 2); print}}&#39; {output.temp} &gt; {output.true} 2&gt;&gt; {log}
else
    cp {output.temp} {output.true}
fi
                fi
</pre></div>
</div>
</section>
<section id="infer-strandedness">
<h2>infer_strandedness<a class="headerlink" href="#infer-strandedness" title="Permalink to this headline">¶</a></h2>
<p>use RSeqQC’s infer_experiment.py to determine strandedness af a sample</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>infer_experiment.py -i {input.bam} -r {input.bed} -q {params} 2&gt; {log} | awk NF &gt; {output}
</pre></div>
</div>
</section>
<section id="insert-size-metrics">
<h2>insert_size_metrics<a class="headerlink" href="#insert-size-metrics" title="Permalink to this headline">¶</a></h2>
<p>Get insert size metrics from a (paired-end) bam. This score is then used by
MultiQC in the report.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>picard CollectInsertSizeMetrics INPUT={input} \
OUTPUT={output.tsv} H={output.pdf} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="kallistobus-count">
<h2>kallistobus_count<a class="headerlink" href="#kallistobus-count" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a transcriptome (index) with kallistobus and output a quantification file per sample.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>kb count \
-i {params.basename}.idx \
-t {threads} -g {params.basename}_t2g.txt \
-o {params.outdir} -c1 {params.basename}_cdna_t2c.txt -c2 {params.basename}_intron_t2c.txt \
{params.barcode_arg} {params.options} {input.reads} &gt; {log} 2&gt;&amp;1
\# Validate output
if grep -q &#39;ERROR\|bad_alloc&#39; &quot;{log}&quot;; then
  exit 1
fi
</pre></div>
</div>
</section>
<section id="kallistobus-ref">
<h2>kallistobus_ref<a class="headerlink" href="#kallistobus-ref" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for kallistobus. This index is required for counting.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>kb ref \
{input.fa} {input.gtf} \
-i {params.basename}.idx -g {params.basename}_t2g.txt -f1 {params.basename}_cdna.fa \
-f2 {params.basename}_intron.fa \
-c1 {params.basename}_cdna_t2c.txt -c2 {params.basename}_intron_t2c.txt \
{params.options} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="kallistobus-ref-kite">
<h2>kallistobus_ref_kite<a class="headerlink" href="#kallistobus-ref-kite" title="Permalink to this headline">¶</a></h2>
<p>Make a mismatch index for kallistobus. This index is required to count feature barcodes, such as antibody tags.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {params.basename}
kb ref  \
{input.featurebarcodes} \
{params.options} \
-i {params.basename}.idx -g {params.basename}_t2g.txt -f1 {params.basename}_cdna.fa &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="kb-seurat-pp">
<h2>kb_seurat_pp<a class="headerlink" href="#kb-seurat-pp" title="Permalink to this headline">¶</a></h2>
<p>summarize reads to gene level. Outputs a counts table per bam file.</p>
</section>
<section id="keep-mates">
<h2>keep_mates<a class="headerlink" href="#keep-mates" title="Permalink to this headline">¶</a></h2>
<p>In-house script that, after alignment, removes the information that reads are paired.
This can be beneficial when peak calling with macs2 when shifting + extending, since
macs2 in this case only keeps the first in pair.</p>
</section>
<section id="linked-txome">
<h2>linked_txome<a class="headerlink" href="#linked-txome" title="Permalink to this headline">¶</a></h2>
<p>Generate a linked transcriptome for tximeta</p>
<p>Also creates a symlink to the gtf in an Ensembl format (required by tximeta)</p>
<p>Required to converting salmon output (estimated transcript abundances) to gene counts</p>
</section>
<section id="log-normalization">
<h2>log_normalization<a class="headerlink" href="#log-normalization" title="Permalink to this headline">¶</a></h2>
<p>Log1p normalization of a count table.</p>
</section>
<section id="macs2-callpeak">
<h2>macs2_callpeak<a class="headerlink" href="#macs2-callpeak" title="Permalink to this headline">¶</a></h2>
<p>Call peaks using macs2.
Macs2 requires a genome size, which we estimate from the amount of unique kmers of the average read length.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>echo &quot;preparing to run unique-kmers.py with -k $kmer_size&quot; &gt;&gt; {log}
GENSIZE=$(unique-kmers.py {params.genome} -k $kmer_size --quiet 2&gt;&amp;1 | grep -P -o &#39;(?&lt;=\.fa: ).*&#39;)
echo &quot;kmer size: $kmer_size, and effective genome size: $GENSIZE&quot; &gt;&gt; {log}

\# call peaks
macs2 callpeak --bdg -t {input.bam} {params.control} --outdir {config[result_dir]}/macs2/ -n {wildcards.assembly}-{wildcards.sample} \
{params.macs_params} -g $GENSIZE -f {params.format} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="macs-bdgcmp">
<h2>macs_bdgcmp<a class="headerlink" href="#macs-bdgcmp" title="Permalink to this headline">¶</a></h2>
<p>Prepare p-value files for rule macs_cmbreps</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>macs2 bdgcmp -t {input.treatment} -c {input.control} -m qpois -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="macs-cmbreps">
<h2>macs_cmbreps<a class="headerlink" href="#macs-cmbreps" title="Permalink to this headline">¶</a></h2>
<p>Combine biological replicates through Fisher’s method</p>
<p>(Link original peakfile in replicate_processed if there is only 1 sample for a condition)</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                    if [ &quot;{params.nr_reps}&quot; == &quot;1&quot; ]; then
touch {output.tmpbdg} {output.tmppeaks}
mkdir -p $(dirname {output.peaks}); cp {input.treatment} {output.peaks}
                    else
macs2 cmbreps -i {input.bdgcmp} -o {output.tmpbdg} -m fisher &gt; {log} 2&gt;&amp;1
macs2 {params.function} {params.config} -i {output.tmpbdg} -o {output.tmppeaks} &gt;&gt; {log} 2&gt;&amp;1
cat {output.tmppeaks} | tail -n +2 &gt; {output.peaks}
                    fi
</pre></div>
</div>
</section>
<section id="mark-duplicates">
<h2>mark_duplicates<a class="headerlink" href="#mark-duplicates" title="Permalink to this headline">¶</a></h2>
<p>Mark all duplicate reads in a bam file with picard MarkDuplicates</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# use the TMPDIR if set, and not given in the config
if [[ ${{TMPDIR:=F}} == &quot;F&quot; ]] || [[ &quot;{params}&quot; == *TMP_DIR* ]]
then
    tmpdir=&quot;&quot;
else
    tmpdir=TMP_DIR=$TMPDIR
fi

picard MarkDuplicates $tmpdir {params} INPUT={input} \
OUTPUT={output.bam} METRICS_FILE={output.metrics} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="mean-center">
<h2>mean_center<a class="headerlink" href="#mean-center" title="Permalink to this headline">¶</a></h2>
<p>Mean centering of a count table.</p>
</section>
<section id="merge-replicates">
<h2>merge_replicates<a class="headerlink" href="#merge-replicates" title="Permalink to this headline">¶</a></h2>
<p>Merge replicates (fastqs) simply by concatenating the files. We also change the name of the read headers to
contain the name of the original replicate.</p>
<p>Must happen after trimming due to trim-galore’s automatic adapter trimming method</p>
<p>If a replicate has only 1 sample in it, simply move the file.</p>
</section>
<section id="minimap2-align">
<h2>minimap2_align<a class="headerlink" href="#minimap2-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with minimap2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>minimap2 -t {threads} -a {input.index} {input.reads} {params} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="minimap2-index">
<h2>minimap2_index<a class="headerlink" href="#minimap2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for minimap2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p $(dirname {output})

minimap2 -t {threads} -d {output} {input} {params} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="mt-nuc-ratio-calculator">
<h2>mt_nuc_ratio_calculator<a class="headerlink" href="#mt-nuc-ratio-calculator" title="Permalink to this headline">¶</a></h2>
<p>Estimate the amount of nuclear and mitochondrial reads in a sample. This metric
is especially important in ATAC-seq experiments where mitochondrial DNA can
be overrepresented. Reads are classified as mitochondrial reads if they map
against either “chrM” or “MT”.</p>
<p>These values are aggregated and displayed in the MultiQC report.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mtnucratio {input.bam} {params.mitochondria}
</pre></div>
</div>
</section>
<section id="multibamsummary">
<h2>multiBamSummary<a class="headerlink" href="#multibamsummary" title="Permalink to this headline">¶</a></h2>
<p>Pre-compute a bam summary with deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>multiBamSummary bins --bamfiles {input.bams} -out {output} {params.names} \
{params.params} -p {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="multiqc-explain">
<h2>multiqc_explain<a class="headerlink" href="#multiqc-explain" title="Permalink to this headline">¶</a></h2>
<p>Generate rename buttons for the multiqc report.</p>
</section>
<section id="multiqc-filter-buttons">
<h2>multiqc_filter_buttons<a class="headerlink" href="#multiqc-filter-buttons" title="Permalink to this headline">¶</a></h2>
<p>Generate filter buttons.</p>
</section>
<section id="multiqc-samplesconfig">
<h2>multiqc_samplesconfig<a class="headerlink" href="#multiqc-samplesconfig" title="Permalink to this headline">¶</a></h2>
<p>Add a section in the multiqc report that reports the samples.tsv and config.yaml</p>
</section>
<section id="multiqc-schema">
<h2>multiqc_schema<a class="headerlink" href="#multiqc-schema" title="Permalink to this headline">¶</a></h2>
<p>Generate a multiqc config schema. Used for the ordering of modules.</p>
</section>
<section id="narrowpeak-summit">
<h2>narrowpeak_summit<a class="headerlink" href="#narrowpeak-summit" title="Permalink to this headline">¶</a></h2>
<p>Convert a narrowpeak file to a “macs2 summits” file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>awk &#39;BEGIN {{OFS=&quot;\t&quot;}} {{ print $1,$2+$10,$2+$10+1,$4,$9; }}&#39; {input} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</section>
<section id="onehot-peaks">
<h2>onehot_peaks<a class="headerlink" href="#onehot-peaks" title="Permalink to this headline">¶</a></h2>
<p>Get onehot encodings of which peaks are found in which samples</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>awk &#39;{{print $1&quot;:&quot;$2&quot;-&quot;$3}}&#39; {input.combinedpeaks} &gt; {output.real} 2&gt; {log}

for brep in {input.narrowpeaks}
do
    bedtools sort -i $brep 2&gt; {log} |
    bedtools intersect -a {input.combinedpeaks} -b stdin -c 2&gt; {log} |
    awk &#39;{{print $4}}&#39; 2&gt; {log} |
    paste {output.real} - &gt; {output.tmp}
    mv {output.tmp} {output.real}
done

echo -e &quot;\# onehot encoding of which condition contains which peaks\\n$(cat {output.real})&quot; &gt; {output.tmp} &amp;&amp; cp {output.tmp} {output.real}
</pre></div>
</div>
</section>
<section id="peak-bigpeak">
<h2>peak_bigpeak<a class="headerlink" href="#peak-bigpeak" title="Permalink to this headline">¶</a></h2>
<p>Convert a narrowpeak file into a bignarrowpeak file.
<a class="reference external" href="https://genome-source.gi.ucsc.edu/gitlist/kent.git/tree/master/src/hg/lib/">https://genome-source.gi.ucsc.edu/gitlist/kent.git/tree/master/src/hg/lib/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# keep first 10 columns, idr adds extra columns we do not need for our bigpeak
cut -d$&#39;\t&#39; -f 1-{params.columns} {input.narrowpeak} |
awk -v OFS=&quot;\t&quot; &#39;{{$5=$5&gt;1000?1000:$5}} {{print}}&#39; |
bedSort /dev/stdin {output.tmp} &gt; {log} 2&gt;&amp;1

bedToBigBed -type={params.type} -as={params.schema} {output.tmp} \
{input.genome_size} {output.out} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="plotcorrelation">
<h2>plotCorrelation<a class="headerlink" href="#plotcorrelation" title="Permalink to this headline">¶</a></h2>
<p>Calculate the correlation between bams with deepTools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotCorrelation --corData {input} --plotFile {output} -c {wildcards.method} \
-p heatmap --plotTitle {params.title} {params.params} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="plotfingerprint">
<h2>plotFingerprint<a class="headerlink" href="#plotfingerprint" title="Permalink to this headline">¶</a></h2>
<p>Plot the “fingerprint” of your bams, using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotFingerprint -b {input.bams} {params} --outRawCounts {output} -p {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="plotheatmap-peak">
<h2>plotHeatmap_peak<a class="headerlink" href="#plotheatmap-peak" title="Permalink to this headline">¶</a></h2>
<p>Plot the peak heatmap using deepTools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotHeatmap --matrixFile {input} --outFileName {output.img} {params.params} --startLabel=&quot;-{params.slop}&quot; --endLabel={params.slop} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="plotpca">
<h2>plotPCA<a class="headerlink" href="#plotpca" title="Permalink to this headline">¶</a></h2>
<p>Plot a PCA between bams using deepTools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotPCA --corData {input} --outFileNameData {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="plotprofile-gene">
<h2>plotProfile_gene<a class="headerlink" href="#plotprofile-gene" title="Permalink to this headline">¶</a></h2>
<p>Plot the gene profile using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotProfile -m {input} --outFileName {output.img} --outFileNameData {output.file} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="prepare-dexseq-annotation">
<h2>prepare_DEXseq_annotation<a class="headerlink" href="#prepare-dexseq-annotation" title="Permalink to this headline">¶</a></h2>
<p>generate a DEXseq annotation.gff from the annotation.gtf</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>current_conda_env=$(conda env list | grep \* | cut -d &quot;*&quot; -f2-)
DEXseq_path=${{current_conda_env}}/lib/R/library/DEXSeq/python_scripts

python ${{DEXseq_path}}/dexseq_prepare_annotation.py {input} {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="quantile-normalization">
<h2>quantile_normalization<a class="headerlink" href="#quantile-normalization" title="Permalink to this headline">¶</a></h2>
<p>Quantile normalization is a type of normalization that makes the distribution
between samples identical. This means that the actual count distribution
within a sample changes. After quantile normalization, samples are CPM
normalized (which has no effect between and within samples), for a fair
comparison between normalisation techniques.</p>
<p>See wikipedia:
<a class="reference external" href="https://en.wikipedia.org/wiki/Quantile_normalization">https://en.wikipedia.org/wiki/Quantile_normalization</a></p>
</section>
<section id="random-subset-peaks">
<h2>random_subset_peaks<a class="headerlink" href="#random-subset-peaks" title="Permalink to this headline">¶</a></h2>
<p>Take a random subset of all summits, and make them all same width.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>shuf -n {wildcards.nrpeaks} {input.peaks} | bedtools slop -i stdin -g {input.sizes} -b {params.slop} &gt; {output}
</pre></div>
</div>
</section>
<section id="rename-sample">
<h2>rename_sample<a class="headerlink" href="#rename-sample" title="Permalink to this headline">¶</a></h2>
<p>Rename local samples with incompatible naming formats</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mv {input[0]} {output[0]}
</pre></div>
</div>
</section>
<section id="run2sra">
<h2>run2sra<a class="headerlink" href="#run2sra" title="Permalink to this headline">¶</a></h2>
<p>Download the SRA of a sample by its unique identifier.</p>
<p>Tries first downloading with the faster ascp protocol, if that fails it
falls back on the slower http protocol.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# move to output dir since somehow prefetch sometimes puts files in the cwd...
\# and remove the top level folder since prefetch will assume we are done otherwise
mkdir -p {params.outdir}; cd {params.outdir}; rm -r {wildcards.run}

\# three attempts
for i in {{1..3}}
do
    \# acquire a lock
    (
        flock --timeout 30 200 || continue
        sleep 2
    ) 200&gt;{pysradb_cache_lock}

    \# dump
    prefetch --max-size 999999999999 --output-directory ./ --log-level debug --progress {wildcards.run} &gt;&gt; {log} 2&gt;&amp;1 &amp;&amp; break
    sleep 10
done

\# TODO: this is the strangest bug, in that on some machines (ocimum) prefetch downloads
\# to a different location. Not sure what causes this, but this should fix that. Could simply
\# be a global setting that we haven&#39;t discovered yet...
\# bug report: https://github.com/ncbi/sra-tools/issues/533
if [[ -f &quot;{params.outdir}/{wildcards.run}.sra&quot; ]]; then
    mkdir {params.outdir}/{wildcards.run}
    mv {params.outdir}/{wildcards.run}.sra {output}
fi
</pre></div>
</div>
</section>
<section id="runs2sample">
<h2>runs2sample<a class="headerlink" href="#runs2sample" title="Permalink to this headline">¶</a></h2>
<p>Concatenate a single run or multiple runs together into a fastq</p>
</section>
<section id="salmon-decoy-aware-index">
<h2>salmon_decoy_aware_index<a class="headerlink" href="#salmon-decoy-aware-index" title="Permalink to this headline">¶</a></h2>
<p>Generate a decoy aware transcriptome index for Salmon.</p>
</section>
<section id="salmon-index">
<h2>salmon_index<a class="headerlink" href="#salmon-index" title="Permalink to this headline">¶</a></h2>
<p>Generate a transcriptome index for Salmon.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p {output}

salmon index -t {input} -i {output} {params} --threads {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="salmon-quant">
<h2>salmon_quant<a class="headerlink" href="#salmon-quant" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a transcriptome (index) with Salmon (mapping-based mode) and output a quantification file per sample.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>salmon quant -i {input.index} -l A {params.input} {params.params} -o {output.dir} \
--threads {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="sam2bam">
<h2>sam2bam<a class="headerlink" href="#sam2bam" title="Permalink to this headline">¶</a></h2>
<p>Convert a file in sam format into a bam format (binary)</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools view -b {input} &gt; {output}
</pre></div>
</div>
</section>
<section id="sambamba-sort">
<h2>sambamba_sort<a class="headerlink" href="#sambamba-sort" title="Permalink to this headline">¶</a></h2>
<p>Sort the result of alignment or sieving with the sambamba sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sambamba view --nthreads {threads} -f bam  {input} -o /dev/stdout  2&gt; {log} |
sambamba sort --nthreads {threads} {params} /dev/stdin -o {output}  2&gt; {log}
</pre></div>
</div>
</section>
<section id="samtools-index">
<h2>samtools_index<a class="headerlink" href="#samtools-index" title="Permalink to this headline">¶</a></h2>
<p>Create an index of a bam file which can be used for e.g. visualization.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools index {params} {input} {output}
</pre></div>
</div>
</section>
<section id="samtools-index-cram">
<h2>samtools_index_cram<a class="headerlink" href="#samtools-index-cram" title="Permalink to this headline">¶</a></h2>
<p>Generate the index for a cram file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools index {input} {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</section>
<section id="samtools-presort">
<h2>samtools_presort<a class="headerlink" href="#samtools-presort" title="Permalink to this headline">¶</a></h2>
<p>(Pre)sort the result of alignment with the samtools sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# we set this trap to remove temp files when prematurely ending the rule
trap &quot;rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam&quot; INT;
rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam 2&gt; {log}

samtools sort -@ {threads} {params.memory} {input} -o {output} \
-T {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp 2&gt; {log}
</pre></div>
</div>
</section>
<section id="samtools-sort">
<h2>samtools_sort<a class="headerlink" href="#samtools-sort" title="Permalink to this headline">¶</a></h2>
<p>Sort the result of shiftsieving with the samtools sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# we set this trap to remove temp files when prematurely ending the rule
trap &quot;rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam&quot; INT;
rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam 2&gt; {log}

samtools sort {params.sort_order} -@ {threads} {params.memory} {input} -o {output} \
-T {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp 2&gt; {log}
</pre></div>
</div>
</section>
<section id="samtools-stats">
<h2>samtools_stats<a class="headerlink" href="#samtools-stats" title="Permalink to this headline">¶</a></h2>
<p>Get general stats from bam files like percentage mapped.</p>
</section>
<section id="setup-blacklist">
<h2>setup_blacklist<a class="headerlink" href="#setup-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Combine the encode blacklist with mitochrondrial dna depending on config.</p>
</section>
<section id="sieve-bam">
<h2>sieve_bam<a class="headerlink" href="#sieve-bam" title="Permalink to this headline">¶</a></h2>
<p>“Sieve” a bam.</p>
<p>This rule does (at least) one of these:</p>
<ul class="simple">
<li><p>filtering on minimum mapping quality</p></li>
<li><p>tn5 shift adjustment</p></li>
<li><p>remove multimappers</p></li>
<li><p>remove reads inside the blacklist</p></li>
<li><p>remove duplicates</p></li>
<li><p>filter paired-end reads on transcript length
<a href="#id2"><span class="problematic" id="id3">``</span></a>`
samtools view -h {params.prim_align} {params.minqual} {params.blacklist} {input.bam} | {params.atacshift} {params.sizesieve}
samtools view -b &gt; {output.final} 2&gt; {log}</p></li>
</ul>
<p># single-end reads never output allsizes so just touch the file when filtering on size
{params.sizesieve_touch}</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### softmask_track_1
Generate a track of all softmasked regions

source: https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py

#### softmask_track_2
Generate a track of all softmasked regions

source: https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py
</pre></div>
</div>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bedSort {input.mask_unsorted} {output.maskbed} &gt;&gt; {log} 2&gt;&amp;1

bedToBigBed -type=bed3 {output.maskbed} {input.sizes} {output.mask} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### sra2fastq_PE
Downloaded (raw) SRAs are converted to paired-end fastq files.
Forward and reverse samples will be switched if forward/reverse names are not lexicographically ordered.
</pre></div>
</div>
<p># move to output dir since somehow parallel-fastq-dump sometimes puts files in the cwd…
mkdir -p {output.tmpdir}; cd {output.tmpdir}</p>
<p># acquire the lock
(</p>
<blockquote>
<div><p>flock –timeout 30 200 || exit 1
sleep 3</p>
</div></blockquote>
<p>) 200&gt;{pysradb_cache_lock}</p>
<p># dump to tmp dir
parallel-fastq-dump -s {input} -O {output.tmpdir} –threads {threads} –split-e –skip-technical –dumpbase –readids –clip –read-filter pass –defline-seq <a class="reference external" href="mailto:'&#37;&#52;&#48;$ac&#46;$si&#46;$sg/$ri">‘<span>&#64;</span>$ac<span>&#46;</span>$si<span>&#46;</span>$sg/$ri</a>’ –defline-qual ‘+’ –gzip &gt;&gt; {log} 2&gt;&amp;1</p>
<p># rename file and move to output dir
mv {output.tmpdir}/<em>_1</em> {output.fastq[0]}
mv {output.tmpdir}/<em>_2</em> {output.fastq[1]}</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### sra2fastq_SE
Downloaded (raw) SRAs are converted to single-end fastq files.
</pre></div>
</div>
<p># move to output dir since somehow parallel-fastq-dump sometimes puts files in the cwd…
mkdir -p {output.tmpdir}; cd {output.tmpdir}</p>
<p># acquire the lock
(</p>
<blockquote>
<div><p>flock –timeout 30 200 || exit 1
sleep 3</p>
</div></blockquote>
<p>) 200&gt;{pysradb_cache_lock}</p>
<p># dump to tmp dir
parallel-fastq-dump -s {input} -O {output.tmpdir} –threads {threads} –split-spot –skip-technical –dumpbase –readids –clip –read-filter pass –defline-seq <a class="reference external" href="mailto:'&#37;&#52;&#48;$ac&#46;$si&#46;$sg/$ri">‘<span>&#64;</span>$ac<span>&#46;</span>$si<span>&#46;</span>$sg/$ri</a>’ –defline-qual ‘+’ –gzip &gt;&gt; {log} 2&gt;&amp;1</p>
<p># rename file and move to output dir
mv {output.tmpdir}/<a href="#id4"><span class="problematic" id="id5">*</span></a>.fastq.gz {output.fastq}</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### star_align
Align reads against a genome (index) with STAR, and pipe the output to the required sorter(s).
</pre></div>
</div>
<p>trap “find {log} -type f ! -name Log* -exec rm {{}} ;” EXIT
mkdir -p {log}
mkdir -p {output.dir}</p>
<p>STAR –genomeDir {input.index} –readFilesIn {params.input} –readFilesCommand gunzip -c –outSAMtype BAM Unsorted –outStd BAM_Unsorted –outFileNamePrefix {log}/ –outTmpDir {output.dir}/STARtmp –runThreadN {threads} {params.params} &gt; {output.pipe} 2&gt; {log}/Log.stderr.out</p>
<p># move all non-log files to output directory (this way the log files are kept on error)
find {log} -type f ! -name Log* -exec mv {{}} {output.dir} ;</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### star_index
Make a genome index for STAR.

Troubleshooting:
1. sufficient RAM &amp; disk space?
2. increase the RAM available (--limitGenomeGenerateRAM)
3. reduce the number of threads (seq2science -j 5)
4. reduce accuracy (--genomeSAsparseD 2)

In your config.yaml, add:
 aligner:
  star:
   index: --limitGenomeGenerateRAM 60000000000 --genomeSAsparseD 1
</pre></div>
</div>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>function log2 {{
</pre></div>
</div>
<p>local x=0
for (( y=$1-1 ; $y &gt; 0; y &gt;&gt;= 1 )) ; do</p>
<blockquote>
<div><p>let x=$x+1</p>
</div></blockquote>
<p>done
echo $x</p>
<blockquote>
<div><p>}}</p>
</div></blockquote>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>\# set genome dependent variables
NBits=&quot;&quot;
NBases=&quot;&quot;
GenomeLength=$(awk -F&quot;\t&quot; &#39;{{x+=$2}}END{{printf &quot;%i&quot;, x}}&#39; {input.sizefile})
NumberOfReferences=$(awk &#39;END{{print NR}}&#39; {input.sizefile})
if [ $NumberOfReferences -gt 5000 ]; then
    \# for large genomes, --genomeChrBinNbits should be scaled to min(18,log2[max(GenomeLength/NumberOfReferences,ReadLength)])
    \# ReadLength is skipped here, as it is unknown
    LpR=$(log2 $((GenomeLength / NumberOfReferences)))
    NBits=&quot;--genomeChrBinNbits $(($LpR&lt;18 ? $LpR : 18))&quot;
    printf &quot;NBits: $NBits\n\n&quot; &gt;&gt; {log} 2&gt;&amp;1
fi

if [ $GenomeLength -lt 268435456 ]; then
    \# for small genomes, --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2-1)
    logG=$(( $(log2 $GenomeLength) / 2 - 1 ))
    NBases=&quot;--genomeSAindexNbases $(( $logG&lt;14 ? $logG : 14 ))&quot;
    printf &quot;NBases: $NBases\n\n&quot; &gt;&gt; {log} 2&gt;&amp;1
fi

mkdir -p {output}

STAR --runMode genomeGenerate --genomeFastaFiles {input.genome} --sjdbGTFfile {input.gtf} \
--genomeDir {output} --outFileNamePrefix {output}/ \
--runThreadN {threads} $NBits $NBases {params} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### trackhub
Generate a UCSC track hub/assembly hub.

To view the hub, the output directory must be hosted on an web accessible location,
and the location of the hub.txt file given the UCSC genome browser at
My Data &gt; Track Hubs &gt; My Hubs

#### trackhub_index
Generate a searchable annotation &amp; index for each assembly

sources: https://genome.ucsc.edu/goldenPath/help/hubQuickStartSearch.html,
https://www.biostars.org/p/272649/
</pre></div>
</div>
<p># generate annotation files
gtfToGenePred -allErrors -ignoreGroupsWithoutExons -geneNameAsName2 -genePredExt {input.gtf} {output.genePred} -infoOut={output.info} &gt;&gt; {log} 2&gt;&amp;1</p>
<p># check if gtf has gene_names &gt; if l!=0, use gene_name, else transcript_id
l=$(head -n 100 {input.gtf} | (grep gene_name || true) | wc -l)</p>
<p># switch columns 1 (transcript_id) and 12 (gene_name)
awk -v len=$l ‘BEGIN {{ FS = “t” }}; {{ if(len!=”0”) {{ t = $1; $1 = $12; $12 = t; print; }} else {{ print; }} }}’ {output.genePred} &gt; {output.genePrednamed}</p>
<p># remove lines with missing headers<span class="raw-html-m2r"><br></span>
grep -v “^ ” {output.genePrednamed} &gt; {output.genePred}</p>
<p>genePredToBed {output.genePred} {output.genePredbed} &gt;&gt; {log} 2&gt;&amp;1</p>
<p>bedSort {output.genePredbed} {output.genePredbed} &gt;&gt; {log} 2&gt;&amp;1</p>
<p>bedToBigBed -extraIndex=name {output.genePredbed} {input.sizes} {output.genePredbigbed} &gt;&gt; {log} 2&gt;&amp;1</p>
<p># generate searchable indexes (by 1: transcriptID, 2: geneId, 8: proteinID, 9: geneName, 10: transcriptName)
grep -v “^#” {output.info} | awk -v len=$l ‘BEGIN {{ FS = “t” }} ; {{ if(len!=”0”) {{print $9, $1, $2, $8, $9, $10}} else {{print $1, $1, $2, $8, $9, $10}} }}’ &gt; {output.indexinfo}</p>
<p>ixIxx {output.indexinfo} {output.ix} {output.ixx} &gt;&gt; {log} 2&gt;&amp;1</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### trimgalore_PE
Automated adapter detection, adapter trimming, and quality trimming through trim galore (paired-end).
</pre></div>
</div>
<p>trim_galore –paired -j {threads} {params.config} -o $(dirname {output.r1}) {input.r1} {input.r2} &gt; {log} 2&gt;&amp;1</p>
<p># now rename to proper output
for f in $(find “$(dirname {output.r1})/” -name “{wildcards.sample}<a href="#id6"><span class="problematic" id="id7">**</span></a>val*<a href="#id8"><span class="problematic" id="id9">*</span></a>.fq.gz”); do</p>
<blockquote>
<div><p>mv “$f” “$(echo “$f” | sed s/.fq/.{params.fqsuffix}/)”; done</p>
</div></blockquote>
<dl class="simple">
<dt>for f in $(find “$(dirname {output.r1})/” -maxdepth 1 -name “{wildcards.sample}_*val_*.{params.fqsuffix}.gz”); do</dt><dd><p>mv “$f” “$(echo “$f” | sed s/<em>val</em>./_trimmed/)”; done</p>
</dd>
</dl>
<p># move the trimming reports to qc directory
for f in $(find “$(dirname {output.r1})/” -name “{wildcards.sample}_*.{params.fqsuffix}.gz_trimming_report.txt”); do</p>
<blockquote>
<div><p>mv “$f” “$(dirname {output.qc[0]})/$(basename $f)”; done</p>
</div></blockquote>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>#### trimgalore_SE
Automated adapter detection, adapter trimming, and quality trimming through trim galore (single-end).
</pre></div>
</div>
<p>trim_galore -j {threads} {params.config} -o $(dirname {output.se}) {input} &gt; {log} 2&gt;&amp;1</p>
<p># now rename to proper output
if [ “{params.fqsuffix}” != “fq” ]; then</p>
<blockquote>
<div><p>mv “$(dirname {output.se})/{wildcards.sample}_trimmed.fq.gz” {output.se}</p>
</div></blockquote>
<p>fi</p>
<p># move the trimming report to qc directory
report=$(dirname {output.se})/{wildcards.sample}.{params.fqsuffix}.gz_trimming_report.txt
mv $report {output.qc}
<a href="#id10"><span class="problematic" id="id11">``</span></a><a href="#id12"><span class="problematic" id="id13">`</span></a></p>
</section>
<section id="twobit">
<h2>twobit<a class="headerlink" href="#twobit" title="Permalink to this headline">¶</a></h2>
<p>Generate a 2bit file for each assembly</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cli.html" class="btn btn-neutral float-left" title="Command line interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="schemas.html" class="btn btn-neutral float-right" title="All configurable options" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>