

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Changelog &mdash; seq2science  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Contributing to seq2science" href="../CONTRIBUTING.html" />
    <link rel="prev" title="All configurable options" href="schemas.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="extensive_docs.html">Extensive docs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="all_rules.html">Per rule explanation</a></li>
<li class="toctree-l2"><a class="reference internal" href="schemas.html">All configurable options</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Changelog</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Unreleased</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">0.5.0 - 2021-03-03</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#changed">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#added">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fixed">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id5">0.4.3 - 2021-01-26</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id6">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id7">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id9">0.4.2 - 2021-01-19</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id10">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id11">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id13">0.4.1 - 2020-12-18</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id14">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id15">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id17">0.4.0 - 2020-12-11</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id18">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id19">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id20">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id22">0.3.2 - 2020-11-26</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id23">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id24">Changed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id26">0.3.1 - 2020-11-16</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id27">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id28">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id30">0.3.1 - 2020-11-05</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id31">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id32">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id33">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id35">0.3.0 - 2020-09-22</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id36">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id37">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id38">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id40">0.2.3 - 2020-09-01</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id41">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id42">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id44">0.2.2 - 2020-08-24</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id45">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id46">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id47">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id49">0.2.1 - 2020-08-10</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id50">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id51">Fixed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id52">Changed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id54">0.2.0 - 2020-08-04</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id55">Fixed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id56">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id57">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id58">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id60">0.1.0 - 2020-07-15</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id61">Added</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id62">Changed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id63">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id65">0.0.3 - 2020-07-01</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id66">Fixed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id68">0.0.2 - 2020-06-29</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id69">Fixed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id70">Changed</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id72">0.0.1 - 2020-06-17</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id74">0.0.0 - 2020-06-11</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="extensive_docs.html">Extensive docs</a> &raquo;</li>
        
      <li>Changelog</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/CHANGELOG.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">Â¶</a></h1>
<p>All notable changes to <code class="docutils literal notranslate"><span class="pre">seq2science</span></code> will be documented in this file.</p>
<p>The format is based on <a class="reference external" href="https://keepachangelog.com/en/1.0.0/">Keep a Changelog</a>,
and this project adheres to <a class="reference external" href="https://semver.org/spec/v2.0.0.html">Semantic Versioning</a>.</p>
<p>All changed fall under either one of these types: <code class="docutils literal notranslate"><span class="pre">Added</span></code>, <code class="docutils literal notranslate"><span class="pre">Changed</span></code>, <code class="docutils literal notranslate"><span class="pre">Deprecated</span></code>, <code class="docutils literal notranslate"><span class="pre">Removed</span></code>, <code class="docutils literal notranslate"><span class="pre">Fixed</span></code> or <code class="docutils literal notranslate"><span class="pre">Security</span></code>.</p>
<div class="section" id="id1">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/master...v0.4.2">Unreleased</a><a class="headerlink" href="#id1" title="Permalink to this headline">Â¶</a></h2>
</div>
<div class="section" id="id3">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.5.0...v0.4.3">0.5.0</a> - 2021-03-03<a class="headerlink" href="#id3" title="Permalink to this headline">Â¶</a></h2>
<p>Version 0.5.0 brings many quality of life improvements, such as seq2science automatically inferring what needs to be re-run when changing the samples.tsv and/or the config.yaml, differential peak analysis for chip/atac workflows and tab-completion!</p>
<p>To (hopefully) clear things up we changed the way technical and biological replicates are called, now technical and biological replicate, before technical replicate and condition.</p>
<p>It is important to note that the RNA-seq workflow DOES NOT remove duplicate reads anymore as a <em>default</em>, and that the sc/bulk ATAC-seq workflows now filters reads on the nucleosome-free region as a <em>default</em>.</p>
<div class="section" id="changed">
<h3>Changed<a class="headerlink" href="#changed" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Keep all duplicate reads in RNA-seq by default</p></li>
<li><p>Slimmed down the config printed at the start of a run</p></li>
<li><p>Changed some rules into localrules when executed on a cluster</p></li>
<li><p>moved onehot peaks to counts_dir</p></li>
<li><p>DESeq2 contrasts now accept any column names</p>
<ul>
<li><p>groups still cannot contain underscores</p></li>
<li><p>no longer accepts one group name</p></li>
<li><p>more examples added to the docs!</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="added">
<h3>Added<a class="headerlink" href="#added" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>dupRadar module to analyse read duplication types in RNA-seq</p></li>
<li><p>Differential peak analysis for ATAC- and ChIP-seq!</p></li>
<li><p>Options to filter bams by minimum and maximum insert sizes (added to config of bulk/sc atac)</p></li>
<li><p>Support experiment ids for EBI ENA and DDBJ for downloading public samples</p></li>
<li><p>More robust expression handling for BUS format detection from kb-python arguments</p></li>
<li><p>Short-hand BUS syntax for indrop v1/v2</p></li>
<li><p>Seq2science now supports tab-completion</p></li>
<li><p>Seq2science now outputs a logfile in the directory it is run</p></li>
</ul>
</div>
<div class="section" id="fixed">
<h3>Fixed<a class="headerlink" href="#fixed" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>renamed more old âreplicateâ variables to the new âtechnical_replicateâ</p></li>
<li><p>minor logging tweak</p></li>
<li><p>Chipseeker now works without defining descriptive name column</p></li>
<li><p>fix bug in resources parsing of profiles</p></li>
<li><p>small bug when naming a column condition in non peak-calling workflows</p></li>
</ul>
</div>
</div>
<div class="section" id="id5">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.4.3...v0.4.2">0.4.3</a> - 2021-01-26<a class="headerlink" href="#id5" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id6">
<h3>Changed<a class="headerlink" href="#id6" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>updated tximeta to 1.6.3 and related packages to fit (now uses R 4)</p></li>
<li><p>RNA-seq: sample distance matrix font scales with number of samples (should improve readability)</p></li>
</ul>
</div>
<div class="section" id="id7">
<h3>Fixed<a class="headerlink" href="#id7" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>RNA-seq: added sample distance matrix back to MulitQC</p></li>
<li><p>RNA-seq: sample distance matrix legend fixed</p></li>
<li><p>combine peaks with biological_replicates: keep now uses the correct peaks</p></li>
</ul>
</div>
</div>
<div class="section" id="id9">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.4.2...v0.4.1">0.4.2</a> - 2021-01-19<a class="headerlink" href="#id9" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id10">
<h3>Changed<a class="headerlink" href="#id10" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Updated kb-python to 0.25.1</p></li>
<li><p>RNA-seq with Salmon will still use bam-related QC files if bams are generated (create_trackhub = True)</p></li>
</ul>
</div>
<div class="section" id="id11">
<h3>Fixed<a class="headerlink" href="#id11" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>gimmemotifs not working with newest pandas, now a fixed pandas version</p></li>
</ul>
</div>
</div>
<div class="section" id="id13">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.4.1...v0.4.0">0.4.1</a> - 2020-12-18<a class="headerlink" href="#id13" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id14">
<h3>Added<a class="headerlink" href="#id14" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>more explanations for rules</p></li>
</ul>
</div>
<div class="section" id="id15">
<h3>Fixed<a class="headerlink" href="#id15" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>custom genome annotations for single-cell RNA-seq workflow</p></li>
<li><p>trackhubs no longer looking for reversed strands if none are present</p></li>
</ul>
</div>
</div>
<div class="section" id="id17">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.4.0...v0.3.2">0.4.0</a> - 2020-12-11<a class="headerlink" href="#id17" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id18">
<h3>Added<a class="headerlink" href="#id18" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>new workflow: (BETA) single cell RNA!</p></li>
</ul>
</div>
<div class="section" id="id19">
<h3>Changed<a class="headerlink" href="#id19" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>replicate renamed to technical_replicate and condition renamed to biological_replicate</p></li>
<li><p>bwa-mem2 default aligner for genomic workflows, instead of bwa-mem</p></li>
<li><p>interactive deeptools correlation heatmaps with static dendrograms in multiqc report</p></li>
<li><p>trackhub file permissions are set to 755 so to host the files online you donât have to change those anymore</p></li>
</ul>
</div>
<div class="section" id="id20">
<h3>Fixed<a class="headerlink" href="#id20" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>bug in chip/atac trackhub generation where peaks and bigwigs used the same name, resulting in collisions and a trackhub that does not want to load</p></li>
<li><p>(literal) genome edge-case where taking the slop of peaks results in identical peaks. One of the duplicates is removed.</p></li>
<li><p>IDR should work again</p></li>
</ul>
</div>
</div>
<div class="section" id="id22">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.3.2...v0.3.1">0.3.2</a> - 2020-11-26<a class="headerlink" href="#id22" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id23">
<h3>Added<a class="headerlink" href="#id23" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>a check to see if the downloaded fastq from ENA is not empty. Related to a recent internal error (guess) at the side of ENA sending empty fastq files</p></li>
<li><p>a custom message when a rule fails, that redirect to docs</p></li>
</ul>
</div>
<div class="section" id="id24">
<h3>Changed<a class="headerlink" href="#id24" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>sample layout lookup is split up in 100âs, to avoid a jsondecodeerror which results from very long lists of samples</p></li>
<li><p>the multiqc samples &amp; config tables are generated in a script with its own environment to make base env smaller</p></li>
<li><p>keep_mates for macs2 turned into a script with ts own environment ot make the base env smaller</p></li>
<li><p>seq2science cache now respects the xdg cache</p></li>
<li><p>moved genome downloading rules into scripts instead of run directives, should result in user-friendlier errors</p></li>
</ul>
</div>
</div>
<div class="section" id="id26">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.3.1...v0.3.0">0.3.1</a> - 2020-11-16<a class="headerlink" href="#id26" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id27">
<h3>Added<a class="headerlink" href="#id27" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Added support for multiple scrna-seq platforms (Kallistobus)</p></li>
<li><p>Fastp detects the correct mate for trimming based on BUS settings.</p></li>
<li><p>Support for Kallistobus short-hand syntax.</p></li>
</ul>
</div>
<div class="section" id="id28">
<h3>Fixed<a class="headerlink" href="#id28" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>make a trackhub index when the gene_name is not present in gtf file</p></li>
<li><p>make a trackhub index when the gene_name is not present in gene all entries</p></li>
<li><p>update Salmon &amp; salmon rules</p></li>
</ul>
</div>
</div>
<div class="section" id="id30">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.3.1...v0.3.0">0.3.1</a> - 2020-11-05<a class="headerlink" href="#id30" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id31">
<h3>Added<a class="headerlink" href="#id31" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>trackhub: automatic color selection</p></li>
<li><p>trackhub: specify colors with the âcolorsâ column in the samples.tsv. Accepts RGB and matplotlib colors.</p></li>
<li><p>trackhub: grouped samples in a composite track with sample filters and composite control</p></li>
</ul>
</div>
<div class="section" id="id32">
<h3>Changed<a class="headerlink" href="#id32" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>updated genomepy to 0.9.1: genomes will have alternative regions removed (if designated with âaltâ in the name)</p></li>
<li><p>trackhub: better defaults for each track</p></li>
<li><p>layouts are stored per version, as to not have collisions in the way these are stored between versions.</p></li>
<li><p>scATAC no longer supports trackhub</p></li>
<li><p>bigwigs are now (BPM) normalized by default</p></li>
</ul>
</div>
<div class="section" id="id33">
<h3>Fixed<a class="headerlink" href="#id33" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>markduplicates now uses $TMP_DIR, if it is defined</p></li>
<li><p>RNA-seq cluster figures werent displaying text on some platforms</p></li>
<li><p>not using the local annotation files</p></li>
<li><p>not recognizing a mix of gzipped and unzipped annotation files</p></li>
<li><p>bigwigs are now correctly labelled forward/reverse (when protocol was stranded)</p></li>
<li><p>trackhub: RNA-seq trackhub now displays both strands of the bigwig (when protocol was stranded)</p></li>
<li><p>trackhub: track order is now identical to the samples.tsv (was alphabetical for ChIP-/ATAC-seq)</p></li>
<li><p>trackhub: assembly hub index now returns gene_name instead of transcript_id.</p></li>
<li><p>bug with edgeR (upperquartile) normalization failed. Not sure why it fails, but when is does, it now returns a dataframe of nan instead of failing the rule, and thus the whole pipeline.</p></li>
<li><p>use gimmemotifs 0.15.0, so gimme.combine_peaks works with numeric chromosome names</p></li>
<li><p>s2s is slightly more lenient with an edge-case when running seq2science in parallel</p></li>
<li><p>clearer error message when trying samples that can not be found</p></li>
<li><p>edge case with trying to dump sra from empty directory</p></li>
<li><p>now give a nice error message when a technical replicate consists of a mix of paired-end and single-end samples</p></li>
<li><p>issue with large number of inputs for multiqc exceeding the os command max length</p></li>
<li><p>bug with downloading only SRR/DRR samples (but no GSM)</p></li>
<li><p>issue with async generation of genome support files</p></li>
<li><p>checking for sequencing runs when sample is already downloaded</p></li>
</ul>
</div>
</div>
<div class="section" id="id35">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.2.3...v0.3.0">0.3.0</a> - 2020-09-22<a class="headerlink" href="#id35" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id36">
<h3>Added<a class="headerlink" href="#id36" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>fastp as aligner (default), makes trimgalore optional other aligner</p></li>
<li><p>you can now specify an url for your samples file</p></li>
<li><p>RNA-seq: gene_id to gene_name conversion table will be output for downstream analysis</p>
<ul>
<li><p>(may be empty if gtf didnât contain both fields or wrong formatting)</p></li>
</ul>
</li>
<li><p>RNA-seq: quantifying with salmon will now also output a gene length table</p>
<ul>
<li><p>(gene lengths, tpms and gene counts can still be found together in the SingleCellExperiment object)</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="id37">
<h3>Changed<a class="headerlink" href="#id37" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>make use of pysradb for quering layout and SRR ids instead of API and web-scraping</p></li>
<li><p>markduplicates now removes duplicates as default</p></li>
<li><p>testing: clear genomepy caches between runs</p></li>
<li><p>add parallel-fastq-dump fallback to fasterq-dump</p></li>
<li><p>configuration rules split into more sections</p></li>
<li><p>DESeq2 options renamed (from <code class="docutils literal notranslate"><span class="pre">diffexp</span></code> to <code class="docutils literal notranslate"><span class="pre">deseq2</span></code> and <code class="docutils literal notranslate"><span class="pre">contrasts</span></code>)</p></li>
<li><p>DESeq2 will now generate batch corrected counts (and TPMs for Salmon) for all samples, based on the set condition column.</p>
<ul>
<li><p>(batch corrected output is still meant for downstream analysis that cannot model batch effects independently, e.g. plotting)</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="id38">
<h3>Fixed<a class="headerlink" href="#id38" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>issue with control and technical replicates</p></li>
<li><p>now also SRR numbers can be directly downloaded from ENA</p></li>
<li><p>python3.8 syntaxwarnings</p></li>
<li><p>chipseeker missing gtf input</p></li>
<li><p>bugs with explain</p></li>
<li><p>bwa-mem2 not working with less than 12 cores</p></li>
<li><p>batch corrected TPMs no longer break when samples/rows are subset.</p></li>
</ul>
</div>
</div>
<div class="section" id="id40">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.2.2...v0.2.3">0.2.3</a> - 2020-09-01<a class="headerlink" href="#id40" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id41">
<h3>Changed<a class="headerlink" href="#id41" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>retry mechanic for genomepy functions</p></li>
<li><p>moved RNA-seq sample clustering to the MultiQC</p></li>
<li><p>updated genomepy</p></li>
</ul>
</div>
<div class="section" id="id42">
<h3>Fixed<a class="headerlink" href="#id42" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>suffix being overwritten by layouts</p></li>
<li><p>issue with combining conditions and ruleorder for macs2</p></li>
<li><p>Assembly hub correctly showing annotations</p></li>
<li><p>.fa.sizes staying empty</p></li>
</ul>
</div>
</div>
<div class="section" id="id44">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.2.1...v0.2.2">0.2.2</a> - 2020-08-24<a class="headerlink" href="#id44" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id45">
<h3>Added<a class="headerlink" href="#id45" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>option to add custom files to each assembly (such as ERCC spike ins for scRNA-seq)</p></li>
</ul>
</div>
<div class="section" id="id46">
<h3>Changed<a class="headerlink" href="#id46" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>assemblies are now checked in the configuration, similar to samples</p></li>
<li><p>get_genome was split in 3 rules, allowing for less reruns</p></li>
<li><p>Profiles are now parsed by the s2s wrapper</p></li>
<li><p>Checking for validity of samples.tsv now happens with pandasschema</p></li>
<li><p>Explicit priority arguments to all group jobs (aligner + samtools_presort)</p></li>
<li><p>Snakemake version (5.22.1)</p></li>
<li><p>Reduced threads on salmon indexing (matching aligners)</p></li>
<li><p>Make use of fasterq-dump instead of parallel-fastq-dump</p></li>
</ul>
</div>
<div class="section" id="id47">
<h3>Fixed<a class="headerlink" href="#id47" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Test no longer use old cache files</p></li>
<li><p>Profiles no longer overwrite command line arguments</p></li>
<li><p>Fixed edge-case with condition column in samples but no peak-calling</p></li>
<li><p>Downloading sra with prefetch tries multiple times to correct for lost connection</p></li>
<li><p>Ambiguity exception with rule narrowpeak_summit</p></li>
<li><p>combine_peaks makes use of biological replicateâs peaks, not technical replicateâs peaks</p></li>
<li><p>Bug with direct peak-calling on conditions</p></li>
</ul>
</div>
</div>
<div class="section" id="id49">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.2.0...v0.2.1">0.2.1</a> - 2020-08-10<a class="headerlink" href="#id49" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id50">
<h3>Added<a class="headerlink" href="#id50" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Chipseeker images in MultiQC report</p></li>
<li><p>Samples that are on ENA are now directly downloaded from ENA as fastq. This means we skip the CPU instensive dumping step!</p></li>
</ul>
</div>
<div class="section" id="id51">
<h3>Fixed<a class="headerlink" href="#id51" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Fixed issue with some samples not being findable/downloadable with s2s</p></li>
<li><p>Fixed has_annotation always looking for annotation even if local files present</p></li>
<li><p>Fixed bug where scatac-seq workflow was making fastqc reports per sample</p></li>
</ul>
</div>
<div class="section" id="id52">
<h3>Changed<a class="headerlink" href="#id52" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>will try to UCSC gene annotations in Ensembl format (which uses gene IDs for the gene_id field, contrary to the UCSC format that uses transcript IDs. Wild huh?)</p></li>
</ul>
</div>
</div>
<div class="section" id="id54">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.1.0...v0.2.0">0.2.0</a> - 2020-08-04<a class="headerlink" href="#id54" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id55">
<h3>Fixed<a class="headerlink" href="#id55" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Allow for same condition name across different assemblies &amp; different controls</p></li>
</ul>
</div>
<div class="section" id="id56">
<h3>Added<a class="headerlink" href="#id56" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>HISAT2 as aligner for RNA-seq</p></li>
<li><p>splice-aware HISAT2 indexing for RNA-seq</p></li>
<li><p>quantifier HTSeq for RNA-seq</p></li>
<li><p>quantifier featurecounts for RNA-seq</p></li>
<li><p>Salmon will output a gene-level TPM matrix as well</p></li>
<li><p>added/expanded <code class="docutils literal notranslate"><span class="pre">seq2science</span> <span class="pre">explain</span></code> info (now covers RNA- and scATAC-seq too)</p></li>
<li><p>sequencing strandedness may now be inferred automatically (unless specified in the config/samples.tsv)</p></li>
<li><p>strandedness results are displayed in the multiQC under âStrandednessâ</p></li>
<li><p>a DEXSeq counts matrixs can now be generated with <code class="docutils literal notranslate"><span class="pre">dexseq:</span> <span class="pre">True</span></code></p></li>
<li><p>seq2science CLI now has the same reason flag as snakemake (-r/âreason flag)</p></li>
<li><p>(re)added fnwi + rimls logos to the qc reports that went missing in seq2science migration</p></li>
</ul>
</div>
<div class="section" id="id57">
<h3>Changed<a class="headerlink" href="#id57" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>rules and script names in RNA-seq. ex: <code class="docutils literal notranslate"><span class="pre">txi.R</span></code> is now <code class="docutils literal notranslate"><span class="pre">quant_to_counts.R</span></code> to better reflect its function</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">quant_to_counts.R</span></code> now converts salmon transcript abundances to gene counts identically to DESeq2</p></li>
<li><p>STAR no longer outputs counts, and is no longer found under <code class="docutils literal notranslate"><span class="pre">quantifiers</span></code></p></li>
<li><p>gene counts are generated from (filtered) bams when using either STAR or HISAT2 as aligner and HTSeq or featureCounts are quantifier</p></li>
<li><p>batch corrected gene counts are generated if a DESeq2 design contrast inclused a batch</p></li>
<li><p>batch corrected TPM are generated if a DESeq2 design contrast inclused a batch, and quantification was performed using Salmon</p>
<ul>
<li><p>for us in ANANSE, for instance</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq2science</span> <span class="pre">explain</span></code> now retrieves messages from <code class="docutils literal notranslate"><span class="pre">explain.smk</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq2science</span> <span class="pre">explain</span></code> now used profiles and snakemakeOptions.</p></li>
</ul>
</div>
<div class="section" id="id58">
<h3>Fixed<a class="headerlink" href="#id58" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>the alignment workflow no longer uses strandedness</p></li>
<li><p>seq2science CLI can now be run without cores with a dryrun or profile with cores</p></li>
<li><p>Jenkins code style (now used mamba to install flake8)</p></li>
</ul>
</div>
</div>
<div class="section" id="id60">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.0.3...v0.1.0">0.1.0</a> - 2020-07-15<a class="headerlink" href="#id60" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id61">
<h3>Added<a class="headerlink" href="#id61" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>bwa-mem2 as aligner</p></li>
<li><p>new command-line option <code class="docutils literal notranslate"><span class="pre">explain</span></code>, which explains what has been done, and writes your material &amp; methods section for you!</p></li>
</ul>
</div>
<div class="section" id="id62">
<h3>Changed<a class="headerlink" href="#id62" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>change the workflow names, replaced _ by -. (download_fastq to download-fastq, chip_seq to chip-seq, atac_seq to atac-seq, scatac_seq to scatac-seq, and rna_seq to rna-seq)</p></li>
<li><p>changed the way seq2science is called. Moved all the logic from bin/seq2science to seq2science/cli.py</p></li>
</ul>
</div>
<div class="section" id="id63">
<h3>Fixed<a class="headerlink" href="#id63" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>Bug when merging replicates and having controls</p></li>
</ul>
</div>
</div>
<div class="section" id="id65">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.0.2...v0.0.3">0.0.3</a> - 2020-07-01<a class="headerlink" href="#id65" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id66">
<h3>Fixed<a class="headerlink" href="#id66" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>bug when specifying 2 cores, which rounded down to zero cores for samtools sorting and crash</p></li>
<li><p>edger environment was incompatible</p></li>
<li><p>seq2science cache on sensible location + seq2science clean fixed</p></li>
<li><p>only lookup sample layout when not local, opens up for slightly better tests in bioconda recipe</p></li>
</ul>
</div>
</div>
<div class="section" id="id68">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.0.1...v0.0.2">0.0.2</a> - 2020-06-29<a class="headerlink" href="#id68" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="id69">
<h3>Fixed<a class="headerlink" href="#id69" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>samtools using the correct nr of threads after update to v1.10</p></li>
</ul>
</div>
<div class="section" id="id70">
<h3>Changed<a class="headerlink" href="#id70" title="Permalink to this headline">Â¶</a></h3>
<ul class="simple">
<li><p>The count table for ATAC/ChIP-seq peaks is now made from finding all peaks within a range of 200 bp, and taking the most significant one (gimmemotifsâ combine_peaks) and extending the remaining peaks 200 bp. On this count table quantile normalisation, TMM, RLE and upperquartile normalisation with CPM is done. Downstream steps log transform these and mean center them. This however means that for broadpeaks no count_table is generated.</p></li>
<li><p>Snakefmt -l 121 applied</p></li>
</ul>
</div>
</div>
<div class="section" id="id72">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/compare/v0.0.0...v0.0.1">0.0.1</a> - 2020-06-17<a class="headerlink" href="#id72" title="Permalink to this headline">Â¶</a></h2>
<p>Many minor bug- and quality of life fixes.</p>
</div>
<div class="section" id="id74">
<h2><a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/releases/tag/v0.0.0">0.0.0</a> - 2020-06-11<a class="headerlink" href="#id74" title="Permalink to this headline">Â¶</a></h2>
<p>First release of seq2science!</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../CONTRIBUTING.html" class="btn btn-neutral float-right" title="Contributing to seq2science" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="schemas.html" class="btn btn-neutral float-left" title="All configurable options" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright Maarten van der Sande, Siebren FrÃ¶lich, Jos Smits, Rebecca Snabel, Tilman SchÃ¤fers, &amp; Simon van Heeringen..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>