<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Frequently Asked Questions (FAQ) &mdash; seq2science  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Seq2science profiles" href="seq2science_profiles.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Frequently Asked Questions (FAQ)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#one-of-the-rules-failed-and-the-log-turns-red">One of the rules failed and the log turns red!</a></li>
<li class="toctree-l2"><a class="reference internal" href="#the-pipeline-starts-creating-conda-environments-and-crashes-createcondaenvironmentexception">The pipeline starts creating conda environments and crashes (CreateCondaEnvironmentException)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#encountered-problems-while-solving-nothing-provides-needed-by-seq2science">Encountered problems while solving (nothing provides … needed by seq2science)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-if-i-change-the-configuration-or-samples-file-after-running-seq2science">What if I change the configuration or samples file after running seq2science?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#failed-to-call-external-services">Failed to call external services.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#unknown-provider">Unknown provider</a></li>
<li class="toctree-l2"><a class="reference internal" href="#storage-exhausted-no-space-left-on-device">storage exhausted / No Space Left on Device</a></li>
<li class="toctree-l2"><a class="reference internal" href="#i-want-to-run-the-pipeline-until-a-certain-rule-but-not-farther">I want to run the pipeline until a certain rule, but not farther</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rna-seq-this-gene-should-be-expressed">RNA-seq: this gene should be expressed!</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scrna-seq-kallisto-bustools-warn-no-reads-pseudoaligned">scRNA-seq: Kallisto|Bustools: [~warn] no reads pseudoaligned</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scrna-seq-kallisto-bustools-what-std-bad-alloc">scRNA-seq: Kallisto|Bustools: what():  std::bad_alloc</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Frequently Asked Questions (FAQ)</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/faq.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="frequently-asked-questions-faq">
<h1>Frequently Asked Questions (FAQ)<a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline">¶</a></h1>
<section id="one-of-the-rules-failed-and-the-log-turns-red">
<h2>One of the rules failed and the log turns red!<a class="headerlink" href="#one-of-the-rules-failed-and-the-log-turns-red" title="Permalink to this headline">¶</a></h2>
<p>Even though we (the developers) think that seq2science works perfectly in all situations, this, unfortunately, is not always the case. Luckily most rules keep a log, where the error message of that rule is being printed to. When a rule fails seq2science prints a couple things, and one of those thing is where the log is stored. Open that file and check the error message. Sometimes rules fail because for instance the storage of your server is full or the internet connection during the downloading of a genome/sample got interrupted. If you think the rule failed because of a mistake on our side, please check our <a class="reference external" href="https://github.com/vanheeringen-lab/seq2science/issues">issues page</a> and see if someone already made an issue about this. If not, we invite you to start a new issue and we’ll try to help you as soon as possible.</p>
<p>p.s. updating to the newest version (if you did not already) also might solve your problem!</p>
</section>
<section id="the-pipeline-starts-creating-conda-environments-and-crashes-createcondaenvironmentexception">
<h2>The pipeline starts creating conda environments and crashes (CreateCondaEnvironmentException)<a class="headerlink" href="#the-pipeline-starts-creating-conda-environments-and-crashes-createcondaenvironmentexception" title="Permalink to this headline">¶</a></h2>
<p>The first thing the pipeline does is making separate conda environments for the rules that will be run. One thing that causes this error is that there isn’t enough space available on the device to install these environments. Sometimes, even when there is enough space on the device, the installation still fails, and we haven’t been able to pinpoint exactly what is causing this. What usually seems to work is just to remove the installed environemnts (<code class="docutils literal notranslate"><span class="pre">seq2science</span> <span class="pre">clean</span></code>) and try again.</p>
<p>A different type of CreateCondaEnvironmentException occurs when you have conda configured to strict, and this will give a UnsatisfiableError. We haven’t been able to solve all problems with this, and recommend to try again with conda set to flexible instead of strict:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">user@comp:~$ </span>conda config --set channel_priority flexible
</pre></div>
</div>
</section>
<section id="encountered-problems-while-solving-nothing-provides-needed-by-seq2science">
<h2>Encountered problems while solving (nothing provides … needed by seq2science)<a class="headerlink" href="#encountered-problems-while-solving-nothing-provides-needed-by-seq2science" title="Permalink to this headline">¶</a></h2>
<p>This error usually has something to do with not having all the channels properly added. Running this should resolve it:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">user@comp:~$ </span>conda config --add channels defaults
<span class="gp">user@comp:~$ </span>conda config --add channels bioconda
<span class="gp">user@comp:~$ </span>conda config --add channels conda-forge
</pre></div>
</div>
</section>
<section id="what-if-i-change-the-configuration-or-samples-file-after-running-seq2science">
<h2>What if I change the configuration or samples file after running seq2science?<a class="headerlink" href="#what-if-i-change-the-configuration-or-samples-file-after-running-seq2science" title="Permalink to this headline">¶</a></h2>
<p>Seq2science starts each run with checking if it was already run before, and if so, if any settings were changed. Seq2science then automatically derives which files need to be changed and only starts the necessary jobs.</p>
<p>This comes in handy when for example you want to add more data later on in the analysis or change for instance the colors in the trackhub.</p>
</section>
<section id="failed-to-call-external-services">
<h2>Failed to call external services.<a class="headerlink" href="#failed-to-call-external-services" title="Permalink to this headline">¶</a></h2>
<p>When downloading samples / looking up their layout online, seq2science makes use of online resources. Sometimes it can happen that those services are not online at the moment of running seq2science, you do not have internet, or you lose connection with the service. Usually just re-running seq2science solves these issues, either directly or a couple hours later.</p>
</section>
<section id="unknown-provider">
<h2>Unknown provider<a class="headerlink" href="#unknown-provider" title="Permalink to this headline">¶</a></h2>
<p>When downloading a genome assembly, seq2science uses genomepy to look up online which assemblies exist, and whether or not the one you try to download can actually be downloaded. In some rare cases this fails, which gives the error <code class="docutils literal notranslate"><span class="pre">Unknown</span> <span class="pre">provider</span></code>. This is ussually resolved by rerunning seq2science, either directly or a couple hours later.</p>
</section>
<section id="storage-exhausted-no-space-left-on-device">
<h2>storage exhausted / No Space Left on Device<a class="headerlink" href="#storage-exhausted-no-space-left-on-device" title="Permalink to this headline">¶</a></h2>
<p>Some rules in seq2science make a lot of temporary intermediate files. Generally your operating system stores these temporary files on the location of variable <code class="docutils literal notranslate"><span class="pre">$TMPDIR</span></code> which usually is on <code class="docutils literal notranslate"><span class="pre">/tmp</span></code>. Some servers/computers have a small <code class="docutils literal notranslate"><span class="pre">$TMPDIR</span></code> and seq2science does not handle those cases well! You can manually set the tmpdir to another location with commands:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mkdir -p /scratch/${USER}/tmp
export TMPDIR=/scratch/${USER}/tmp
</pre></div>
</div>
<p>These commands make a folder tmp on the scratch disc (assuming this disc exists), and set the <code class="docutils literal notranslate"><span class="pre">$TMPDIR</span></code> to there.</p>
<p>Another, more simple reason why these errors can happen is that there simply is no more space/storage left on the server. This can simply be resolved by deleting unused/old files you no longer need.</p>
</section>
<section id="i-want-to-run-the-pipeline-until-a-certain-rule-but-not-farther">
<h2>I want to run the pipeline until a certain rule, but not farther<a class="headerlink" href="#i-want-to-run-the-pipeline-until-a-certain-rule-but-not-farther" title="Permalink to this headline">¶</a></h2>
<p>let’s say you want to download and trim reads with seq2science, but you do not want to align and do all the other stuff since you have your own methods for that. You can simply tell seq2science to run until some rules, and not farther:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>seq2science run alignment --cores 48 --snakemakeOptions until=[&quot;trim_galore_PE&quot;,&quot;trim_galore_SE&quot;]
</pre></div>
</div>
</section>
<section id="rna-seq-this-gene-should-be-expressed">
<h2>RNA-seq: this gene should be expressed!<a class="headerlink" href="#rna-seq-this-gene-should-be-expressed" title="Permalink to this headline">¶</a></h2>
<p>Seq2science supports 3 gene-quantifiers: HTSeq-count, FeatureCounts and Salmon.
All three are valid methods, and a large population of the expressed genes is found by all quantifiers.
There are differences between these methods however, and caution is advised with the genes not found by another method.</p>
<p>Additionally, Seq2science employs strict quality checks by default.
You could try reducing the <code class="docutils literal notranslate"><span class="pre">min_mapping_quality</span></code> (for HTSeq and FeatureCounts) or change the parameters to the quantifier.</p>
<p>The biggest differences in expressed genes can be found by switching gene annotation, followed by switching quantifier (to or from Salmon).</p>
</section>
<section id="scrna-seq-kallisto-bustools-warn-no-reads-pseudoaligned">
<h2>scRNA-seq: Kallisto|Bustools: [~warn] no reads pseudoaligned<a class="headerlink" href="#scrna-seq-kallisto-bustools-warn-no-reads-pseudoaligned" title="Permalink to this headline">¶</a></h2>
<p>This error means that no reads were (pseudo)aligned to the reference genome!
Oh no!
This is probably because you did not specify the chemistry correctly.
Take a look at how to <a class="reference external" href="https://vanheeringen-lab.github.io/seq2science/content/workflows/scrna_seq.html#bus-barcode-umi-set-format">specify the BUS format</a> for kallisto|bustools.</p>
</section>
<section id="scrna-seq-kallisto-bustools-what-std-bad-alloc">
<h2>scRNA-seq: Kallisto|Bustools: what():  std::bad_alloc<a class="headerlink" href="#scrna-seq-kallisto-bustools-what-std-bad-alloc" title="Permalink to this headline">¶</a></h2>
<p>bad alloc generally means that kallisto|bustools tried to reserve some memory, but there was none left to be reserved…
Are there other programs running that take up a lot of memory?
For us (the developers) it also seems as if this rule sometimes happens, seemingly at random.
Simply restarting seq2science might just solve the problem!</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="seq2science_profiles.html" class="btn btn-neutral float-left" title="Seq2science profiles" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>