

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Using the results &mdash; seq2science  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Extensive docs" href="extensive_docs.html" />
    <link rel="prev" title="scATAC-seq" href="workflows/scatac_seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using the results</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#multiqc-quality-report">MultiQC quality report</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rename-hide-buttons">Rename &amp; hide buttons</a></li>
<li class="toctree-l3"><a class="reference internal" href="#general-statistics">General statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#metrics-tools">Metrics &amp; Tools</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trackhub">Trackhub</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">UCSC genome browser</a></li>
<li class="toctree-l3"><a class="reference internal" href="#integrative-genomics-viewer">Integrative Genomics Viewer</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bigwigs">BigWigs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene-annotations">Gene annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#supporting-tracks">Supporting tracks</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Using the results</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/results.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="using-the-results">
<h1>Using the results<a class="headerlink" href="#using-the-results" title="Permalink to this headline">¶</a></h1>
<div class="section" id="multiqc-quality-report">
<h2>MultiQC quality report<a class="headerlink" href="#multiqc-quality-report" title="Permalink to this headline">¶</a></h2>
<p>All pipelines (except from the <code class="docutils literal notranslate"><span class="pre">download_fastq</span></code> pipeline) output a <a class="reference external" href="https://multiqc.info/">multiQC</a> report. The report is located under <code class="docutils literal notranslate"><span class="pre">{qc_dir}/multiqc_{assembly}.html</span></code> and we highly recommend always checking out the report after a pipeline run. What is reported inside the report differs per pipeline and input.</p>
<div class="section" id="rename-hide-buttons">
<h3>Rename &amp; hide buttons<a class="headerlink" href="#rename-hide-buttons" title="Permalink to this headline">¶</a></h3>
<p>Seq2science reports quality control metrics for all your samples. However sample names on the SRA or from your own files aren’t always as easy to read as you’d like. The report automatically allows for renaming of samples by the big blue buttons at the start of the report. We recommend adding a <code class="docutils literal notranslate"><span class="pre">descriptive_name</span></code> column to the samples.tsv file, where you put a human-readable sample name for each sample. This column is then automatically used by other tools as well (e.g. trackhub).</p>
<p>When dealing with technical replicates and paired-end data you might find that multiqc reports way too many samples and info to properly check the report. The report also automatically generates “hide” buttons, which gives you more control over which samples are shown/hidden.</p>
</div>
<div class="section" id="general-statistics">
<h3>General statistics<a class="headerlink" href="#general-statistics" title="Permalink to this headline">¶</a></h3>
<p>The general statistics table shows a quick summary of your data. The table is interactive and you can sort it, add or remove columns (there are many hidden columns!) by clicking on the <code class="docutils literal notranslate"><span class="pre">configure</span> <span class="pre">columns</span></code> button.</p>
</div>
<div class="section" id="metrics-tools">
<h3>Metrics &amp; Tools<a class="headerlink" href="#metrics-tools" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>FastQC</strong> (raw):</p>
<ul>
<li><p>Here the results of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> are displayed <strong>before</strong> trimming.</p></li>
</ul>
</li>
<li><p><strong>Cutadapt</strong>:</p>
<ul>
<li><p>The pipeline makes use of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">trimgalore!</a> for automatic adapter detection and adapter &amp; quality trimming, which under the hood makes use of <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a> for adapter trimming. This gives us a metric of the percentage of reads that have been trimmed.</p></li>
</ul>
</li>
<li><p><strong>FastQC</strong> (trimmed):</p>
<ul>
<li><p>Here the results of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> are displayed <strong>after</strong> trimming.</p></li>
</ul>
</li>
<li><p><strong>Picard</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://broadinstitute.github.io/picard/">Picard</a> is a suite of tools that can do many useful things. We use it to mark (optical and PCR) duplicates, and to get the sizes of (paired-end) inserts (useful to check for e.g. over-digestion in ATAC-seq).</p></li>
</ul>
</li>
<li><p><strong>Samtools Stats</strong>:</p>
<ul>
<li><p><a class="reference external" href="http://www.htslib.org/doc/samtools-1.6.html">Samtools</a> Stats is part of the SamTools suite. This gives us different metrics about our aligned reads.</p></li>
</ul>
</li>
<li><p><strong>deeptools</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deepTools</a> is a suite of tools to process and analyze deep sequencing data.</p></li>
</ul>
</li>
<li><p><strong>Strandedness</strong></p>
<ul>
<li><p>RNA-seq sample strandedness is inferred by <a class="reference external" href="http://rseqc.sourceforge.net/">RSeQC</a>’s infer_experiment.py to improve gene counting accuracy. Results can be reviewed in the MultiQC graph (and can be used to update the samples.tsv if you disagree with the inference results).</p></li>
</ul>
</li>
<li><p><strong>macs2 / genrich _frips</strong>:</p>
<ul>
<li><p>When calling peaks on your data the fraction reads in peaks score (frips) can be insightful about how well your experiment was performed. This is calculated with <a class="reference external" href="http://subread.sourceforge.net/">featurecounts</a> of the subread module.</p></li>
</ul>
</li>
<li><p><strong>mtnucratio</strong>:</p>
<ul>
<li><p>In some experiments there can be a lot of contamination from mitochondrial reads and it is good to be aware of this. The <code class="docutils literal notranslate"><span class="pre">M</span> <span class="pre">MT/Genome</span> <span class="pre">reads</span></code> values give the number of reads (in millions) that are mapped to the mitochondria or genome, calculated by <a class="reference external" href="https://github.com/apeltzer/MTNucRatioCalculator">MTNucRatioCalculator</a>. You can remove mitochondrial reads with the <code class="docutils literal notranslate"><span class="pre">remove_mito</span></code> flag in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p></li>
</ul>
</li>
</ul>
</div>
</div>
<hr class="docutils" />
<div class="section" id="trackhub">
<h2>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h2>
<p>It is often good to ‘eyeball’ the data, and check if e.g. peak calling went alright. One of the features of the pipeline is that it can generate a trackhub. You can host the trackhub yourself on a web accessible location, and visualize on the <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgHubConnect">UCSC genome browser</a>. Alternatively, you can visualize the files locally in <a class="reference external" href="https://software.broadinstitute.org/software/igv/">IGV</a>.</p>
<p>Generation of the trackhub files is optional for all workflows that support it, and is turned off by default. Set <code class="docutils literal notranslate"><span class="pre">create_trackhub:</span> <span class="pre">True</span></code> in the config to start generating your hub.</p>
<div class="section" id="id2">
<h3>UCSC genome browser<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>If you move the <em>trackhub</em> folder to a web-accessible location, you can upload the URL to the <code class="docutils literal notranslate"><span class="pre">hub.txt</span></code> file on the <a class="reference external" href="https://genome-euro.ucsc.edu/cgi-bin/hgHubConnect#unlistedHubs">UCSC genome browser</a> to gain access to your personalized hub!</p>
<p>If you don’t have access to a web-accessible location, the bigwig files can be manually uploaded on a UCSC trackhub, as long as the assembly used is recognized by UCSC.</p>
</div>
<div class="section" id="integrative-genomics-viewer">
<h3>Integrative Genomics Viewer<a class="headerlink" href="#integrative-genomics-viewer" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://software.broadinstitute.org/software/igv/">IGV</a> is a locally run genome browser with baseline functionalities for read and sequence inspection. It is an excellent alternative for quick jobs or if you do not have access to a (large enough) web-accessible location.</p>
</div>
<div class="section" id="bigwigs">
<h3>BigWigs<a class="headerlink" href="#bigwigs" title="Permalink to this headline">¶</a></h3>
<p>Bigwigs visualize the sequencing depth per base and form the core of the trackhub. Bigwigs are stored in workflow-dependent locations, and linked in the <em>trackhub</em> folder.</p>
<ul class="simple">
<li><p>Bigwig files generated by the Alignment- and RNA-seq workflow are collected in the <em>bigwigs</em> folder. Each aligned sample (or merged sample in case of technical replicates) is converted to a bigwig file.</p></li>
<li><p>Bigwig files generated by the ATAC- and ChIP-seq workflow are collected in the peak-caller directory (<em>macs, genrich, hmmratac</em>). Each condition is converted to a bigwig file.</p></li>
</ul>
<p>See the <a class="reference external" href="https://vanheeringen-lab.github.io/seq2science/content/workflows/atac_seq.html#technical-replicates">replicate handling page</a> for more information on sample replicates and conditions.</p>
</div>
<div class="section" id="genome">
<h3>Genome<a class="headerlink" href="#genome" title="Permalink to this headline">¶</a></h3>
<p>If your genome assembly is not recognized on UCSC, a number of files must be generated to map your bigwigs to. Seq2science does this for you! It creates the required <em>genome.2bit</em>, as well as a <em>cytobands</em> file. With just these files you can search your assembly by coordinates.</p>
</div>
<div class="section" id="gene-annotations">
<h3>Gene annotations<a class="headerlink" href="#gene-annotations" title="Permalink to this headline">¶</a></h3>
<p>If gene annotations are available, these are added as <em>annotations.bigBed</em>. This a visible as a separate track containing genes.
Additionally, the <em>genome.2bit</em> is indexed to allow you to search your assembly by gene name (if the annotation file was formatted properly).</p>
</div>
<div class="section" id="supporting-tracks">
<h3>Supporting tracks<a class="headerlink" href="#supporting-tracks" title="Permalink to this headline">¶</a></h3>
<p>Track depicting the GC-percentage and the softmasked regions of the genome are generated, similarly to <a class="reference external" href="https://github.com/Gaius-Augustus/MakeHub">MakeHub</a>.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="extensive_docs.html" class="btn btn-neutral float-right" title="Extensive docs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="workflows/scatac_seq.html" class="btn btn-neutral float-left" title="scATAC-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>