

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using the results &mdash; seq2science  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Extensive docs" href="extensive_docs.html" />
    <link rel="prev" title="scRNA-seq" href="workflows/scrna_seq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using the results</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#multiqc-quality-report">MultiQC quality report</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rename-hide-buttons">Rename &amp; hide buttons</a></li>
<li class="toctree-l3"><a class="reference internal" href="#general-statistics">General statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#metrics-tools">Metrics &amp; Tools</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#trackhub">Trackhub</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id5">UCSC genome browser</a></li>
<li class="toctree-l3"><a class="reference internal" href="#integrative-genomics-viewer">Integrative Genomics Viewer</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bigwigs">BigWigs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genome">Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene-annotations">Gene annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#supporting-tracks">Supporting tracks</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Using the results</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/results.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-the-results">
<h1>Using the results<a class="headerlink" href="#using-the-results" title="Link to this heading"></a></h1>
<section id="multiqc-quality-report">
<h2>MultiQC quality report<a class="headerlink" href="#multiqc-quality-report" title="Link to this heading"></a></h2>
<p>All pipelines (except from the <code class="docutils literal notranslate"><span class="pre">download_fastq</span></code> pipeline) output a <a class="reference external" href="https://multiqc.info/">multiQC</a> report. The report is located under <code class="docutils literal notranslate"><span class="pre">{qc_dir}/multiqc_{assembly}.html</span></code> and we highly recommend always checking out the report after a pipeline run. What is reported inside the report differs per pipeline and input. In this section we go over the generic output of a seq2science QC report and how to make use of it.</p>
<section id="rename-hide-buttons">
<h3>Rename &amp; hide buttons<a class="headerlink" href="#rename-hide-buttons" title="Link to this heading"></a></h3>
<p>Seq2science reports quality control metrics for all your samples. However sample names on the SRA or from your own files aren’t always as easy to read as you’d like. The report automatically allows for renaming of samples by the big blue buttons at the start of the report. We recommend adding a <code class="docutils literal notranslate"><span class="pre">descriptive_name</span></code> column to the samples.tsv file, where you put a human-readable sample name for each sample. This column is then automatically used by other tools as well (e.g. trackhub).</p>
<p>When dealing with technical replicates and paired-end data you might find that multiqc reports way too many samples and info to properly check the report. The report also automatically generates “hide” buttons, which gives you more control over which samples are shown/hidden.</p>
</section>
<section id="general-statistics">
<h3>General statistics<a class="headerlink" href="#general-statistics" title="Link to this heading"></a></h3>
<p>The general statistics table shows a quick summary of your data. The table is interactive and you can sort it, add or remove columns (there are many hidden columns!) by clicking on the <code class="docutils literal notranslate"><span class="pre">configure</span> <span class="pre">columns</span></code> button. A very useful function of the general statistics table is that it allows for the plotting of two columns against each other!</p>
</section>
<section id="metrics-tools">
<h3>Metrics &amp; Tools<a class="headerlink" href="#metrics-tools" title="Link to this heading"></a></h3>
<p>The sections in the QC report follows a chronological order. It starts with e.g. read trimming, then report alignment metrics, and later differential analysis. The sections in detail are:</p>
<ul class="simple">
<li><p><strong>Assembly stats</strong></p>
<ul>
<li><p>The assembly stats are generated by seq2science. They function as a quick check whether the assembly you are using looks good (e.g. enough genes in annotation, or contigs (chromosomes) are big).</p></li>
</ul>
</li>
<li><p><strong>FastQC</strong> (raw) (<em>only present when using trim galore as trimmer</em>):</p>
<ul>
<li><p>Here the results of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> are displayed <strong>before</strong> trimming.</p></li>
</ul>
</li>
<li><p><strong>Cutadapt</strong> (<em>only present when using trim galore as trimmer</em>):</p>
<ul>
<li><p>The pipeline makes use of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">trimgalore!</a> for automatic adapter detection and adapter &amp; quality trimming, which under the hood makes use of <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a> for adapter trimming. This gives us a metric of the percentage of reads that have been trimmed.</p></li>
</ul>
</li>
<li><p><strong>FastQC</strong> (trimmed) (<em>only present when using trim galore as trimmer</em>):</p>
<ul>
<li><p>Here the results of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> are displayed <strong>after</strong> trimming.</p></li>
</ul>
</li>
<li><p><strong>Fastp</strong> (<em>only present when using fastp as trimmer</em>):</p>
<ul>
<li><p>Here the results of <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty560">fastp</a> are displayed. Fastp is a trimmer that also reports a selection of read quality metrics.</p></li>
</ul>
</li>
<li><p><strong>Picard</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://broadinstitute.github.io/picard/">Picard</a> is a suite of tools that can do many useful things. We use it to mark (optical and PCR) duplicates, and to get the sizes of (paired-end) inserts (useful to check for e.g. over-digestion in ATAC-seq).</p></li>
</ul>
</li>
<li><p><strong>Samtools pre-sieve</strong>:</p>
<ul>
<li><p><a class="reference external" href="http://www.htslib.org/doc/samtools-1.6.html">Samtools</a> Stats is part of the SamTools suite. This gives us different metrics about our aligned reads. Samtools is ran twice on our aligned reads, once before removing (we call this sieving) e.g. poorly aligned reads and duplicates, and once after.</p></li>
</ul>
</li>
<li><p><strong>Samtools post-sieve</strong>:</p>
<ul>
<li><p><a class="reference external" href="http://www.htslib.org/doc/samtools-1.6.html">Samtools</a> Stats is part of the SamTools suite. This gives us different metrics about our aligned reads. Samtools is ran twice on our aligned reads, once before removing (we call this sieving) e.g. poorly aligned reads and duplicates, and once after.</p></li>
</ul>
</li>
<li><p><strong>deeptools</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deepTools</a> is a suite of tools to process and analyze deep sequencing data. Contains a PCA, fingerprint, and profile plot.</p></li>
</ul>
</li>
<li><p><strong>Strandedness</strong>:</p>
<ul>
<li><p>RNA-seq sample strandedness is inferred by <a class="reference external" href="http://rseqc.sourceforge.net/">RSeQC</a>‘s infer_experiment.py to improve gene counting accuracy. Results can be reviewed in the MultiQC graph (and can be used to update the samples.tsv if you disagree with the inference results).</p></li>
</ul>
</li>
<li><p><strong>macs2 / genrich _frips</strong> (<em>only present with peak calling</em>):</p>
<ul>
<li><p>When calling peaks on your data the fraction reads in peaks score (frips) can be insightful about how well your experiment was performed. This is calculated with <a class="reference external" href="http://subread.sourceforge.net/">featurecounts</a> of the subread module.</p></li>
</ul>
</li>
<li><p><strong>deeptools correlation heatmaps</strong>:</p>
<ul>
<li><p><a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deepTools</a> is a suite of tools to process and analyze deep sequencing data. These plots show the binned correlation along the genome.</p></li>
</ul>
</li>
<li><p><strong>Peak distribution</strong> (<em>only present with peak calling</em>):</p>
<ul>
<li><p>Shows the distribution of reads along each peak. Figure made with <a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deepTools</a>.</p></li>
</ul>
</li>
<li><p><strong>Feature/peak distribution</strong> (<em>only present with peak calling</em>):</p>
<ul>
<li><p>Shows where peaks are called relative to regulatory sequences in the genome. Figures are made with <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv145">chipseeker</a></p></li>
</ul>
</li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="trackhub">
<h2>Trackhub<a class="headerlink" href="#trackhub" title="Link to this heading"></a></h2>
<p>It is often good to ‘eyeball’ the data, and check if e.g. peak calling went alright.
One of the features of the pipeline is that it can generate a trackhub.
You can host the trackhub yourself on a web accessible location, and visualize on the <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgHubConnect">UCSC genome browser</a>.
Alternatively, you can visualize the files locally in <a class="reference external" href="https://software.broadinstitute.org/software/igv/">IGV</a>.</p>
<p>Generation of the trackhub files is optional for all workflows that support it, and is turned off by default.
Set <code class="docutils literal notranslate"><span class="pre">create_trackhub:</span> <span class="pre">True</span></code> in the config to start generating your hub.</p>
<section id="id5">
<h3>UCSC genome browser<a class="headerlink" href="#id5" title="Link to this heading"></a></h3>
<p>If you move the <em>trackhub</em> folder to a web-accessible location, you can upload the URL to the <code class="docutils literal notranslate"><span class="pre">hub.txt</span></code> file <a class="reference external" href="https://genome-euro.ucsc.edu/cgi-bin/hgHubConnect#unlistedHubs">here</a> to gain access to your personalized hub!</p>
<p>If you don’t have access to a web-accessible location, the bigwig files can be manually uploaded on a UCSC trackhub, as long as the assembly used is recognized by UCSC.</p>
</section>
<section id="integrative-genomics-viewer">
<h3>Integrative Genomics Viewer<a class="headerlink" href="#integrative-genomics-viewer" title="Link to this heading"></a></h3>
<p><a class="reference external" href="https://software.broadinstitute.org/software/igv/">IGV</a> is a locally run genome browser with baseline functionalities for read and sequence inspection.
It is an excellent alternative for quick jobs or if you do not have access to a (large enough) web-accessible location.</p>
</section>
<section id="bigwigs">
<h3>BigWigs<a class="headerlink" href="#bigwigs" title="Link to this heading"></a></h3>
<p>Bigwigs visualize the sequencing depth per base and form the core of the trackhub. Bigwigs are stored in workflow-dependent locations, and linked in the <em>trackhub</em> folder.</p>
<ul class="simple">
<li><p>Bigwig files generated by the Alignment- and RNA-seq workflow are collected in the <em>bigwigs</em> folder. Each aligned sample (or merged sample in case of technical replicates) is converted to a bigwig file.</p></li>
<li><p>Bigwig files generated by the ATAC- and ChIP-seq workflow are collected in the peak-caller directory (<em>macs, genrich, hmmratac</em>). Each biological replicate is converted to a bigwig file.</p></li>
</ul>
<p>See the <a class="reference external" href="./workflows/atac_seq.html#technical-replicates">replicate handling page</a> for more information on sample replicates and conditions.</p>
</section>
<section id="genome">
<h3>Genome<a class="headerlink" href="#genome" title="Link to this heading"></a></h3>
<p>If your genome assembly is not recognized on UCSC, a number of files must be generated to map your bigwigs to.
Seq2science does this for you!
It creates the required <em>genome.2bit</em>, as well as a <em>cytobands</em> file.
With just these files you can search your assembly by coordinates.</p>
</section>
<section id="gene-annotations">
<h3>Gene annotations<a class="headerlink" href="#gene-annotations" title="Link to this heading"></a></h3>
<p>If gene annotations are available, these are added as <em>annotations.bigBed</em>.
This a visible as a separate track containing genes.
Additionally, the <em>genome.2bit</em> is indexed to allow you to search your assembly by gene name (if the annotation file was formatted properly).</p>
</section>
<section id="supporting-tracks">
<h3>Supporting tracks<a class="headerlink" href="#supporting-tracks" title="Link to this heading"></a></h3>
<p>Track depicting the GC-percentage and the softmasked regions of the genome are generated, similarly to <a class="reference external" href="https://github.com/Gaius-Augustus/MakeHub">MakeHub</a>.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="workflows/scrna_seq.html" class="btn btn-neutral float-left" title="scRNA-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="extensive_docs.html" class="btn btn-neutral float-right" title="Extensive docs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>