<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>All configurable options &mdash; seq2science  documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Changelog" href="CHANGELOG.html" />
    <link rel="prev" title="Per rule explanation" href="all_rules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="extensive_docs.html">Extensive docs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="all_rules.html">Per rule explanation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">All configurable options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#general">General</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#samples">samples</a></li>
<li class="toctree-l4"><a class="reference internal" href="#rule-dir">rule_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#result-dir">result_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome-dir">genome_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#log-dir">log_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#benchmark-dir">benchmark_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fqext1">fqext1</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fqext2">fqext2</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fqsuffix">fqsuffix</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cpulimit">cpulimit</a></li>
<li class="toctree-l4"><a class="reference internal" href="#email">email</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download">Download</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sra-dir">sra_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-dir">fastq_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ncbi-key">ncbi_key</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ncbi-requests">ncbi_requests</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ascp-path">ascp_path</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ascp-key">ascp_key</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-general">Alignment general</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#trimmed-dir">trimmed_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#qc-dir">qc_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#final-bam-dir">final_bam_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trimmer">trimmer</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates">technical_replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#biological-replicates">biological_replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#provider">provider</a></li>
<li class="toctree-l4"><a class="reference internal" href="#aligner">aligner</a></li>
<li class="toctree-l4"><a class="reference internal" href="#samtools-index">samtools_index</a></li>
<li class="toctree-l4"><a class="reference internal" href="#markduplicates">markduplicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#min-mapping-quality">min_mapping_quality</a></li>
<li class="toctree-l4"><a class="reference internal" href="#tn5-shift">tn5_shift</a></li>
<li class="toctree-l4"><a class="reference internal" href="#remove-blacklist">remove_blacklist</a></li>
<li class="toctree-l4"><a class="reference internal" href="#remove-mito">remove_mito</a></li>
<li class="toctree-l4"><a class="reference internal" href="#only-primary-align">only_primary_align</a></li>
<li class="toctree-l4"><a class="reference internal" href="#remove-dups">remove_dups</a></li>
<li class="toctree-l4"><a class="reference internal" href="#create-qc-report">create_qc_report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bam-sort-mem">bam_sort_mem</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-genome-extension">custom_genome_extension</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-annotation-extension">custom_annotation_extension</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-assembly-suffix">custom_assembly_suffix</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deeptools-qc">deeptools_qc</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deeptools-plotcorrelation">deeptools_plotcorrelation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deeptools-multibamsummary">deeptools_multibamsummary</a></li>
<li class="toctree-l4"><a class="reference internal" href="#min-template-length">min_template_length</a></li>
<li class="toctree-l4"><a class="reference internal" href="#max-template-length">max_template_length</a></li>
<li class="toctree-l4"><a class="reference internal" href="#subsample">subsample</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-alignment">Workflow: Alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#bam-sorter">bam_sorter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cram-no-bam">cram_no_bam</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-chip-atac-seq">Workflow: ChIP &amp; ATAC-seq</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#peak-caller">peak_caller</a></li>
<li class="toctree-l4"><a class="reference internal" href="#macs2-keep-mates">macs2_keep_mates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#peak-windowsize">peak_windowsize</a></li>
<li class="toctree-l4"><a class="reference internal" href="#slop">slop</a></li>
<li class="toctree-l4"><a class="reference internal" href="#logbase">logbase</a></li>
<li class="toctree-l4"><a class="reference internal" href="#heatmap-slop">heatmap_slop</a></li>
<li class="toctree-l4"><a class="reference internal" href="#heatmap-npeaks">heatmap_npeaks</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deeptools-heatmap-options">deeptools_heatmap_options</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-rna-seq">Workflow: RNA-seq</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#quantifier">quantifier</a></li>
<li class="toctree-l4"><a class="reference internal" href="#tpm2counts">tpm2counts</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ignore-strandedness">ignore_strandedness</a></li>
<li class="toctree-l4"><a class="reference internal" href="#dexseq">dexseq</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-single-cell-atac-seq">Workflow: Single-cell ATAC-seq</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#snaptools-opt">snaptools_opt</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bin-opt">bin_opt</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#workflow-single-cell-rna-seq">Workflow: Single-cell RNA-seq</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fastq-clean-dir">fastq_clean_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-pair">fastq-pair</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id1">quantifier</a></li>
<li class="toctree-l4"><a class="reference internal" href="#barcodefile">barcodefile</a></li>
<li class="toctree-l4"><a class="reference internal" href="#run-kb-seurat-pp">run_kb_seurat_pp</a></li>
<li class="toctree-l4"><a class="reference internal" href="#export-seu-objects">export_seu_objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#seurat-object">seurat_object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#seurat">seurat</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#differential-gene-peak-analysis">Differential gene/peak analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#counts-dir">counts_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deseq2-dir">deseq2_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deseq2">deseq2</a></li>
<li class="toctree-l4"><a class="reference internal" href="#contrasts">contrasts</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#trackhub">Trackhub</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-trackhub">create_trackhub</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub-dir">trackhub_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bigwig-dir">bigwig_dir</a></li>
<li class="toctree-l4"><a class="reference internal" href="#deeptools-bamcoverage">deeptools_bamcoverage</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
<li class="toctree-l2"><a class="reference internal" href="../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="extensive_docs.html">Extensive docs</a> &raquo;</li>
      <li>All configurable options</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/schemas.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="all-configurable-options">
<h1>All configurable options<a class="headerlink" href="#all-configurable-options" title="Permalink to this headline">¶</a></h1>
<p>This is an automatically generated summary of all configurable options for seq2science. These options are loosely grouped around workflows/topics, however they are generally also shared across workflows. So it is possible that  tunable configuration settings are not mentioned in their topic. At the start of each seq2science run the complete configuration is printed to stdout. You can use that printed configuration as the complete list of tunable configuration settings.</p>
<p>We believe that all our default settings are reasonable, and manual finetuning is generally not required.</p>
<div class="section" id="general">
<h2>General<a class="headerlink" href="#general" title="Permalink to this headline">¶</a></h2>
<div class="section" id="samples">
<h3>samples<a class="headerlink" href="#samples" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samples:
    description: tab-separated file specifying the samples to run through the workflow
    default: samples.tsv
    type: string
</pre></div>
</div>
</div>
<div class="section" id="rule-dir">
<h3>rule_dir<a class="headerlink" href="#rule-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>rule_dir:
    description: directory where to find the workflow rules
    default: ../../rules
    type: string
</pre></div>
</div>
</div>
<div class="section" id="result-dir">
<h3>result_dir<a class="headerlink" href="#result-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>result_dir:
    description: where to store the pipeline&#39;s results (by default)
    default: ./results
    type: string
</pre></div>
</div>
</div>
<div class="section" id="genome-dir">
<h3>genome_dir<a class="headerlink" href="#genome-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>genome_dir:
    description: where to store the pipeline&#39;s genome assemblies
    default: ./genomes
    type: string
</pre></div>
</div>
</div>
<div class="section" id="log-dir">
<h3>log_dir<a class="headerlink" href="#log-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>log_dir:
    description: directory where to store the logs (defaults inside the result dir)
    default: log
    type: string
</pre></div>
</div>
</div>
<div class="section" id="benchmark-dir">
<h3>benchmark_dir<a class="headerlink" href="#benchmark-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>benchmark_dir:
    description: directory where to store the logs (defaults inside the result dir)
    default: benchmark
    type: string
</pre></div>
</div>
</div>
<div class="section" id="fqext1">
<h3>fqext1<a class="headerlink" href="#fqext1" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fqext1:
    description: filename suffix when handling paired-end data, describing the forward read
    default: R1
    type: string
</pre></div>
</div>
</div>
<div class="section" id="fqext2">
<h3>fqext2<a class="headerlink" href="#fqext2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fqext2:
    description: filename suffix when handling paired-end data, describing the reverse read
    default: R2
    type: string
</pre></div>
</div>
</div>
<div class="section" id="fqsuffix">
<h3>fqsuffix<a class="headerlink" href="#fqsuffix" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fqsuffix:
    description: file descriptor for fastq files (often fastq or fq)
    default: fastq
    type: string
</pre></div>
</div>
</div>
<div class="section" id="cpulimit">
<h3>cpulimit<a class="headerlink" href="#cpulimit" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cpulimit:
    description: whether or not to make use of cpulimit to enfoce max thread usage for some rules
    default: True
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="email">
<h3>email<a class="headerlink" href="#email" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>email:
    description: email to reach you after pipeline finished, required for trackhub and multiqc report
    type: string
</pre></div>
</div>
</div>
</div>
<div class="section" id="download">
<h2>Download<a class="headerlink" href="#download" title="Permalink to this headline">¶</a></h2>
<div class="section" id="sra-dir">
<h3>sra_dir<a class="headerlink" href="#sra-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sra_dir:
    description: directory where to store the workflow sras (defaults inside the result dir)
    default: sra
    type: string
</pre></div>
</div>
</div>
<div class="section" id="fastq-dir">
<h3>fastq_dir<a class="headerlink" href="#fastq-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastq_dir:
    description: directory where to store the workflow fastqs (defaults inside the result dir)
    default: fastq
    type: string
</pre></div>
</div>
</div>
<div class="section" id="ncbi-key">
<h3>ncbi_key<a class="headerlink" href="#ncbi-key" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ncbi_key:
    description: e-utilities key for faster lookup interval
    type: string
</pre></div>
</div>
</div>
<div class="section" id="ncbi-requests">
<h3>ncbi_requests<a class="headerlink" href="#ncbi-requests" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ncbi_requests:
    description: the number of lookups per second the key permits
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="ascp-path">
<h3>ascp_path<a class="headerlink" href="#ascp-path" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ascp_path:
    description: the pipeline supports downloading through ascp if this variable is set.
    type: string
</pre></div>
</div>
</div>
<div class="section" id="ascp-key">
<h3>ascp_key<a class="headerlink" href="#ascp-key" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ascp_key:
    description: the public key of ascp
    type: string
</pre></div>
</div>
</div>
</div>
<div class="section" id="alignment-general">
<h2>Alignment general<a class="headerlink" href="#alignment-general" title="Permalink to this headline">¶</a></h2>
<div class="section" id="trimmed-dir">
<h3>trimmed_dir<a class="headerlink" href="#trimmed-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trimmed_dir:
    description: directory where to store the workflow trimmed fastqs (defaults inside the result dir)
    default: fastq_trimmed
    type: string
</pre></div>
</div>
</div>
<div class="section" id="qc-dir">
<h3>qc_dir<a class="headerlink" href="#qc-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>qc_dir:
    description: directory where to store the workflow quality controls (defaults inside the result dir)
    default: qc
    type: string
</pre></div>
</div>
</div>
<div class="section" id="final-bam-dir">
<h3>final_bam_dir<a class="headerlink" href="#final-bam-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>final_bam_dir:
    description: directory where to store the workflow&#39;s final deduplicated &amp; sieved bams (defaults inside the result dir)
    default: final_bam
    type: string
</pre></div>
</div>
</div>
<div class="section" id="trimmer">
<h3>trimmer<a class="headerlink" href="#trimmer" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trimmer:
    description: which adapter trimmer to use
    properties:
        trimgalore:
            description: trim galore! settings (note that the adapter is autodetected, so you do not have to set it here)
            properties:
                trimoptions:
                    default: --quality 10 --length 20 --consider_already_trimmed 10
            default:
                trimoptions: --quality 10 --length 20 --consider_already_trimmed 10
        fastp:
            description: fastp settings (note that the adapter is autodetected, so you do not have to set it here)
            properties:
                trimoptions:
                    default:
            default:
                trimoptions:
    default:
        fastp:
            trimoptions:
    type: object
</pre></div>
</div>
</div>
<div class="section" id="technical-replicates">
<h3>technical_replicates<a class="headerlink" href="#technical-replicates" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>technical_replicates:
    description: concatenates samples depending on the names given in the &#39;replicates&#39; column of samples.tsv
    default: merge
    enum: [&#39;merge&#39;, &#39;keep&#39;]
</pre></div>
</div>
</div>
<div class="section" id="biological-replicates">
<h3>biological_replicates<a class="headerlink" href="#biological-replicates" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>biological_replicates:
    description: which method to use to combine replicates (fisher (n &gt;=2), idr (n==2), or keep)
    default: keep
    enum: [&#39;fisher&#39;, &#39;idr&#39;, &#39;keep&#39;]
    type: string
</pre></div>
</div>
</div>
<div class="section" id="provider">
<h3>provider<a class="headerlink" href="#provider" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>provider:
    description: Specify provider to download new genomes from. Will try GENCODE &gt; Ensembl &gt; UCSC &gt; NCBI if left blank.
    enum: [&#39;GENCODE&#39;, &#39;Ensembl&#39;, &#39;UCSC&#39;, &#39;NCBI&#39;]
    type: string
</pre></div>
</div>
</div>
<div class="section" id="aligner">
<h3>aligner<a class="headerlink" href="#aligner" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>aligner:
    description: which aligner to use
    properties:
        bowtie2:
            properties:
                index:
                    default:
                align:
                    default:
            default:
                index:
                align:
        bwa-mem:
            properties:
                index:
                    default: -a bwtsw
                align:
                    default: -M
            default:
                index: -a bwtsw
                align: -M
        bwa-mem2:
            properties:
                index:
                    default:
                align:
                    default: -M
            default:
                index:
                align:
        hisat2:
            properties:
                index:
                    default:
                align:
                    default:
            default:
                index:
                align:
        minimap2:
            properties:
                index:
                    default:
                align:
                    default: -a sr --secondary=no
            default:
                index:
                align:
        star:
            properties:
                index:
                    default: --limitGenomeGenerateRAM 37000000000 --genomeSAsparseD 1
                align:
                    default:
            default:
                index: --limitGenomeGenerateRAM 37000000000 --genomeSAsparseD 1
                align:
    default:
        bwa-mem2:
            index:
            align:
</pre></div>
</div>
</div>
<div class="section" id="samtools-index">
<h3>samtools_index<a class="headerlink" href="#samtools-index" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools_index:
    description: samtools index settings
    default:
    type: string
</pre></div>
</div>
</div>
<div class="section" id="markduplicates">
<h3>markduplicates<a class="headerlink" href="#markduplicates" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>markduplicates:
    description: the parameters of picard markduplicates
    default: -Xms4G -Xmx6G MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=999
    type: string
</pre></div>
</div>
</div>
<div class="section" id="min-mapping-quality">
<h3>min_mapping_quality<a class="headerlink" href="#min-mapping-quality" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>min_mapping_quality:
    description: minimum mapq to keep aligned reads
    default: 0
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="tn5-shift">
<h3>tn5_shift<a class="headerlink" href="#tn5-shift" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>tn5_shift:
    description: whether or not to shift reads to adjust for the tn5 bias
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="remove-blacklist">
<h3>remove_blacklist<a class="headerlink" href="#remove-blacklist" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>remove_blacklist:
    description: whether or not to remove reads in the encode blacklisted regions
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="remove-mito">
<h3>remove_mito<a class="headerlink" href="#remove-mito" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>remove_mito:
    description: whether or not to remove reads aligned to the mitochondria
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="only-primary-align">
<h3>only_primary_align<a class="headerlink" href="#only-primary-align" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>only_primary_align:
    description: whether or not to only keep primary aligned reads (no multimappers)
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="remove-dups">
<h3>remove_dups<a class="headerlink" href="#remove-dups" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>remove_dups:
    description: whether or not to remove duplicate reads (either optical or PCR)
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="create-qc-report">
<h3>create_qc_report<a class="headerlink" href="#create-qc-report" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>create_qc_report:
    description: whether or not to generate a multiqc report
    default: True
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="bam-sort-mem">
<h3>bam_sort_mem<a class="headerlink" href="#bam-sort-mem" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bam_sort_mem:
    description: The max memory used in gigabytes by samtools/sambamba when sorting bams
    default: 2
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="custom-genome-extension">
<h3>custom_genome_extension<a class="headerlink" href="#custom-genome-extension" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>custom_genome_extension:
    description: File(s) to append to the genome (must be in fasta format)
</pre></div>
</div>
</div>
<div class="section" id="custom-annotation-extension">
<h3>custom_annotation_extension<a class="headerlink" href="#custom-annotation-extension" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>custom_annotation_extension:
    description: File(s) to append to the gene annotations (must be in GTF format)
</pre></div>
</div>
</div>
<div class="section" id="custom-assembly-suffix">
<h3>custom_assembly_suffix<a class="headerlink" href="#custom-assembly-suffix" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>custom_assembly_suffix:
    description: What to append to the assembly name to indicate that custom data has been added to either genome or assembly.
    default: _custom
    type: string
</pre></div>
</div>
</div>
<div class="section" id="deeptools-qc">
<h3>deeptools_qc<a class="headerlink" href="#deeptools-qc" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deeptools_qc:
    description: Whether or not to perform deeptools QC. With many samples this can take a long time and might be undesirable
    default: True
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="deeptools-plotcorrelation">
<h3>deeptools_plotcorrelation<a class="headerlink" href="#deeptools-plotcorrelation" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deeptools_plotcorrelation:
    description: Deeptools plotCorrelation params
    default: --colorMap RdYlBu_r --plotNumbers
    type: string
</pre></div>
</div>
</div>
<div class="section" id="deeptools-multibamsummary">
<h3>deeptools_multibamsummary<a class="headerlink" href="#deeptools-multibamsummary" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deeptools_multibamsummary:
    description: Deeptools multiBamSummary params
    default: --distanceBetweenBins 9000 --binSize 1000
    type: string
</pre></div>
</div>
</div>
<div class="section" id="min-template-length">
<h3>min_template_length<a class="headerlink" href="#min-template-length" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>min_template_length:
    description: the minimum template length for paired-end reads
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="max-template-length">
<h3>max_template_length<a class="headerlink" href="#max-template-length" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>max_template_length:
    description: the maximum template length for paired-end reads
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="subsample">
<h3>subsample<a class="headerlink" href="#subsample" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>subsample:
    description: the maximum amount of reads allowed in the final bam file. When setting a value lower than zero, no subsampling happens (default).
    type: integer
    default: -1
</pre></div>
</div>
</div>
</div>
<div class="section" id="workflow-alignment">
<h2>Workflow: Alignment<a class="headerlink" href="#workflow-alignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="bam-sorter">
<h3>bam_sorter<a class="headerlink" href="#bam-sorter" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bam_sorter:
    description: which sorting software to use
    type: object
    properties:
        description: sort order of aligned reads
        default: coordinate
        enum: [&#39;queryname&#39;, &#39;coordinate&#39;]
        type: string
    default:
        samtools: coordinate
</pre></div>
</div>
</div>
<div class="section" id="cram-no-bam">
<h3>cram_no_bam<a class="headerlink" href="#cram-no-bam" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cram_no_bam:
    description: whether to store cram instead of bam
    type: boolean
    default: False
</pre></div>
</div>
</div>
</div>
<div class="section" id="workflow-chip-atac-seq">
<h2>Workflow: ChIP &amp; ATAC-seq<a class="headerlink" href="#workflow-chip-atac-seq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="peak-caller">
<h3>peak_caller<a class="headerlink" href="#peak-caller" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>peak_caller:
    type: object
    properties:
        genrich:
            type: string
        hmmratac:
            type: string
        macs2:
            type: string
    minProperties: 1
    additionalProperties: False
    default:
        macs2: --shift -100 --extsize 200 --nomodel --keep-dup 1 --buffer-size 10000
</pre></div>
</div>
</div>
<div class="section" id="macs2-keep-mates">
<h3>macs2_keep_mates<a class="headerlink" href="#macs2-keep-mates" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>macs2_keep_mates:
    description: macs2 normally removes paired-end mates when doing --nomodel shift peak calling. When this flag is on, the paired-end bam is converted to sngle-end bam, so all reads are counted. This is not conventional to do, but might help with low sequencing depth.
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="peak-windowsize">
<h3>peak_windowsize<a class="headerlink" href="#peak-windowsize" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>peak_windowsize:
    description: when merging narrowpeaks, how much each summit should be extended by (on each side) when considering overlapping peaks. If for instance a peak_windowsize of 100 is chosen, all peaks within 100 bps of each other will be merged.
    default: 100
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="slop">
<h3>slop<a class="headerlink" href="#slop" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>slop:
    description: how much each summit is extended by (on both sides) when making the final peak. For e.g. count tables it is important that all peaks are of the same size. If for instance a slop of 100 is chosen, the final peak width (for all peaks) is 200.
    type: integer
    default: 100
</pre></div>
</div>
</div>
<div class="section" id="logbase">
<h3>logbase<a class="headerlink" href="#logbase" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>logbase:
    description: The log base of the number of reads under peak count tables
    default: 2
    type: number
</pre></div>
</div>
</div>
<div class="section" id="heatmap-slop">
<h3>heatmap_slop<a class="headerlink" href="#heatmap-slop" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>heatmap_slop:
    description: how much the summits should be extended by (on each side) for making a heatmap of peaks in the multiQC report. If for instance a slop of 1000 is chosen, the final peak&#39;s width is 2000.
    default: 1000
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="heatmap-npeaks">
<h3>heatmap_npeaks<a class="headerlink" href="#heatmap-npeaks" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>heatmap_npeaks:
    description: The number of random peaks used for the heatmap. This at most visualizes all peaks found, so picking a number higher than the number of peaks has no effect. The reason to pick a smaller number is to limit computational resources.
    default: 20000
    type: integer
</pre></div>
</div>
</div>
<div class="section" id="deeptools-heatmap-options">
<h3>deeptools_heatmap_options<a class="headerlink" href="#deeptools-heatmap-options" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deeptools_heatmap_options:
    description: the options passed to deeptools when visualizing the heatmap
    default: --kmeans 6 --xAxisLabel &quot;Summit distance (bp)&quot;
    type: string
</pre></div>
</div>
</div>
</div>
<div class="section" id="workflow-rna-seq">
<h2>Workflow: RNA-seq<a class="headerlink" href="#workflow-rna-seq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="quantifier">
<h3>quantifier<a class="headerlink" href="#quantifier" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>quantifier:
    description: which quantifier to use
    properties:
        salmon:
            properties:
                quantifier_index:
                    default: -k 31
                quantifier_flags:
                    default: --seqBias --gcBias --validateMappings --recoverOrphans
                decoy_aware_index:
                    default: False
            default:
                quantifier_index: -k 31
                quantifier_flags: --seqBias --gcBias --validateMappings --recoverOrphans
                decoy_aware_index: False
        htseq:
            properties:
                htseq_flags:
                    default:
            default:
                htseq_flags:
        featurecounts:
            properties:
                featurecounts_flags:
                    default:
            default:
                featurecounts_flags:
    default:
        htseq:
            htseq_flags:
</pre></div>
</div>
</div>
<div class="section" id="tpm2counts">
<h3>tpm2counts<a class="headerlink" href="#tpm2counts" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>tpm2counts:
    description: how do you wish to convert TPMs to gene counts?
    properties:
        tximeta:
            description: Tximeta uses Ensembl/GENCODE assemblies to convert TPMs to gene counts, and produces an additional R object with the various contents.
            properties:
                txi_source:
                    description: the source of transcriptome
                    enum: [&#39;Ensembl&#39;, &#39;GENCODE&#39;, &#39;de-novo&#39;]
                    default: de-novo
                txi_organism:
                    description: organism (e.g. &quot;Homo sapiens&quot;)
                    default: tximeta_file
                txi_release:
                    description: release number of the source database (e.g. &quot;101&quot;)
                    default: 42
            default:
                txi_source: de-novo
                txi_organism: tximeta_file
                txi_release: 42
        pytxi:
            description: Pytxi uses genomepy assemblies to convert TPMs to gene counts (using either the GTF file or MyGene.info).
            properties:
                tx2gene_from_gtf:
                    description: try to convert to gene names (symbols) using the GTF file. If false (or the GTF cannot help), pytxi uses MyGene.info.
                    type: boolean
                    default: True
            default:
                tx2gene_from_gtf: True
    default:
        pytxi:
            tx2gene_from_gtf: True
</pre></div>
</div>
</div>
<div class="section" id="ignore-strandedness">
<h3>ignore_strandedness<a class="headerlink" href="#ignore-strandedness" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ignore_strandedness:
    description: ignore/dont infer strand-specificity?
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="dexseq">
<h3>dexseq<a class="headerlink" href="#dexseq" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>dexseq:
    description: output an exon counts matrix for use in DEXSeq.
    default: False
    type: boolean
</pre></div>
</div>
</div>
</div>
<div class="section" id="workflow-single-cell-atac-seq">
<h2>Workflow: Single-cell ATAC-seq<a class="headerlink" href="#workflow-single-cell-atac-seq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="snaptools-opt">
<h3>snaptools_opt<a class="headerlink" href="#snaptools-opt" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>snaptools_opt:
    description: snaptools optional arguments for snaptools snap-pre, from snaptools (https://github.com/r3fang/SnapTools). For more info run snaptools snap-pre -h
    default: --min-flen=0 --max-flen=1000 --keep-single=FALSE --keep-secondary=FALSE --overwrite=True --min-cov=100 --verbose=True
</pre></div>
</div>
</div>
<div class="section" id="bin-opt">
<h3>bin_opt<a class="headerlink" href="#bin-opt" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bin_opt:
    description: snaptools optional arguments for snaptools snap-add-bmat, from snaptools (https://github.com/r3fang/SnapTools). The genome will be divided into bins of the equal size of --bin-size-list to create the cell x bin count matrix. For more info run snaptools snap-add-bmat -h
    default: --bin-size-list 5000 --verbose=True
</pre></div>
</div>
</div>
</div>
<div class="section" id="workflow-single-cell-rna-seq">
<h2>Workflow: Single-cell RNA-seq<a class="headerlink" href="#workflow-single-cell-rna-seq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="fastq-clean-dir">
<h3>fastq_clean_dir<a class="headerlink" href="#fastq-clean-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastq_clean_dir:
    description: directory where to find the workflow rules
    default: fastq_cleaned
    type: string
</pre></div>
</div>
</div>
<div class="section" id="fastq-pair">
<h3>fastq-pair<a class="headerlink" href="#fastq-pair" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastq-pair:
    description: fastq synchronization
    default:
</pre></div>
</div>
</div>
<div class="section" id="id1">
<h3>quantifier<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>quantifier:
    description: which quantifier to use
    properties:
        citeseqcount:
            properties:
                count:
                    default: -cbf 9 -cbl 16 -umif 1 -umil 8 -cells 372 --max-error 1 --bc_collapsing_dist 1 --umi_collapsing_dist 1
        kallistobus:
            properties:
                ref:
                    default: --workflow lamanno
                count:
                    default: -x 10XV3 --verbose --h5ad --workflow lamanno
    default:
        kallistobus:
            ref: --workflow lamanno
            count: -x 10XV3 --verbose --h5ad --workflow lamanno
</pre></div>
</div>
</div>
<div class="section" id="barcodefile">
<h3>barcodefile<a class="headerlink" href="#barcodefile" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>barcodefile:
    type: string
</pre></div>
</div>
</div>
<div class="section" id="run-kb-seurat-pp">
<h3>run_kb_seurat_pp<a class="headerlink" href="#run-kb-seurat-pp" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>run_kb_seurat_pp:
    default: False
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="export-seu-objects">
<h3>export_seu_objects<a class="headerlink" href="#export-seu-objects" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>export_seu_objects:
    description: Export scRNA count tables to Seurat objects and save in RDS format
    type: boolean
    default: False
</pre></div>
</div>
</div>
<div class="section" id="seurat-object">
<h3>seurat_object<a class="headerlink" href="#seurat-object" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>seurat_object:
    description: Seurat object settings
    properties:
        project_name:
            description: Project name for merged Seurat object
            type: string
        min_cells:
            description: Include features detected in at least this many cells
            type: integer
        min_features:
            description: Include cells where at least this many features are detected
            type: integer
        default:
            project_name: merged
            min_cells: 0
            min_features: 0
</pre></div>
</div>
</div>
<div class="section" id="seurat">
<h3>seurat<a class="headerlink" href="#seurat" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>seurat:
    description: Seurat rmd for exploratory data analysis
    properties:
        method:
            description: experimental method
            enum: [&#39;384plate&#39;, &#39;droplet&#39;]
        barcode_file:
            description: .tab delimited file with 2-col layout for plate based methods (with DNA barcode and well-id)
            type: string
        mt_genes_file:
            description: .txt file containing one column with mitchondrial genes
            type: string
        run.sct:
            description: perform SCTransform normalization, FALSE runs log1p NormalizeData, HVG selection and ScaleData in Seurat
            type: boolean
        run.jackstraw:
            description: perform Jackstraw analysis (does not work with SCTransform)
            type: boolean
        add.spikes.ercc:
            label: use ERCC RNA spike-ins (if present)
            type: boolean
        ercc_pct_max:
            description: max percentage ERCC spike-ins counts
            type: integer
        add.spikes.mt:
            description: use mitochondrial gene-list for qc
            type: boolean
        mt_pct_max:
            description: max percentage mitochondrial genes counts
            type: integer
        gene_tresh:
            description: gene filter threshold for genes considered expressed
            type: integer
        amount_cells_expr:
            description: gene filter threshold for minimal amount of cells a gene should be expressed in
            type: integer
        total_counts_tresh:
            description: cell filter threshold for minimal amount of UMI counts detected in a cell
            type: integer
        total_feat_tresh:
            description: cell filter threshold for minimal amount of features (genes) detected in a cell
            type: integer
        nhvg:
            description: number of Highly Variable Genes, used in Seurat&#39;s FindVariableFeatures
            type: integer
        pcs_for_overview:
            description: perform UMAP on the defined list of pc&#39;s (seperated by comma) and generate combined UMAP plot
            type: string
        pcs_max_hvg:
            description: max number of principal components to visualize
            type: integer
        default:
            method: 384plate
            barcode_file:
            mt_genes_file:
            run.sct: True
            run.jackstraw: False
            add.spikes.ercc: True
            add.spikes.mt: True
            ercc_pct_max: 20
            mt_pct_max: 50
            gene_tresh: 0
            amount_cells_expr: 0
            total_counts_tresh: 1000
            total_feat_tresh: 500
            nhvg: 2000
            pcs_max_hvg: 70
            pcs_for_overview: 10,20,30,40,50
</pre></div>
</div>
</div>
</div>
<div class="section" id="differential-gene-peak-analysis">
<h2>Differential gene/peak analysis<a class="headerlink" href="#differential-gene-peak-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="counts-dir">
<h3>counts_dir<a class="headerlink" href="#counts-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>counts_dir:
    description: directory where to store the counts for DESeq2 analysis (defaults inside the result dir)
    default: counts
    type: string
</pre></div>
</div>
</div>
<div class="section" id="deseq2-dir">
<h3>deseq2_dir<a class="headerlink" href="#deseq2-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deseq2_dir:
    description: directory where to store the DESeq2 output (defaults inside the result dir)
    default: deseq2
    type: string
</pre></div>
</div>
</div>
<div class="section" id="deseq2">
<h3>deseq2<a class="headerlink" href="#deseq2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deseq2:
    description: DESeq2 settings. See the DESeq2 vignette for details
    properties:
        multiple_testing_procedure:
            default: BH
            enum: [&#39;BH&#39;, &#39;IHW&#39;]
            type: string
        alpha_value:
            default: 0.1
            type: number
        shrinkage_estimator:
            default: apeglm
            enum: [&#39;apeglm&#39;, &#39;ashr&#39;, &#39;normal&#39;]
            type: string
    default:
        multiple_testing_procedure: BH
        alpha_value: 0.1
        shrinkage_estimator: apeglm
</pre></div>
</div>
</div>
<div class="section" id="contrasts">
<h3>contrasts<a class="headerlink" href="#contrasts" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>contrasts:
    description: deseq2 design contrast(s)
</pre></div>
</div>
</div>
</div>
<div class="section" id="trackhub">
<h2>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h2>
<div class="section" id="create-trackhub">
<h3>create_trackhub<a class="headerlink" href="#create-trackhub" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>create_trackhub:
    description: generate a trackhub (and required files)?
    default: True
    type: boolean
</pre></div>
</div>
</div>
<div class="section" id="trackhub-dir">
<h3>trackhub_dir<a class="headerlink" href="#trackhub-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trackhub_dir:
    description: directory where to store the trackhub (defaults inside the result dir)
    default: trackhub
    type: string
</pre></div>
</div>
</div>
<div class="section" id="bigwig-dir">
<h3>bigwig_dir<a class="headerlink" href="#bigwig-dir" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bigwig_dir:
    description: directory where to store the bigwigs used by the trackhub (defaults inside the result dir)
    default: bigwigs
    type: string
</pre></div>
</div>
</div>
<div class="section" id="deeptools-bamcoverage">
<h3>deeptools_bamcoverage<a class="headerlink" href="#deeptools-bamcoverage" title="Permalink to this headline">¶</a></h3>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>deeptools_bamcoverage:
    description: flags for bam to bigwig conversions with deeptools bamCoverage
    default: --normalizeUsing BPM --binSize 1
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="all_rules.html" class="btn btn-neutral float-left" title="Per rule explanation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="CHANGELOG.html" class="btn btn-neutral float-right" title="Changelog" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>