<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Alignment &mdash; seq2science  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="ATAC-seq" href="atac_seq.html" />
    <link rel="prev" title="Downloading fastqs" href="download_fastq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#workflow-overview-simplified">Workflow overview (simplified)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-trimming">Read trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignment-sorting">Alignment &amp; Sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-assembly-extensions">Custom assembly extensions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mark-duplicates">Mark duplicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#samtools-index">Samtools index</a></li>
<li class="toctree-l4"><a class="reference internal" href="#quality-report">Quality report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#descriptive-name-column">Descriptive_name column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates-column">technical_replicates column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#colors-column">Colors column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#final-notes">Final notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#bam-mapping-quality-filtering">BAM mapping quality filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cram-support">Cram support</a></li>
<li class="toctree-l4"><a class="reference internal" href="#subsampling">subsampling</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_seq.html">RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scrna_seq.html">scRNA-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="../workflows.html">Workflows</a></li>
      <li class="breadcrumb-item active">Alignment</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/alignment.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="alignment">
<h1>Alignment<a class="headerlink" href="#alignment" title="Permalink to this heading"></a></h1>
<p>Aligning samples has never been easier!</p>
<section id="workflow-overview-simplified">
<h2>Workflow overview (simplified)<a class="headerlink" href="#workflow-overview-simplified" title="Permalink to this heading"></a></h2>
<p align="center">
  <img src="../../_static/alignment.png" style="width:auto;height:500px;">
</p><section id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Permalink to this heading"></a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples.
You control which samples are used in the <a class="reference external" href="#filling-out-the-samples-tsv">samples.tsv</a>.
Background on public data can be found <a class="reference external" href="./download_fastq.html#download-sra-file">here</a>.</p>
</section>
<section id="read-trimming">
<h3>Read trimming<a class="headerlink" href="#read-trimming" title="Permalink to this heading"></a></h3>
<p>The pipeline starts by trimming the reads with Trim Galore! or Fastp (the default).
The trimmer will automatically trim the low quality 3’ ends of reads, and removes short reads.
During the quality trimming it automatically detects which adapter was used, and trims this as well.
Trimming parameters for the pipeline can be set in the configuration.</p>
</section>
<section id="alignment-sorting">
<h3>Alignment &amp; Sorting<a class="headerlink" href="#alignment-sorting" title="Permalink to this heading"></a></h3>
<p>After trimming the reads are aligned against an assembly. Currently we support <em>bowtie2</em>, <em>bwa</em>, <em>bwa-mem2</em>, <em>hisat2</em> and <em>STAR</em> as aligners. Choosing which aligner is as easy as setting the <em>aligner</em> variable in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, for example: <code class="docutils literal notranslate"><span class="pre">aligner:</span> <span class="pre">bwa</span></code>. Sensible defaults have been set for every aligner, but can be overwritten for either (or both) the indexing and alignment by specifying them in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aligner</span><span class="p">:</span>
  <span class="n">bwa</span><span class="o">-</span><span class="n">mem</span><span class="p">:</span>
    <span class="n">index</span><span class="p">:</span> <span class="s1">&#39;-a bwtsw&#39;</span>
    <span class="n">align</span><span class="p">:</span> <span class="s1">&#39;-M&#39;</span>
</pre></div>
</div>
<p>The pipeline will check if the assembly you specified is present in the <em>genome_dir</em>, and otherwise will download it for you through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>.
All these aligners require an index to be formed first for each assembly, but don’t worry, the pipeline does this for you.</p>
<p>The outputted alignment.bam is immediately sorted by either samtools or sambamba (<em>bam_sorter</em>) either in <em>queryname</em> or <em>coordinate</em> (default) order.</p>
</section>
<section id="custom-assembly-extensions">
<h3>Custom assembly extensions<a class="headerlink" href="#custom-assembly-extensions" title="Permalink to this heading"></a></h3>
<p>The genome and/or gene annotation can be extended with custom files, such as ERCC spike-ins for scRNA-seq. To do so, add <code class="docutils literal notranslate"><span class="pre">custom_genome_extension:</span> <span class="pre">path/to/spike_in.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">custom_annotation_extension:</span> <span class="pre">path/to/spike_in.gtf</span></code> to the config.
Seq2science will place the customized assembly in a separate folder in the <code class="docutils literal notranslate"><span class="pre">genome_dir</span></code>.</p>
</section>
<section id="mark-duplicates">
<h3>Mark duplicates<a class="headerlink" href="#mark-duplicates" title="Permalink to this heading"></a></h3>
<p>After aligning &amp; sorting the bam duplicate reads are being ‘marked’ by picard Markduplicates. You can change the call by setting <em>markduplicates</em> in config.yaml.</p>
</section>
<section id="samtools-index">
<h3>Samtools index<a class="headerlink" href="#samtools-index" title="Permalink to this heading"></a></h3>
<p>Many downstream tools require an index of the deduplicated bam. The pipeline automatically generates these for you.</p>
</section>
<section id="quality-report">
<h3>Quality report<a class="headerlink" href="#quality-report" title="Permalink to this heading"></a></h3>
<p>It is always a good idea to check the quality of your samples. Along the way different quality control steps are taken, and are outputted in a single <a class="reference external" href="https://multiqc.info/">multiqc report</a> in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> folder. Make sure to always check the report, and take a look at <a class="reference external" href="../results.html#multiqc-quality-report">interpreting the multiqc report</a>!</p>
</section>
<section id="trackhub">
<h3>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this heading"></a></h3>
<p>A UCSC compatible trackhub can be generated for this workflow.
See the <a class="reference external" href="../results.html#trackhub">trackhub page</a><span class="raw-html-m2r"><!-- @IGNORE PREVIOUS: link --></span> for more information!</p>
</section>
</section>
<section id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this heading"></a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on.
Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs.
As an example, the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">technical_replicates</span>    <span class="n">descriptive_name</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">heart_1</span>      <span class="n">heart_merged</span>
<span class="n">GSM321</span>    <span class="n">GRCh38</span>      <span class="n">heart_1</span>      <span class="n">heart_merged</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">heart_2</span>      <span class="n">heart_not_merged</span>
<span class="n">GSMxzy</span>    <span class="n">danRer11</span>    <span class="n">stage_8</span>      <span class="n">stage_8</span>
<span class="n">GSM890</span>    <span class="n">danRer11</span>    <span class="n">stage_9</span>      <span class="n">stage_9</span>
</pre></div>
</div>
<section id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Permalink to this heading"></a></h3>
<p>If you use the pipeline on <strong>public data</strong> this should be the name of the accession (e.g. GSM2837484).
Accepted formats start with “GSM”, “SRR”, “SRX”, “DRR”, “DRX”, “ERR” or “ERX”.</p>
<p>If you use the pipeline on <strong>local data</strong> this should be the <em>basename</em> of the file without the <em>extension(s)</em>.
For example:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample1.fastq.gz</span></code> ——-&gt; <code class="docutils literal notranslate"><span class="pre">sample1</span></code> for single-ended data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> ┬&gt; <code class="docutils literal notranslate"><span class="pre">sample2</span></code> for paired-ended data <span class="raw-html-m2r"><br></span> <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> ┘</p></li>
</ul>
<p>For <strong>local data</strong>, some fastq files may have slightly different naming formats.
For instance, Illumina may produce a sample named <code class="docutils literal notranslate"><span class="pre">sample3_S1_L001_R1_001.fastq.gz</span></code> (and the <code class="docutils literal notranslate"><span class="pre">R2</span></code> fastq).
Seq2science will attempt to recognize these files based on the sample name <code class="docutils literal notranslate"><span class="pre">sample3</span></code>.</p>
<p>For <strong>both local and public data</strong>, identifiers used to recognize fastq files are the fastq read extensions (<code class="docutils literal notranslate"><span class="pre">R1</span></code> and <code class="docutils literal notranslate"><span class="pre">R2</span></code> by default) and the fastq suffix (<code class="docutils literal notranslate"><span class="pre">fastq</span></code> by default).
The directory where seq2science will store (or look for) fastqs is determined by the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> config option.
In the example above, the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> should be set to <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs</span></code>.
These setting can be changed in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
</section>
<section id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this heading"></a></h3>
<p>Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</section>
<section id="descriptive-name-column">
<h3>Descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Permalink to this heading"></a></h3>
<p>The descriptive_name column is used for the trackhub and multiqc report.
In the trackhub your tracks will be called after the descriptive name, and in the multiqc report there will be a button to rename your samples after this column.
The descriptive name can not contain ‘-’ characters, but underscores ‘_’ are allowed.</p>
</section>
<section id="technical-replicates-column">
<h3>technical_replicates column<a class="headerlink" href="#technical-replicates-column" title="Permalink to this heading"></a></h3>
<p>Technical replicates, or any fastq file you may wish to merge, are set using the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column in the samples.tsv file.
All samples with the same name in the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column will be concatenated into one file with the replicate name.</p>
<p>Example <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> utilizing replicate merging:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">technical_replicates</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">heart</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">heart</span>
<span class="n">GSMxzy</span>    <span class="n">GRCh38</span>      <span class="n">stage8</span>
<span class="n">GSM890</span>    <span class="n">GRCh38</span>
</pre></div>
</div>
<p>Using this file in the alignment workflow will output <em>heart.bam</em>, <em>stage8.bam</em> and <em>GSM890.bam</em>.
The MultiQC will inform you of the trimming steps performed on all samples, and subsequent information of the ‘replicate’ files (of which only <em>heart</em> is merged).</p>
<p>Note: If you are working with multiple assemblies in one workflow, replicate names have to be unique between assemblies (you will receive a warning if names overlap).</p>
<section id="keep">
<h4>keep<a class="headerlink" href="#keep" title="Permalink to this heading"></a></h4>
<p>Replicate merging is turned on by default. It can be turned off by setting <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to <code class="docutils literal notranslate"><span class="pre">keep</span></code>.</p>
</section>
</section>
<section id="colors-column">
<h3>Colors column<a class="headerlink" href="#colors-column" title="Permalink to this heading"></a></h3>
<p>If you are visualizing your data on the UCSC trackhub you can optionally specify the colors of each track.
To do so, you can add the color by name (google “matplotlib colors” for the options), or RGB values, in the “colors” column. Empty fields are considered black.</p>
</section>
<section id="final-notes">
<h3>Final notes<a class="headerlink" href="#final-notes" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Make sure that the samples.tsv is a tab separated values file when running the pipeline.</p></li>
<li><p>Feel free to delete or add columns to your liking.</p></li>
</ul>
</section>
</section>
<section id="filling-out-the-config-yaml">
<h2>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this heading"></a></h2>
<p>Every workflow has many configurable options, and can be set in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. In each <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) <strong>reasonable default</strong> values.</p>
<p>When a workflow starts it prints the complete configuration, and (almost) all these values can be added in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> and changed to your liking. You can see the complete set of configurable options in the <a class="reference external" href="../schemas.html">extensive docs</a>.</p>
</section>
<section id="best-practices">
<h2>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this heading"></a></h2>
<section id="bam-mapping-quality-filtering">
<h3>BAM mapping quality filtering<a class="headerlink" href="#bam-mapping-quality-filtering" title="Permalink to this heading"></a></h3>
<p>All aligners pass a mapping quality (MAPQ) score to the reads.
This score reflects the certainty that a read belongs to a certain genomic position.
This score can then be used by <a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deeptools</a> to only keep reads of which the aligner is certain where it aligns, and to limit multimappers.
You can set the minimum mapq score by using the configuration variable <code class="docutils literal notranslate"><span class="pre">min_mapping_quality</span></code>.
Unfortunately, most aligners use a different scoring system.
This problem has been nicely summarized <a class="reference external" href="https://sequencing.qcfail.com/articles/mapq-values-are-really-useful-but-their-implementation-is-a-mess/">in this article</a>.</p>
<p>For the default aligners (BWA and STAR), a minimal quality score has been set to select uniquely mapped reads only. For other aligners, we suggest the following scores to filter for uniquely mapped reads:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Aligner</p></th>
<th class="head"><p>min. MAPQ scores for uniquely mapped reads</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Bowtie2</p></td>
<td><p>41</p></td>
</tr>
<tr class="row-odd"><td><p>Bwa-mem(2)</p></td>
<td><p>30</p></td>
</tr>
<tr class="row-even"><td><p>Hisat2</p></td>
<td><p>44</p></td>
</tr>
<tr class="row-odd"><td><p>Minimap</p></td>
<td><p>0, but use <code class="docutils literal notranslate"><span class="pre">--secondary=no</span></code> as alignment option</p></td>
</tr>
<tr class="row-even"><td><p>STAR</p></td>
<td><p>255</p></td>
</tr>
</tbody>
</table>
</section>
<section id="cram-support">
<h3>Cram support<a class="headerlink" href="#cram-support" title="Permalink to this heading"></a></h3>
<p>For the edge-case where you want to work with cram files instead of bams, there is an option <code class="docutils literal notranslate"><span class="pre">cram_no_bam</span></code> which will convert your bam files to cram (saves around 60% storage).</p>
</section>
<section id="subsampling">
<h3>subsampling<a class="headerlink" href="#subsampling" title="Permalink to this heading"></a></h3>
<p>In some cases you might want to have the same number of reads between samples. By adding e.g. <code class="docutils literal notranslate"><span class="pre">subsample:1_000_000</span></code> seq2science will make sure that each sample contains at most a million reads. Seq2science always downsamples, never upsamples.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="download_fastq.html" class="btn btn-neutral float-left" title="Downloading fastqs" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="atac_seq.html" class="btn btn-neutral float-right" title="ATAC-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>