<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>ATAC-seq &mdash; seq2science  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../_static/doctools.js?v=888ff710"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="ChIP-seq" href="chip_seq.html" />
    <link rel="prev" title="Alignment" href="alignment.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">ATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#workflow-overview-simplified">Workflow overview (simplified)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#downloading-and-indexing-of-assembly-s">Downloading and indexing of assembly(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-trimming">Read trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignment">Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bam-sieving">Bam sieving</a></li>
<li class="toctree-l4"><a class="reference internal" href="#peak-calling">Peak calling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#quality-report">Quality report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#count-table">Count table</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-motif-analysis">Differential motif analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-peak-analysis">Differential peak analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#control-input-column">Control (input) column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#descriptive-name-column">Descriptive_name column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates-column">Technical_replicates column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#biological-replicates">Biological replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#a-mix-of-biological-and-technical-replicates">A mix of biological and technical replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#colors-column">Colors column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#final-notes">Final notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#macs2-and-shift-with-paired-end-data">MACS2 and –shift with paired-end data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#irreproducible-discovery-rate">Irreproducible Discovery Rate</a></li>
<li class="toctree-l4"><a class="reference internal" href="#subsampling">subsampling</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_seq.html">RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scrna_seq.html">scRNA-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../CONTRIBUTING.html">Contributing to seq2science</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../workflows.html">Workflows</a></li>
      <li class="breadcrumb-item active">ATAC-seq</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/atac_seq.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="atac-seq">
<h1>ATAC-seq<a class="headerlink" href="#atac-seq" title="Link to this heading"></a></h1>
<p>Preprocessing of ATAC-seq has never been easier!</p>
<section id="workflow-overview-simplified">
<h2>Workflow overview (simplified)<a class="headerlink" href="#workflow-overview-simplified" title="Link to this heading"></a></h2>
<p align="center">
  <img src="../../_static/atac_seq.png" style="width:auto;height:500px;">
</p><p>An example quality control report can be <a class="reference external" href="https://vanheeringen-lab.github.io/seq2science/_static/atac_seq_multiqc.html">viewed here</a>.</p>
<section id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Link to this heading"></a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples.
The downloading of samples is integrated into each workflow, so you don’t have to start a download workflow first.
You control which samples are used in the <a class="reference external" href="#filling-out-the-samples-tsv">samples.tsv</a>.
Background on public data can be found <a class="reference external" href="./download_fastq.html#download-sra-file">here</a>.</p>
</section>
<section id="downloading-and-indexing-of-assembly-s">
<h3>Downloading and indexing of assembly(s)<a class="headerlink" href="#downloading-and-indexing-of-assembly-s" title="Link to this heading"></a></h3>
<p>Depending on whether the assembly and its index you align your samples against already exist seq2science will start with downloading of the assembly through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>.</p>
</section>
<section id="read-trimming">
<h3>Read trimming<a class="headerlink" href="#read-trimming" title="Link to this heading"></a></h3>
<p>The pipeline starts by trimming the reads with Trim Galore! or Fastp (the default).
The trimmer will automatically trim the low quality 3’ ends of reads, and removes short reads.
During quality trimming it automatically detects which sequencing adapter was used (if it wasn’t trimmed off yet), and then trims this as well.
Trimming parameters for the pipeline can be set in the configuration, for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">trimmer</span><span class="p">:</span>
  <span class="n">fastp</span><span class="p">:</span>
    <span class="n">trimoptions</span><span class="p">:</span> <span class="o">--</span><span class="n">trim_front1</span> <span class="mi">3</span> <span class="o">--</span><span class="n">trim_front2</span> <span class="mi">14</span> <span class="o">--</span><span class="n">trim_poly_x</span>
</pre></div>
</div>
</section>
<section id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Link to this heading"></a></h3>
<p>After trimming the reads are aligned against an assembly. Currently we support <em>bowtie2</em>, <em>bwa</em>, <em>bwa-mem2</em>, <em>hisat2</em>, <em>minimap2</em> and <em>STAR</em> as aligners. Choosing which aligner is as easy as setting the <em>aligner</em> variable in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, for example: <code class="docutils literal notranslate"><span class="pre">aligner:</span> <span class="pre">bwa</span></code>. Sensible defaults have been set for every aligner, but can be overwritten for either (or both) the indexing and alignment by specifying them in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aligner</span><span class="p">:</span>
  <span class="n">bwa</span><span class="o">-</span><span class="n">mem</span><span class="p">:</span>
    <span class="n">index</span><span class="p">:</span> <span class="s1">&#39;-a bwtsw&#39;</span>
    <span class="n">align</span><span class="p">:</span> <span class="s1">&#39;-M&#39;</span>
</pre></div>
</div>
<p>The pipeline will check if the assembly you specified is present in the <em>genome_dir</em>, and otherwise will download it for you through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>. All these aligners require an index to be formed first for each assembly, but don’t worry, the pipeline does this for you.</p>
</section>
<section id="bam-sieving">
<h3>Bam sieving<a class="headerlink" href="#bam-sieving" title="Link to this heading"></a></h3>
<p>After aligning the bam you can choose to remove unmapped reads, low quality mappings, duplicates, and multimappers. It is also advisable to remove mitochondrially mapped reads for ATAC-seq analysis, plus to get slightly more accurate peaks you can set set the workflow to apply a tn5 shift. Finally you are probably interested in reads from the nucleosome-free region, so by selecting paired-end reads with a short insert-size (+/- 150 bp) you filter out reads from the mono-di-tri-etc nucleosomes.</p>
</section>
<section id="peak-calling">
<h3>Peak calling<a class="headerlink" href="#peak-calling" title="Link to this heading"></a></h3>
<p>Currently we support two peak callers for the ATAC-seq workflow, MACS2 and genrich.</p>
<section id="macs2">
<h4>MACS2<a class="headerlink" href="#macs2" title="Link to this heading"></a></h4>
<p><a class="reference external" href="https://github.com/taoliu/MACS">MACS2</a> is the successor of MACS, created by <a class="reference external" href="https://github.com/taoliu">taoliu</a> (also one of the co-authors of hmmratac). MACS2 generates a ‘<em>pileup</em>‘ of all the reads. You can imagine this pileup as laying all reads on top of each other on the genome, and counting how high your pile gets for each basepair. MACS2 then models ‘<em>background</em>‘ values, based on the total length of all your reads and the genome length, to a <a class="reference external" href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson distribution</a> and decides whether your experimental pileup is significant compared to the poisson distribution. MACS2 is one of the, if not the, most used peak caller.</p>
<p>For the calculation of peaks MACS2 requires that an “effective” genome size is being passed as one of its arguments. For the more common assemblies (e.g. mm10 and human) these numbers can be found online. However we found googling these numbers quite the hassle, and for lesser studied species these numbers can’t be found online. Therefore the pipeline automatically estimates the effective genome size, we calculate this as <strong>the number of unique kmers of the average read length</strong>.</p>
</section>
<section id="genrich">
<h4>Genrich<a class="headerlink" href="#genrich" title="Link to this heading"></a></h4>
<p><a class="reference external" href="https://github.com/jsh58/Genrich">Genrich</a> is a spiritual successor of MACS2, created by <a class="reference external" href="https://github.com/jsh58">John M. Gaspar</a>. Just like MACS2 is generates a ‘<em>pileup’</em>. However the author of genrich realized that the distribution of pileup never follows a poisson distribution. Genrich then uses a <a class="reference external" href="https://en.wikipedia.org/wiki/Log-normal_distribution">log-normal distribution</a> to model the background.</p>
</section>
<section id="hmmratac-stub-not-fully-supported">
<h4>HMMRATAC  (stub, not fully supported!)<a class="headerlink" href="#hmmratac-stub-not-fully-supported" title="Link to this heading"></a></h4>
<p><a class="reference external" href="https://github.com/LiuLabUB/HMMRATAC">HMMRATAC</a> is a peak caller specifically made for ATAC-seq data. HMMRATAC does not generate a so-called pileup. It however looks at the length distribution of reads. Since the tn5 transposase prefers to cut at nucleosome free regions in the DNA, we see that our reads are either short (at open DNA), or the length of 1 (or 2, 3, 4, etc.) nucleosomes long. HMMRATAC then classifies regions in the genome based on the length distribution of the reads in open chromatin, flanking regions, non-flanking regions and background.</p>
</section>
</section>
<section id="quality-report">
<h3>Quality report<a class="headerlink" href="#quality-report" title="Link to this heading"></a></h3>
<p>It is always a good idea to check the quality of your samples. Along the way different quality control steps are taken, and are outputted in a single <a class="reference external" href="https://multiqc.info/">multiqc report</a> in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> folder. Make sure to always check the report, and take a look at <a class="reference external" href="../results.html#multiqc-quality-report">interpreting the multiqc report</a>!</p>
</section>
<section id="count-table">
<h3>Count table<a class="headerlink" href="#count-table" title="Link to this heading"></a></h3>
<p>A useful result the pipeline outputs is the count table, located at {counts_dir}/{peak_caller}/.
For each narrowpeak (across all samples) its summit is taken, and all other summits within range <code class="docutils literal notranslate"><span class="pre">peak_windowsize</span></code> (default 100) are taken together, and the summit with the highest q-value is taken as the “true” peak.
The remaining peaks are extended by <code class="docutils literal notranslate"><span class="pre">slop</span></code> on each side and for each sample the number of reads are counted under this peak.
This file is stored as {assembly}_raw.tsv, and looks something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                <span class="n">sample1</span>     <span class="n">sample2</span>
<span class="n">chr1</span><span class="p">:</span><span class="mi">5649</span><span class="o">-</span><span class="mi">5849</span>  <span class="mf">75.00000</span>    <span class="mf">0.00000</span>
<span class="n">chr1</span><span class="p">:</span><span class="mi">7399</span><span class="o">-</span><span class="mi">7599</span>  <span class="mf">47.00000</span>    <span class="mf">1.00000</span>
</pre></div>
</div>
<p>Seq2science currently supports four different normalisation methods: quantile normalisation, TMM, RLE, and upperquartile normalisation, and does count-per-million (counts under peaks) normalisation on each before normalisation.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Quantile_normalization">**Quantile normalization**</a> is a type of normalization that makes the distribution between samples identical. This means that the actual count distribution within a sample changes. This normalisation is especially powerful when comparing results from different labs/experiments/experimental methods.</p></li>
<li><p><strong>TMM</strong> is the weighted trimmed mean of M-values proposed by Robinson and Oshlack (2010).</p></li>
<li><p><strong>RLE</strong> is the scaling factor method proposed by Anders and Huber (2010). DEseq2’s standard normalisation is based on this.</p></li>
<li><p><strong>Upper quartile</strong> is the upper-quartile normalization method of Bullard et al (2010).</p></li>
</ul>
<p>After these normalisations the counts are log normalised, and the base can be set with <code class="docutils literal notranslate"><span class="pre">logbase</span></code> and defaults to 2.
As a final step the count tables are mean-centered.
Note that this table contains <strong>all</strong> peaks, and no selection on differential peaks has been made.</p>
</section>
<section id="differential-motif-analysis">
<h3>Differential motif analysis<a class="headerlink" href="#differential-motif-analysis" title="Link to this heading"></a></h3>
<p>Seq2science supports differential motif analysis!
This analysis is based on the tool <a class="reference external" href="https://gimmemotifs.readthedocs.io/en/master/reference.html#command-gimme-maelstrom">gimme maelstrom</a>.
As input to gimme maelstrom is the log 2 transformed quantile normalized count table of the biological reps (average between reps) in the consensus peakset.
It then tries to solve a system of linear equations, where the output is the read counts in a peak, and the input is the motif score in the peak times the “motif activity”.
This motif activity can then be compared across biological replicates for differential motifs.</p>
</section>
<section id="differential-peak-analysis">
<h3>Differential peak analysis<a class="headerlink" href="#differential-peak-analysis" title="Link to this heading"></a></h3>
<p>On top of that, Seq2science can also use the raw peak counts table to perform differential peak analysis with DESeq2.
See <a class="reference external" href="../DESeq2.html">Differential gene/peak analysis page</a> for more information!</p>
</section>
<section id="trackhub">
<h3>Trackhub<a class="headerlink" href="#trackhub" title="Link to this heading"></a></h3>
<p>A UCSC compatible trackhub can be generated for this workflow. See the <a class="reference external" href="../results.html#trackhub">trackhub page</a><span class="raw-html-m2r"><!-- @IGNORE PREVIOUS: link --></span> for more information!</p>
</section>
</section>
<section id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Link to this heading"></a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on. Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs. As an example, the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">technical_replicates</span>    <span class="n">descriptive_name</span>    <span class="n">control</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">heart_1</span>      <span class="n">heart_merged</span>        <span class="n">GSM234</span>
<span class="n">GSM321</span>    <span class="n">GRCh38</span>      <span class="n">heart_1</span>      <span class="n">heart_merged</span>        <span class="n">GSM234</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">heart_2</span>      <span class="n">heart_not_merged</span>    <span class="n">GSM234</span>
<span class="n">GSMxzy</span>    <span class="n">danRer11</span>    <span class="n">stage_8</span>      <span class="n">stage_8</span>             <span class="n">GSM234</span>
<span class="n">GSM890</span>    <span class="n">danRer11</span>    <span class="n">stage_9</span>      <span class="n">stage_9</span>             <span class="n">GSM234</span>
</pre></div>
</div>
<section id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Link to this heading"></a></h3>
<p>If you use the pipeline on <strong>public data</strong> this should be the name of the accession (e.g. GSM2837484).
Accepted formats start with “GSM”, “SRR”, “SRX”, “DRR”, “DRX”, “ERR” or “ERX”.</p>
<p>If you use the pipeline on <strong>local data</strong> this should be the <em>basename</em> of the file without the <em>extension(s)</em>.
For example:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample1.fastq.gz</span></code> ——-&gt; <code class="docutils literal notranslate"><span class="pre">sample1</span></code> for single-ended data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> ┬&gt; <code class="docutils literal notranslate"><span class="pre">sample2</span></code> for paired-ended data <span class="raw-html-m2r"><br></span> <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> ┘</p></li>
</ul>
<p>For <strong>local data</strong>, some fastq files may have slightly different naming formats.
For instance, Illumina may produce a sample named <code class="docutils literal notranslate"><span class="pre">sample3_S1_L001_R1_001.fastq.gz</span></code> (and the <code class="docutils literal notranslate"><span class="pre">R2</span></code> fastq).
Seq2science will attempt to recognize these files based on the sample name <code class="docutils literal notranslate"><span class="pre">sample3</span></code>.</p>
<p>For <strong>both local and public data</strong>, identifiers used to recognize fastq files are the fastq read extensions (<code class="docutils literal notranslate"><span class="pre">R1</span></code> and <code class="docutils literal notranslate"><span class="pre">R2</span></code> by default) and the fastq suffix (<code class="docutils literal notranslate"><span class="pre">fastq</span></code> by default).
The directory where seq2science will store (or look for) fastqs is determined by the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> config option.
In the example above, the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> should be set to <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs</span></code>.
These setting can be changed in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
</section>
<section id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Link to this heading"></a></h3>
<p>Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</section>
<section id="control-input-column">
<h3>Control (input) column<a class="headerlink" href="#control-input-column" title="Link to this heading"></a></h3>
<p>In the <code class="docutils literal notranslate"><span class="pre">control</span></code> column you can (optionally) add the “sample name” of the input control, although it is not common to generate this for ATAC-seq.
Doing so looks like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">control</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">GSMabc</span>
</pre></div>
</div>
</section>
<section id="descriptive-name-column">
<h3>Descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Link to this heading"></a></h3>
<p>The descriptive_name column is used for the trackhub and multiqc report.
In the trackhub your tracks will be called after the descriptive name, and in the multiqc report there will be a button to rename your samples after this column.
The descriptive name can not contain ‘-’ characters, but underscores ‘_’ are allowed.</p>
</section>
<section id="technical-replicates-column">
<h3>Technical_replicates column<a class="headerlink" href="#technical-replicates-column" title="Link to this heading"></a></h3>
<p>Technical replicates, or any fastq file you may wish to merge on the fastq level, are set using the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column in the samples.tsv file.
All samples with the same name in the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column will be concatenated into one file with the replicate name.</p>
<p>Example <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> utilizing replicate merging:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">technical_replicates</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">heart</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">heart</span>
<span class="n">GSMxzy</span>    <span class="n">GRCh38</span>      <span class="n">stage8</span>
<span class="n">GSM890</span>    <span class="n">GRCh38</span>
</pre></div>
</div>
<p>Using this file in the alignment workflow will output <em>heart.bam</em>, <em>stage8.bam</em> and <em>GSM890.bam</em>.
The MultiQC will inform you of the trimming steps performed on all samples, and subsequent information of the ‘replicate’ files (of which only <em>heart</em> is merged).</p>
<p>Note: If you are working with multiple assemblies in one workflow, replicate names have to be unique between assemblies (you will receive a warning if names overlap).</p>
<section id="keep">
<h4>keep<a class="headerlink" href="#keep" title="Link to this heading"></a></h4>
<p>Replicate merging is turned on by default. It can be turned off by setting <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to <code class="docutils literal notranslate"><span class="pre">keep</span></code>.</p>
</section>
</section>
<section id="biological-replicates">
<h3>Biological replicates<a class="headerlink" href="#biological-replicates" title="Link to this heading"></a></h3>
<p>During ATAC-seq workflows, peak calling can be performed per biological condition, depending on your configuration setting. Biological conditions are determined by the <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code> column in the samples.tsv file. How these samples are handled is specified by configuration variable <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">biological_replicates</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">kidney</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">liver</span>
<span class="n">GSMxzy</span>    <span class="n">GRCh38</span>      <span class="n">liver</span>
<span class="n">GSM890</span>    <span class="n">GRCh38</span>
</pre></div>
</div>
<p>In this case peaks of the two liver samples (GSMabc and GSMxzy) will be combined by e.g. IDR.</p>
<section id="id4">
<h4>Keep<a class="headerlink" href="#id4" title="Link to this heading"></a></h4>
<p>By setting <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code> to keep, all biological replicates are analyzed individually. The <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code> column is ignored.</p>
</section>
<section id="irreproducible-discovery-rate-idr">
<h4>Irreproducible Discovery Rate (IDR)<a class="headerlink" href="#irreproducible-discovery-rate-idr" title="Link to this heading"></a></h4>
<p>One of the more common methods to combine biological replicates is by the irreproducible discovery rate (<a class="reference external" href="https://github.com/kundajelab/idr">idr</a>). Shortly; idr sorts all the peaks of two replicates separately on their significance. Since true peaks should be very significant for both replicates these peaks will be one of the highest sorted peaks. As peaks get less and less true (and thus their significance) their ordering also becomes more random between the samples. The idr method then only keeps the peak that overlap <em>nonrandomly</em> between the samples. The idr method only works for two replicates, so can not be used when you have more than 2. IDR can be turned on by setting <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code> to idr.</p>
</section>
<section id="fisher-s-method">
<h4>Fisher’s method<a class="headerlink" href="#fisher-s-method" title="Link to this heading"></a></h4>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Fisher%27s_method">Fisher’s method</a> simply is a method to ‘combine’ multiple p-values. This method is built-in for genrich and macs2, and allows any number of replicates (<code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">&gt;=</span> <span class="pre">2</span></code>). However a ‘disadvantage’ of this method is that it assumes the p-values a method generates are actual p-values (and not a general indication of significance). MACS2 is slightly notorious for that its p-values are not true p-values. Genrich claims by fitting on a log-normal distribution that their p-values are true. Even though fisher’s method might not be ideal, it’s the only option you have (we provide) within this pipeline for more than 2 replicates. Fisher’s method can be turned on by setting <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code> to fisher.</p>
</section>
</section>
<section id="a-mix-of-biological-and-technical-replicates">
<h3>A mix of biological and technical replicates<a class="headerlink" href="#a-mix-of-biological-and-technical-replicates" title="Link to this heading"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>    <span class="n">assembly</span>    <span class="n">technical_replicates</span>  <span class="n">biological_replicates</span>
<span class="n">GSM123</span>    <span class="n">GRCh38</span>      <span class="n">peter_kidney</span>         <span class="n">kidney</span>
<span class="n">GSMabc</span>    <span class="n">GRCh38</span>      <span class="n">peter_liver</span>          <span class="n">liver</span>
<span class="n">GSMxzy</span>    <span class="n">GRCh38</span>      <span class="n">lindsey_liver</span>        <span class="n">liver</span>
<span class="n">GSM890</span>    <span class="n">GRCh38</span>      <span class="n">lindsey_liver</span>        <span class="n">liver</span>
</pre></div>
</div>
<p>In this example we have data from Peter and Lindsey. We have one liver sample and one kidney sample from Peter. Of Lindsey we only have a liver sample, which was sequenced across two different lanes. With this sample sheet, the samples of Lindsey get merged after trimming. Then peak calling is done separately on each of the three remaining technical replicates (peter_kidney, peter_liver, lindsey_liver), and afterwards all liver technical replicates belonging to the same biological replicate (in this case peter_liver and lindsey_liver) are combined through e.g. IDR.</p>
</section>
<section id="colors-column">
<h3>Colors column<a class="headerlink" href="#colors-column" title="Link to this heading"></a></h3>
<p>If you are visualizing your data on the UCSC trackhub you can optionally specify the colors of each main track. For ATAC- and ChIP-seq, each biological replicate is a main track.
To do so, you can add the color by name (google “matplotlib colors” for the options), or RGB values, in the “colors” column. Empty fields are considered black.</p>
</section>
<section id="final-notes">
<h3>Final notes<a class="headerlink" href="#final-notes" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Make sure that the samples.tsv is a tab separated values file when running the pipeline.</p></li>
<li><p>Feel free to delete or add columns to your liking.</p></li>
</ul>
</section>
</section>
<section id="filling-out-the-config-yaml">
<h2>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Link to this heading"></a></h2>
<p>Every workflow has many configurable options, and can be set in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. In each <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) <strong>reasonable default</strong> values.</p>
<p>When a workflow starts it prints the complete configuration, and (almost) all these values can be added in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> and changed to your liking. You can see the complete set of configurable options in the <a class="reference external" href="../schemas.html">extensive docs</a>.</p>
</section>
<section id="best-practices">
<h2>Best practices<a class="headerlink" href="#best-practices" title="Link to this heading"></a></h2>
<section id="macs2-and-shift-with-paired-end-data">
<h3>MACS2 and –shift with paired-end data<a class="headerlink" href="#macs2-and-shift-with-paired-end-data" title="Link to this heading"></a></h3>
<p>When trying to detect the cutting sites of Tn5, it is common to shift all your reads for instance 100 bp upstream, and extend them to 200 long. This makes the cutting site of a read fall in the middle of this 200 bp region. Unfortunately this is only possible in single-end mode of MACS, and makes MACS2 throw away all(!) of the paired-end mates, resulting in only half the amount of reads…</p>
<p>We made a script that deletes all the information that reads are paired-end after alignment from a bam. In this way we bypass the problem that shift is only possible for single-end reads. If you want to make use of this option then set the <code class="docutils literal notranslate"><span class="pre">macs2_keep_mates</span></code> variable in the configuration to True.</p>
</section>
<section id="irreproducible-discovery-rate">
<h3>Irreproducible Discovery Rate<a class="headerlink" href="#irreproducible-discovery-rate" title="Link to this heading"></a></h3>
<p>For idr to work properly, it needs a large portion of peaks that are actually not true peaks. Therefore we recommend to call peaks ‘loosely’. One way of doing this is setting the <a class="reference external" href="https://github.com/taoliu/MACS#-q--pvalue">peak threshold q-value</a> low for macs2 in the configuration. The q-value controls the ratio of <code class="docutils literal notranslate"><span class="pre">false</span> <span class="pre">positives</span> <span class="pre">/</span> <span class="pre">(false</span> <span class="pre">positives</span> <span class="pre">+</span> <span class="pre">true</span> <span class="pre">positives)</span></code>. Setting a q-value of 1 guarantees you have plenty of false positives in your dataset for IDR to work properly.</p>
</section>
<section id="subsampling">
<h3>subsampling<a class="headerlink" href="#subsampling" title="Link to this heading"></a></h3>
<p>In some cases you might want to have the same number of reads between samples. By adding e.g. <code class="docutils literal notranslate"><span class="pre">subsample:1_000_000</span></code> seq2science will make sure that each sample contains at most a million reads. Seq2science always downsamples, never upsamples.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="alignment.html" class="btn btn-neutral float-left" title="Alignment" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chip_seq.html" class="btn btn-neutral float-right" title="ChIP-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>