<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNA-seq &mdash; seq2science  documentation</title><link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="scATAC-seq" href="scatac_seq.html" />
    <link rel="prev" title="ChIP-seq" href="chip_seq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> seq2science
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#downloading-and-indexing-of-assembly-s">Downloading and indexing of assembly(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-trimming">Read trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignment">Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bam-sieving">Bam sieving</a></li>
<li class="toctree-l4"><a class="reference internal" href="#strandedness">Strandedness</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gene-quantification-methods">Gene quantification methods</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-gene-expression-analysis">Differential gene expression analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-transcript-usage">Differential transcript usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-exon-usage">Differential exon usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#descriptive-name-column">Descriptive_name column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates-column">technical_replicates column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#colors-column">Colors column</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#genome-assembly-and-gene-annotation">Genome assembly and gene annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#aligners-star-vs-hisat2">Aligners: STAR vs HISAT2</a></li>
<li class="toctree-l4"><a class="reference internal" href="#quantifier-counts-vs-quantification">Quantifier: counts vs quantification</a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome-aligner-quantifier">Genome/Aligner/Quantifier</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reviewing-the-results">Reviewing the results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scrna_seq.html">scRNA-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../workflows.html">Workflows</a> &raquo;</li>
      <li>RNA-seq</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/rna_seq.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="rna-seq">
<h1>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h1>
<p>Running an RNA-seq analysis has never been easier!</p>
<div class="section" id="pipeline-steps">
<h2>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">¶</a></h2>
<p align="center">
  <img src="../../_static/rna_seq.png">
</p><div class="section" id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples.
You control which samples are used in the <a class="reference external" href="#filling-out-the-samples-tsv">samples.tsv</a>.
Background on public data can be found <a class="reference external" href="./download_fastq.html#download-sra-file">here</a>.</p>
</div>
<div class="section" id="downloading-and-indexing-of-assembly-s">
<h3>Downloading and indexing of assembly(s)<a class="headerlink" href="#downloading-and-indexing-of-assembly-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the assembly and its index you align your samples against already exist seq2science will start with downloading of the assembly through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>.</p>
</div>
<div class="section" id="read-trimming">
<h3>Read trimming<a class="headerlink" href="#read-trimming" title="Permalink to this headline">¶</a></h3>
<p>The pipeline starts by trimming the reads with Trim Galore! or Fastp (the default).
The trimmer will automatically trim the low quality 3’ ends of reads, and removes short reads.
During the quality trimming it automatically detects which adapter was used, and trims this as well.
Trimming parameters for the pipeline can be set in the configuration.</p>
</div>
<div class="section" id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h3>
<p>Reads are aligned using <code class="docutils literal notranslate"><span class="pre">HISAT2</span></code> or <code class="docutils literal notranslate"><span class="pre">STAR</span></code> (the default).
Sensible defaults have been set, but can be overwritten for either (or both) the indexing and alignment by specifying them in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>The pipeline will check if the assembly you specified is present in the <em>genome_dir</em>, and otherwise will download it for you through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>.
All these aligners require an index to be formed first for each assembly, but don’t worry, the pipeline does this for you.</p>
</div>
<div class="section" id="bam-sieving">
<h3>Bam sieving<a class="headerlink" href="#bam-sieving" title="Permalink to this headline">¶</a></h3>
<p>After aligning the bam you can choose to remove unmapped reads, low quality mappings, duplicates, and multimappers.
Again, sensible defaults have been set, but can be overwritten.</p>
</div>
<div class="section" id="strandedness">
<h3>Strandedness<a class="headerlink" href="#strandedness" title="Permalink to this headline">¶</a></h3>
<p>Most sequencing protocols at present are strand-specific.
This specificity can be used to help identify pseudogenes originating from antisense DNA, or genes with overlapping regions on opposite strands without ambiguity.
Strandedness is inferred automatically for all RNA-seq samples.
For aligners it is inferred by RSeQC, the results of which can be reviewed in the MultiQC.
RSeQC inference can be overwritten by column <code class="docutils literal notranslate"><span class="pre">strandedness</span></code> in the samples.tsv.
This column may contain identifiers <code class="docutils literal notranslate"><span class="pre">no</span></code>, <code class="docutils literal notranslate"><span class="pre">forward</span></code> or <code class="docutils literal notranslate"><span class="pre">reverse</span></code>.
If strandedness is unknown (for some samples), fields may be left blank or filled with <code class="docutils literal notranslate"><span class="pre">nan</span></code>.
Setting <code class="docutils literal notranslate"><span class="pre">ignore_strandedness</span></code> in the config.yaml will resulting in gene counting to assume all reads are unstranded.</p>
</div>
<div class="section" id="gene-quantification-methods">
<h3>Gene quantification methods<a class="headerlink" href="#gene-quantification-methods" title="Permalink to this headline">¶</a></h3>
<p>RNA-seq can be performed using gene counting or abundance estimation methods.
Gene counting methods require BAM files, which are generated and processed in the Alignment, Bam sieving and Strandedness steps.
Gene abundances require trimmed fastqs, and are therefore not influenced by the aforementioned steps.</p>
<div class="section" id="gene-counts-with-htseq-featurecounts">
<h4>Gene counts  (with HTSeq/featureCounts)<a class="headerlink" href="#gene-counts-with-htseq-featurecounts" title="Permalink to this headline">¶</a></h4>
<p>Gene counts are obtained from the filtered BAM files using either <code class="docutils literal notranslate"><span class="pre">HTSeq</span></code> or <code class="docutils literal notranslate"><span class="pre">featureCounts</span></code> (default <code class="docutils literal notranslate"><span class="pre">HTSeq</span></code>).
These counts are then combined into a count matrix per assembly for use in downstream analyses.</p>
</div>
<div class="section" id="gene-abundances-with-salmon">
<h4>Gene abundances (with Salmon)<a class="headerlink" href="#gene-abundances-with-salmon" title="Permalink to this headline">¶</a></h4>
<p>Gene abundances can be estimated using <code class="docutils literal notranslate"><span class="pre">Salmon</span></code>.
Reads are aligned against the transcriptome to obtain transcript abundances (sequence strandedness is inferred automatically by Salmon), then summarized to gene-level using tximeta.
The Gene-level counts matrix are output similar to the gene counts method.
Additionally, <code class="docutils literal notranslate"><span class="pre">Salmon</span></code> generates a gene-level TPM matrix and a SingleCellExperiment object which can be opened in <em>R</em>, containing the transcript- and gene-level summaries.</p>
</div>
</div>
<div class="section" id="differential-gene-expression-analysis">
<h3>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Permalink to this headline">¶</a></h3>
<p>Seq2science outputs gene counts matrices for each assembly.
Additionally, it can perform differential expression analysis automatically.
See the <a class="reference external" href="../DESeq2.html">Differential gene/peak analysis page</a> for more information!</p>
</div>
<div class="section" id="differential-transcript-usage">
<h3>Differential transcript usage<a class="headerlink" href="#differential-transcript-usage" title="Permalink to this headline">¶</a></h3>
<p>Quantifying with <code class="docutils literal notranslate"><span class="pre">Salmon</span></code>, the transcript-level summaries in the SingleCellExperiment object <em>should</em> be usable for differential transcript analysis with <code class="docutils literal notranslate"><span class="pre">DEXseq</span></code>, as described <a class="reference external" href="http://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqDTU/inst/doc/rnaseqDTU.html#dexseq">here</a>.</p>
</div>
<div class="section" id="differential-exon-usage">
<h3>Differential exon usage<a class="headerlink" href="#differential-exon-usage" title="Permalink to this headline">¶</a></h3>
<p>Differential exon analysis by <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DEXSeq.html">DEXseq</a> can be automatically prepared by setting <code class="docutils literal notranslate"><span class="pre">dexseq:</span> <span class="pre">True</span></code> in the config.yaml.
This will let seq2science to output an exon counts matrix per assembly, which can be loaded directly into <code class="docutils literal notranslate"><span class="pre">DEXSeqDataSet()</span></code>.</p>
<p>Note: this utilizes scripts implemented by DEXseq, which are <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DEXSeq/inst/doc/DEXSeq.html#24_Preparing_the_annotation">built for Ensembl genomes</a>.</p>
</div>
<div class="section" id="trackhub">
<h3>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h3>
<p>A UCSC compatible trackhub can be generated for this workflow.
See the <a class="reference external" href="../results.html#trackhub">trackhub page</a><span class="raw-html-m2r"><!-- @IGNORE PREVIOUS: link --></span> for more information!</p>
</div>
</div>
<div class="section" id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">¶</a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on.
Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs.
As an example, the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look something like this:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    technical_replicates    descriptive_name
GSM123    GRCh38      heart_1      heart_merged        GSM234
GSM321    GRCh38      heart_1      heart_merged        GSM234
GSMabc    GRCh38      heart_2      heart_not_merged    GSM234
GSMxzy    danRer11    stage_8      stage_8             GSM234
GSM890    danRer11    stage_9      stage_9             GSM234
</pre></div>
</div>
<div class="section" id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Permalink to this headline">¶</a></h3>
<p>If you use the pipeline on <strong>public data</strong> this should be the name of the accession (e.g. GSM2837484).
Accepted formats start with “GSM”, “SRR”, “SRX”, “DRR”, “DRX”, “ERR” or “ERX”.</p>
<p>If you use the pipeline on <strong>local data</strong> this should be the <em>basename</em> of the file without the <em>extension(s)</em>.
For example:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample1.fastq.gz</span></code> ——-&gt; <code class="docutils literal notranslate"><span class="pre">sample1</span></code> for single-ended data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> ┬&gt; <code class="docutils literal notranslate"><span class="pre">sample2</span></code> for paired-ended data <span class="raw-html-m2r"><br></span> <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> ┘</p></li>
</ul>
<p>For <strong>local data</strong>, some fastq files may have slightly different naming formats.
For instance, Illumina may produce a sample named <code class="docutils literal notranslate"><span class="pre">sample3_S1_L001_R1_001.fastq.gz</span></code> (and the <code class="docutils literal notranslate"><span class="pre">R2</span></code> fastq).
Seq2science will attempt to recognize these files based on the sample name <code class="docutils literal notranslate"><span class="pre">sample3</span></code>.</p>
<p>For <strong>both local and public data</strong>, identifiers used to recognize fastq files are the fastq read extensions (<code class="docutils literal notranslate"><span class="pre">R1</span></code> and <code class="docutils literal notranslate"><span class="pre">R2</span></code> by default) and the fastq suffix (<code class="docutils literal notranslate"><span class="pre">fastq</span></code> by default).
The directory where seq2science will store (or look for) fastqs is determined by the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> config option.
In the example above, the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> should be set to <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs</span></code>.
These setting can be changed in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
</div>
<div class="section" id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this headline">¶</a></h3>
<p>Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</div>
<div class="section" id="descriptive-name-column">
<h3>Descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Permalink to this headline">¶</a></h3>
<p>The descriptive_name column is used for the trackhub and multiqc report.
In the trackhub your tracks will be called after the descriptive name, and in the multiqc report there will be a button to rename your samples after this column.
The descriptive name can not contain ‘-‘ characters, but underscores ‘_’ are allowed.</p>
</div>
<div class="section" id="technical-replicates-column">
<h3>technical_replicates column<a class="headerlink" href="#technical-replicates-column" title="Permalink to this headline">¶</a></h3>
<p>Technical replicates, or any fastq file you may wish to merge, are set using the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column in the samples.tsv file.
All samples with the same name in the <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> column will be concatenated into one file with the replicate name.</p>
<p>Example <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> utilizing replicate merging:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    technical_replicates
GSM123    GRCh38      heart
GSMabc    GRCh38      heart
GSMxzy    GRCh38      stage8
GSM890    GRCh38
</pre></div>
</div>
<p>Using this file in the alignment workflow will output <em>heart.bam</em>, <em>stage8.bam</em> and <em>GSM890.bam</em>.
The MultiQC will inform you of the trimming steps performed on all samples, and subsequent information of the ‘replicate’ files (of which only <em>heart</em> is merged).</p>
<p>Note: If you are working with multiple assemblies in one workflow, replicate names have to be unique between assemblies (you will receive a warning if names overlap).</p>
<div class="section" id="keep">
<h4>keep<a class="headerlink" href="#keep" title="Permalink to this headline">¶</a></h4>
<p>Replicate merging is turned on by default.
It can be turned off by setting <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to <code class="docutils literal notranslate"><span class="pre">keep</span></code>.</p>
</div>
</div>
<div class="section" id="colors-column">
<h3>Colors column<a class="headerlink" href="#colors-column" title="Permalink to this headline">¶</a></h3>
<p>If you are producing a UCSC trackhub, seq2science will assign an alternating color gradient to your samples to distinguish them.
You can optionally specify the colors of each track by adding this column.
Colors can be added by name (google “matplotlib colors” for the options), or RGB values.
Empty fields are considered black.</p>
</div>
</div>
<div class="section" id="filling-out-the-config-yaml">
<h2>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this headline">¶</a></h2>
<p>Every workflow has many configurable options, and can be set in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file.
In each <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) <strong>reasonable default</strong> values.</p>
<p>When a workflow starts it prints the configuration variables influencing the workflow, and (almost) all these values can be added in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> and changed to your liking.
You can see the complete set of configurable options in the <a class="reference external" href="../schemas.html">extensive docs</a>.</p>
</div>
<div class="section" id="best-practices">
<h2>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h2>
<div class="section" id="genome-assembly-and-gene-annotation">
<h3>Genome assembly and gene annotation<a class="headerlink" href="#genome-assembly-and-gene-annotation" title="Permalink to this headline">¶</a></h3>
<p>The choice of genome assembly and gene annotation is of significant influence to the downstream analysis.
To explore available options, you could use <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>, which comes installed in the seq2science conda environment.
Installing the desired genome and gene annotation in the {genomes_dir} will cause seq2science to use these files.</p>
<p>If the genome/annotation is missing from the {genomes_dir}, seq2science will attempt to download the named assembly from Ensembl, UCSC and the NCBI (in that order).</p>
</div>
<div class="section" id="aligners-star-vs-hisat2">
<h3>Aligners: STAR vs HISAT2<a class="headerlink" href="#aligners-star-vs-hisat2" title="Permalink to this headline">¶</a></h3>
<p>Both aligners have been found to perform well.
Selection should be dependent on familiarity and configuration options.</p>
</div>
<div class="section" id="quantifier-counts-vs-quantification">
<h3>Quantifier: counts vs quantification<a class="headerlink" href="#quantifier-counts-vs-quantification" title="Permalink to this headline">¶</a></h3>
<p>Both methods have been found valid for differential gene expression analysis, although results vary somewhat.
Caution is advised for the genes found by only one of the two methods.</p>
</div>
<div class="section" id="genome-aligner-quantifier">
<h3>Genome/Aligner/Quantifier<a class="headerlink" href="#genome-aligner-quantifier" title="Permalink to this headline">¶</a></h3>
<p>The most significant choice to be made is the genome assembly and gene annotation.</p>
</div>
<div class="section" id="reviewing-the-results">
<h3>Reviewing the results<a class="headerlink" href="#reviewing-the-results" title="Permalink to this headline">¶</a></h3>
<p>Unless configured not to, Seq2science makes several assumptions on your data which may be incorrect:</p>
<ul class="simple">
<li><p>that strandedness of each sample can be determined automatically.</p></li>
<li><p>that duplicate reads are mostly caused by natural overexpression, not by library artifacts.</p></li>
</ul>
<p>These assumptions can be tested by inspecting the MultiQC.
Should the results be disappointing, they can be overwritten using the <code class="docutils literal notranslate"><span class="pre">strandedness</span></code> column in the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> and the <code class="docutils literal notranslate"><span class="pre">markduplicates</span></code> variable in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> respectively.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chip_seq.html" class="btn btn-neutral float-left" title="ChIP-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="scatac_seq.html" class="btn btn-neutral float-right" title="scATAC-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>