

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>RNA-seq &mdash; seq2science  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="ChIP-seq" href="chip_seq.html" />
    <link rel="prev" title="ATAC-seq" href="atac_seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#downloading-and-indexing-of-assembly-s">Downloading and indexing of assembly(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-trimming">Read trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gene-counting-with-htseq-featurecounts">Gene counting (with HTSeq/featureCounts)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gene-quantification-with-salmon">Gene quantification (with Salmon)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gene-abundances">Gene abundances</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-gene-expression-analysis">Differential gene expression analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-transcript-usage">Differential transcript usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-exon-usage">Differential exon usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#descriptive-name-column">Descriptive_name column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates">Technical replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#colors-column">Colors column</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#deseq2-contrast-designs">DESeq2 contrast designs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#conditions">Conditions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#batches">Batches</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../workflows.html">Workflows</a> &raquo;</li>
        
      <li>RNA-seq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/rna_seq.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rna-seq">
<h1>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h1>
<p>Running an RNA-seq analysis has never been easier!</p>
<div class="section" id="pipeline-steps">
<h2>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">¶</a></h2>
<p align="center">
  <img src="../../_static/rna_seq.png">
</p><div class="section" id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples. Take a look at the <a class="reference external" href="https://vanheeringen-lab.github.io/seq2science/content/workflows/download_fastq.html">downloading_fastq</a> workflow for extensive documentation about downloading of public samples.</p>
</div>
<div class="section" id="downloading-and-indexing-of-assembly-s">
<h3>Downloading and indexing of assembly(s)<a class="headerlink" href="#downloading-and-indexing-of-assembly-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the assembly and its index you align your samples against already exist seq2science will start with downloading of the assembly through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>.</p>
</div>
<div class="section" id="read-trimming">
<h3>Read trimming<a class="headerlink" href="#read-trimming" title="Permalink to this headline">¶</a></h3>
<p>The pipeline starts by trimming the reads with <a class="reference external" href="https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md">trim galore!</a>. Trim galore automatically first trims the low quality 3’ ends of reads, and removes short reads. After the quality trimming trim galore automatically detects which adapter was used, and trims it. The parameters of trim galore! for the pipeline can be set in the configuration by variable <em>trim_galore</em>.</p>
</div>
<div class="section" id="gene-counting-with-htseq-featurecounts">
<h3>Gene counting (with HTSeq/featureCounts)<a class="headerlink" href="#gene-counting-with-htseq-featurecounts" title="Permalink to this headline">¶</a></h3>
<p>RNA-seq can be performed using gene counting methods or gene quantification methods. The following section described the former (and default) method as implemented in seq2science.</p>
<div class="section" id="alignment">
<h4>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h4>
<p>Reads are aligned using <code class="docutils literal notranslate"><span class="pre">STAR</span></code> or <code class="docutils literal notranslate"><span class="pre">HISAT2</span></code> (default <code class="docutils literal notranslate"><span class="pre">STAR</span></code>). Sensible defaults have been set, but can be overwritten for either (or both) the indexing and alignment by specifying them in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>The pipeline will check if the assembly you specified is present in the <em>genome_dir</em>, and otherwise will download it for you through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>. All these aligners require an index to be formed first for each assembly, but don’t worry, the pipeline does this for you.</p>
</div>
<div class="section" id="bam-sieving">
<h4>Bam sieving<a class="headerlink" href="#bam-sieving" title="Permalink to this headline">¶</a></h4>
<p>After aligning the bam you can choose to remove unmapped reads, low quality mappings, duplicates, and multimappers. Again, sensible defaults have been set, but can be overwritten.</p>
</div>
<div class="section" id="strandedness">
<h4>Strandedness<a class="headerlink" href="#strandedness" title="Permalink to this headline">¶</a></h4>
<p>Most sequencing protocols at present are strand-specific. This specificity can be used to help identify pseudogenes originating from antisense DNA, or genes with overlapping regions on opposite strands without ambiguity.
Strandedness is inferred automatically for all RNA-seq samples. For aligners it is inferred by RSeQC, the results of which can be reviewed in the MultiQC.
RSeQC inference can be overwritten by column <code class="docutils literal notranslate"><span class="pre">strandedness</span></code> in the samples.tsv. This column may contain identifiers <code class="docutils literal notranslate"><span class="pre">no</span></code>, <code class="docutils literal notranslate"><span class="pre">forward</span></code> or <code class="docutils literal notranslate"><span class="pre">reverse</span></code>. If strandedness is unknown (for some samples), fields may be left blank or filled with <code class="docutils literal notranslate"><span class="pre">nan</span></code>.
Setting <code class="docutils literal notranslate"><span class="pre">ignore_strandedness</span></code> in the config.yaml will resulting in gene counting to assume all reads are unstranded.</p>
</div>
<div class="section" id="gene-counts">
<h4>Gene counts<a class="headerlink" href="#gene-counts" title="Permalink to this headline">¶</a></h4>
<p>Gene counts are obtained from the filtered bam files using either <code class="docutils literal notranslate"><span class="pre">HTSeq</span></code> or <code class="docutils literal notranslate"><span class="pre">featureCounts</span></code> (default <code class="docutils literal notranslate"><span class="pre">HTSeq</span></code>). These counts are then combined into a count matrix per assembly for use in downstream analyses.</p>
</div>
</div>
<div class="section" id="gene-quantification-with-salmon">
<h3>Gene quantification (with Salmon)<a class="headerlink" href="#gene-quantification-with-salmon" title="Permalink to this headline">¶</a></h3>
<p>RNA-seq can be performed using gene counting methods or gene quantification methods. The following section described the latter method as implemented in seq2science.</p>
</div>
<div class="section" id="gene-abundances">
<h3>Gene abundances<a class="headerlink" href="#gene-abundances" title="Permalink to this headline">¶</a></h3>
<p>Gene abundances can be estimated using <code class="docutils literal notranslate"><span class="pre">Salmon</span></code>. Reads are aligned against the transcriptome to obtain transcript abundances (sequence strandedness is inferred automatically by Salmon), then summarized to gene-level using tximeta.
Additionally, <code class="docutils literal notranslate"><span class="pre">Salmon</span></code> generates a gene-level TPM matrix and a SingleCellExperiment object which can be opened in <em>R</em>, containing the transcript- and gene-level summaries.</p>
</div>
<hr class="docutils" />
<div class="section" id="differential-gene-expression-analysis">
<h3>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Permalink to this headline">¶</a></h3>
<p>Seq2science outputs gene counts matrices for each assembly. Additionally, it can also perform differential expression analysis with <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2</a>.
To do so, add one or more contrast design(s) in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file, with matching identifiers in the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> file. Examples are given below, and the RNA-seq <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> contains a commented-out set of examples.</p>
<p>Note: (additional) design contrasts can be added at any time. After completing the workflow, rerunning Snakemake with new contrasts will only perform these analyses.</p>
<div class="section" id="deseq2-method">
<h4>DESeq2 method<a class="headerlink" href="#deseq2-method" title="Permalink to this headline">¶</a></h4>
<p>DESeq2 automatically performs library bias correction when loading your data. Batch correction is performed if included in the design.
After calculating differentially expressed genes, a multiple testing procedure is applied. This is either the Benjamini-Hochberg procedure (the default) or Independent Hypothesis Weighing.
Expression counts are log transformed (by default using the apeglm method). These defaults can be changed in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
</div>
<div class="section" id="deseq2-output">
<h4>DESeq2 output<a class="headerlink" href="#deseq2-output" title="Permalink to this headline">¶</a></h4>
<p>For each contrast design, the list of <em>all genes</em> is saved to file, with analysis results for expressed genes. Briefly, these include:</p>
<ul class="simple">
<li><p>The column <code class="docutils literal notranslate"><span class="pre">padj</span></code> contains the adjusted p-values after multiple testing. These should be used to identify DE genes.</p></li>
<li><p>The column <code class="docutils literal notranslate"><span class="pre">log2FoldChange</span></code> contains the fold change of each gene between the two conditions. The reference condition is the one <em>last mentioned</em> in the contrast design, so use <code class="docutils literal notranslate"><span class="pre">conditions_treatment_control</span></code>. If you use <code class="docutils literal notranslate"><span class="pre">conditions_control_treatment</span></code> the fold change is <em>inverted</em>.</p></li>
<li><p>Several other columns were kept for sake of completion, such as column <code class="docutils literal notranslate"><span class="pre">pvalue</span></code>, which contains <em>non-adjusted</em> p-values.</p></li>
</ul>
<p>In addition, MA and PCA plots are generated for each contrast design.</p>
<p>If the design includes a batch effect, several PCA plots are generated to visualize the effect of the batch correction.</p>
<p>DESeq2 models the batch effect in their package, but downstream methods may not. For this reason, seq2science will produce a batch-corrected counts matrix (and a batch corrected TPM matrix if quantified with Salmon).</p>
</div>
</div>
<div class="section" id="differential-transcript-usage">
<h3>Differential transcript usage<a class="headerlink" href="#differential-transcript-usage" title="Permalink to this headline">¶</a></h3>
<p>Quantifying with <code class="docutils literal notranslate"><span class="pre">Salmon</span></code>, the transcript-level summaries in the SingleCellExperiment object <em>should</em> be usable for differential transcript analysis with <code class="docutils literal notranslate"><span class="pre">DEXseq</span></code>, as described <a class="reference external" href="http://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqDTU/inst/doc/rnaseqDTU.html#dexseq">here</a>.</p>
</div>
<div class="section" id="differential-exon-usage">
<h3>Differential exon usage<a class="headerlink" href="#differential-exon-usage" title="Permalink to this headline">¶</a></h3>
<p>Differential exon analysis by <cite>``DEXseq`</cite> &lt;<a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DEXSeq.html">https://bioconductor.org/packages/release/bioc/html/DEXSeq.html</a>&gt;`_ can be automatically prepared by setting <code class="docutils literal notranslate"><span class="pre">dexseq:</span> <span class="pre">True</span></code> in the config.yaml.
This will let seq2science to output an exon counts matrix per assembly, which can be loaded directly into <code class="docutils literal notranslate"><span class="pre">DEXSeqDataSet()</span></code>.</p>
<p>Note: this utilizes scripts implemented by DEXseq, which are <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DEXSeq/inst/doc/DEXSeq.html#24_Preparing_the_annotation">built for Ensembl genomes</a>.</p>
</div>
<hr class="docutils" />
<div class="section" id="trackhub">
<h3>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h3>
<p>A UCSC compatible trackhub can be generated for this workflow. See the <a class="reference external" href="../results.html#trackhub">trackhub page</a><span class="raw-html-m2r"><!-- @IGNORE PREVIOUS: link --></span> for more information!</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">¶</a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on. Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs. As an example, the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look something like this:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    replicate    descriptive_name    control
GSM123    GRCh38      heart_1      heart_merged        GSM234
GSM321    GRCh38      heart_1      heart_merged        GSM234
GSMabc    GRCh38      heart_2      heart_not_merged    GSM234
GSMxzy    danRer11    stage_8      stage_8             GSM234
GSM890    danRer11    stage_9      stage_9             GSM234
</pre></div>
</div>
<div class="section" id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows, not just the RNA-seq workflow. If you use the pipeline on public data this should be the name of the accession (e.g. GSM2837484). If you use the pipeline on local data this should be the <em>basename</em> of the file without the <em>extension(s)</em>. For instance, <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample1.fastq.gz</span></code> would be <code class="docutils literal notranslate"><span class="pre">sample1</span></code> (for single-ended data). For paired-ended data <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> would be <code class="docutils literal notranslate"><span class="pre">sample2</span></code>.</p>
</div>
<div class="section" id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows, except the <em>downloading samples</em> workflow. Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</div>
<div class="section" id="descriptive-name-column">
<h3>Descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Permalink to this headline">¶</a></h3>
<p>The descriptive_name column is used for the trackhub and multiqc report. In the trackhub your tracks will be called after the descriptive name, and in the multiqc report there will be a button to rename your samples after this column.</p>
</div>
<div class="section" id="technical-replicates">
<h3>Technical replicates<a class="headerlink" href="#technical-replicates" title="Permalink to this headline">¶</a></h3>
<p>Technical replicates, or any fastq file you may wish to merge, are set using the <code class="docutils literal notranslate"><span class="pre">replicate</span></code> column in the samples.tsv file. All samples with the same name in the <code class="docutils literal notranslate"><span class="pre">replicate</span></code> column will be concatenated into one file with the replicate name.</p>
<p>Example <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> utilizing replicate merging:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    replicate
GSM123    GRCh38      heart
GSMabc    GRCh38      heart
GSMxzy    danRer11    stage8
GSM890    danRer11
</pre></div>
</div>
<p>Using this file in the alignment workflow will output <em>heart.bam</em>, <em>stage8.bam</em> and <em>GSM890.bam</em>. The MultiQC will inform you of the trimming steps performed on all samples, and subsequent information of the ‘replicate’ files (of which only <em>heart</em> is merged).</p>
<p><strong>Note:</strong>
If you are working with multiple assemblies in one workflow, replicate names have to be unique between assemblies (you will receive a warning if names overlap).</p>
<p>Replicate merging is turned on by default. It can be turned off by setting <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to <code class="docutils literal notranslate"><span class="pre">keep</span></code>.</p>
</div>
<div class="section" id="colors-column">
<h3>Colors column<a class="headerlink" href="#colors-column" title="Permalink to this headline">¶</a></h3>
<p>If you are visualizing your data on the UCSC trackhub you can optionally specify the colors of each track.
To do so, you can add the color by name (google “matplotlib colors” for the options), or RGB values, in the “colors” column. Empty fields are considered black.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="deseq2-contrast-designs">
<h2>DESeq2 contrast designs<a class="headerlink" href="#deseq2-contrast-designs" title="Permalink to this headline">¶</a></h2>
<p>The following section will guide you through the process of creating a DESeq2 contrast using only the samples.tsv and the config.yaml.</p>
<p>A contrast contains a <strong>condition</strong>, which is the variable you wish to analyse, and which contains 2 or more states. Optionally, a contrast can contain a <strong>batch</strong> effect, which you want to correct for.</p>
<div class="section" id="conditions">
<h3>Conditions<a class="headerlink" href="#conditions" title="Permalink to this headline">¶</a></h3>
<p>In samples.tsv, add a column for every property of your samples.
Examples:</p>
<ul class="simple">
<li><p>a column named ‘conditions’ with values ‘control’ and ‘treated’.</p></li>
<li><p>a column named ‘conditions’ with values ‘control’, ‘treatmentA’ and ‘treatmentB’.</p></li>
<li><p>a column named ‘stages’ with values 1, 2 and 3.</p></li>
</ul>
<p>Next, in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> add one or more contrasts.</p>
<p>In order to compare two groups from a list (e.a. stage 1 vs stage 3), the contrast condition is the column name, followed by the groups (<code class="docutils literal notranslate"><span class="pre">stages_3_1</span></code>). Note that the <em>last</em> group (stage 1 in this example) is the reference group.</p>
<p>To compare all groups against a reference (e.a. control vs both treatments A and B), the contrast condition is the column name, followed by the reference group (<code class="docutils literal notranslate"><span class="pre">condition_control</span></code>). Alternatively, group ‘all’ may also be added (<code class="docutils literal notranslate"><span class="pre">condition_all_control</span></code>).</p>
</div>
<div class="section" id="batches">
<h3>Batches<a class="headerlink" href="#batches" title="Permalink to this headline">¶</a></h3>
<p>If your data has batch effects, these can be accounted for. Note that this can only be done if the batch effects are spread over different conditions.</p>
<p>Similar to conditions, add a column to samples.tsv and note the batch of each sample:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>sequencing_month</p></th>
<th class="head"><p>conditions</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_1</p></td>
<td><p>jan</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-odd"><td><p>sample_2</p></td>
<td><p>feb</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-even"><td><p>sample_3</p></td>
<td><p>jan</p></td>
<td><p>treatment</p></td>
</tr>
<tr class="row-odd"><td><p>sample_4</p></td>
<td><p>feb</p></td>
<td><p>treatment</p></td>
</tr>
</tbody>
</table>
<p>In this example, the sequencing_month may have caused a batch effect. Since the (potential) batch effect is spread over the test conditions (control vs treatment) it can be taken into account during DE analysis. The contrast design to do this would be <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code></p>
<div class="section" id="more-design-contrast-examples">
<h4>More design contrast examples<a class="headerlink" href="#more-design-contrast-examples" title="Permalink to this headline">¶</a></h4>
<p>The example design <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code> may also be written as <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_all_control</span></code>, <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_control_all</span></code> or <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_control</span></code>. In each case, seq2science will understand “control” is the reference group, and compare it against the other groups in the “conditions” columns.</p>
<p>For consistency with regular DESeq2 contrast designs, designs may start with <code class="docutils literal notranslate"><span class="pre">~</span></code>, example: <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code>.</p>
<p>Our gremlins will carefully unpack all of your input and pass the unique designs to DESeq2.</p>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="chip_seq.html" class="btn btn-neutral float-right" title="ChIP-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="atac_seq.html" class="btn btn-neutral float-left" title="ATAC-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>