

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>scATAC-seq &mdash; seq2science  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="scRNA-seq" href="scrna_seq.html" />
    <link rel="prev" title="RNA-seq" href="rna_seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_seq.html">RNA-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">scATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#merge-cell-fastq-files-based-on-processing">merge cell fastq files based on processing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignment">Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generate-and-bin-a-snap-object-for-downstream-analysis">Generate and bin a SNAP_Object for downstream analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicate-column">technical_replicate column</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="scrna_seq.html">scRNA-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../workflows.html">Workflows</a> &raquo;</li>
        
      <li>scATAC-seq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/scatac_seq.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="scatac-seq">
<h1>scATAC-seq<a class="headerlink" href="#scatac-seq" title="Permalink to this headline">¶</a></h1>
<p>Running a scATAC-seq pipeline has never been easier!</p>
<p>Does your scATAC protocol generate single cell fastq files?
And do they need to be mapped and analysed?
Then this pipeline is for your!
This pipeline takes single cell fastq files, performs extensive (plate based) QC steps, and generates a binned SNAP-object that can be used for further downstream analysis following <a class="reference external" href="https://github.com/r3fang/SnapATAC">this tutorial</a>.</p>
<p>If you have a Cellranger object instead (output 10x), it is relatively easy to directly generate a SNAPobject, look at the documentation on the <a class="reference external" href="https://github.com/r3fang/SnapATAC/wiki/FAQs#10X_snap">SnapATAC github</a> instead.</p>
<p>Alternatively, if you have a large FASTQ file containing a cell specific ID in the header, it might be possible to run this pipeline with some alterations.
The ‘merge_replica’ outputs a fastq trimmed fastq file with the cell_id after the <code class="docutils literal notranslate"><span class="pre">&#64;</span></code> in the header, e.g. <code class="docutils literal notranslate"><span class="pre">&#64;sampleXcellY:NS500173:518:HTYVTBGXB:4:23609:2756:19401</span> <span class="pre">2:N:0:ACTCGCTA+ATTAGACG</span></code>.</p>
<div class="section" id="pipeline-steps">
<h2>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">¶</a></h2>
<p align="center">
  <img src="../../_static/scatac_seq.png">
</p><div class="section" id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples.
You control which samples are used in the <a class="reference external" href="#filling-out-the-samples-tsv">samples.tsv</a>.
Background on public data can be found <a class="reference external" href="./download_fastq.html#download-sra-file">here</a>.</p>
</div>
<div class="section" id="merge-cell-fastq-files-based-on-processing">
<h3>merge cell fastq files based on processing<a class="headerlink" href="#merge-cell-fastq-files-based-on-processing" title="Permalink to this headline">¶</a></h3>
<p>Single cell fastq files get a barcode id corresponding to the sample name. This is added to the fastq header, after which each single cell fastq is merged to a large FASTQ file with all the technical replica’s (for example all cells from a plate). This results in a multi-QC report containing QC per technical replica’s. This makes lower plate quality easy to spot.
Additional single cell QC will be available in downstream analysis of the SNAP-object.
See <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/4.-Replicate-handling">Replicate handling</a> for more information.
See <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/blob/docs/docs/fillingout.md">Filling out the samples.tsv and config.yaml</a> for more information.</p>
</div>
<div class="section" id="alignment">
<h3>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h3>
<p>Make sure to take a look at our <a class="reference external" href="./alignment.html">alignment workflow</a>.</p>
</div>
<div class="section" id="generate-and-bin-a-snap-object-for-downstream-analysis">
<h3>Generate and bin a SNAP_Object for downstream analysis<a class="headerlink" href="#generate-and-bin-a-snap-object-for-downstream-analysis" title="Permalink to this headline">¶</a></h3>
<p>To find cellular heterogeneity before peak calling, SNAP-ATAC binns the genome. Using this for demensionality reduction of the data and finding cell clusters before peak calling. By default the pipeline includes bins of 5kB, but this is changable in the config.yaml file.
After clustering peaks can be called on the aggregated cluster signals. This makes it easier to find peaks present only in a small cluster of cells. This does however mean that the pipeline doesnt perform peakcalling, this is done later in the snapATAC vignette.</p>
</div>
</div>
<div class="section" id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">¶</a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on. Each workflow starts with a samples.tsv as an example, and you should adapt it to your specific needs. One thing you need to check before filling a sample into your sample.tsv file, is that the fastq file is not completely empty! If a fastq file has a size of 0 it will crash the pipeline.
You can run bash code to automatically remove completely empty fastq files from a directory, e.g. run:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cd fastq-dir-here
find . -size 0 -delete
</pre></div>
</div>
<p>After removing all the completely empty fastq files, use the remainder to fill your samples.tsv file with.
As an example, the samples.tsv could look something like this:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample  assembly    technical_replicate
GSM1596256  hg38    H1ESC
GSM1596257  hg38    H1ESC
GSM1596258  hg38    H1ESC
GSM1596259  hg38    H1ESC
GSM1596260  hg38    H1ESC
GSM1596261  hg38    H1ESC
GSM1596262  hg38    H1ESC
GSM1596263  hg38    H1ESC
GSM1596351  hg38    GM
GSM1596352  hg38    GM
GSM1596353  hg38    GM
GSM1596354  hg38    GM
GSM1596355  hg38    GM
GSM1596356  hg38    GM
GSM1596357  hg38    GM
GSM1596358  hg38    GM
GSM1596359  hg38    GM
iPSC-2-KC-diff-plate2-16718m701m511 hg38    plate2
iPSC-2-KC-diff-plate2-plate2-16718m701m513  hg38    plate2
iPSC-2-KC-diff-plate2-plate2-16718m701m515  hg38    plate2
iPSC-2-KC-diff-plate2-plate2-16718m701m516  hg38    plate2
iPSC-2-KC-diff-plate2-plate2-16718m701m517  hg38    plate2
</pre></div>
</div>
<div class="section" id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Permalink to this headline">¶</a></h3>
<p>If you use the pipeline on <strong>public data</strong> this should be the name of the accession (e.g. GSM2837484).
Accepted formats start with “GSM”, “SRR”, “SRX”, “DRR”, “DRX”, “ERR” or “ERX”.</p>
<p>If you use the pipeline on <strong>local data</strong> this should be the <em>basename</em> of the file without the <em>extension(s)</em>.
For example:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample1.fastq.gz</span></code> ——-&gt; <code class="docutils literal notranslate"><span class="pre">sample1</span></code> for single-ended data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> ┬&gt; <code class="docutils literal notranslate"><span class="pre">sample2</span></code> for paired-ended data <span class="raw-html-m2r"><br></span> <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> ┘</p></li>
</ul>
<p>For <strong>local data</strong>, some fastq files may have slightly different naming formats.
For instance, Illumina may produce a sample named <code class="docutils literal notranslate"><span class="pre">sample3_S1_L001_R1_001.fastq.gz</span></code> (and the <code class="docutils literal notranslate"><span class="pre">R2</span></code> fastq).
Seq2science will attempt to recognize these files based on the sample name <code class="docutils literal notranslate"><span class="pre">sample3</span></code>.</p>
<p>For <strong>both local and public data</strong>, identifiers used to recognize fastq files are the fastq read extensions (<code class="docutils literal notranslate"><span class="pre">R1</span></code> and <code class="docutils literal notranslate"><span class="pre">R2</span></code> by default) and the fastq suffix (<code class="docutils literal notranslate"><span class="pre">fastq</span></code> by default).
The directory where seq2science will store (or look for) fastqs is determined by the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> config option.
In the example above, the <code class="docutils literal notranslate"><span class="pre">fastq_dir</span></code> should be set to <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs</span></code>.
These setting can be changed in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
</div>
<div class="section" id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this headline">¶</a></h3>
<p>Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</div>
<div class="section" id="technical-replicate-column">
<h3>technical_replicate column<a class="headerlink" href="#technical-replicate-column" title="Permalink to this headline">¶</a></h3>
<p>Here you put to which technical replica each cell corresponds. E.g. the plate from which the cells were sequenced. Per technical replica QC will be generated. For public data you can put all cells under the same technical replica name.</p>
</div>
</div>
<div class="section" id="filling-out-the-config-yaml">
<h2>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this headline">¶</a></h2>
<p>Every workflow has many configurable options, and can be set in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. In each <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) <strong>reasonable default</strong> values.</p>
<p>When a workflow starts it prints the complete configuration, and (almost) all these values can be added in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> and changed to your liking. You can see the complete set of configurable options in the <a class="reference external" href="../schemas.html">extensive docs</a>.</p>
</div>
<div class="section" id="best-practices">
<h2>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h2>
<p>Take a look at the <a class="reference external" href="../scATAC_postprocessing.html">postprocessing example</a>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="scrna_seq.html" class="btn btn-neutral float-right" title="scRNA-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="rna_seq.html" class="btn btn-neutral float-left" title="RNA-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>