

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>scRNA-seq &mdash; seq2science  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Using the results" href="../results.html" />
    <link rel="prev" title="scATAC-seq" href="scatac_seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="rna_seq.html">RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">scRNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloading-of-sample-s">Downloading of sample(s)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#downloading-and-indexing-of-assembly-s">Downloading and indexing of assembly(s)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#read-trimming">Read trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#quantification">Quantification</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a></li>
<li class="toctree-l3"><a class="reference internal" href="#how-to-get-the-pipeline-started">How to get the pipeline started?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-column">Sample column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#assembly-column">Assembly column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#descriptive-name-column">Descriptive_name column</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-assembly-extensions">Custom assembly extensions</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensive_docs.html">Extensive docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../workflows.html">Workflows</a> &raquo;</li>
        
      <li>scRNA-seq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/scrna_seq.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="scrna-seq">
<h1>scRNA-seq<a class="headerlink" href="#scrna-seq" title="Permalink to this headline">¶</a></h1>
<p>Running a scRNA-seq pipeline has never been easier!</p>
<div class="section" id="pipeline-steps">
<h2>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">¶</a></h2>
<p align="center">
  <img src="../../_static/scrna_seq.png">
</p><div class="section" id="downloading-of-sample-s">
<h3>Downloading of sample(s)<a class="headerlink" href="#downloading-of-sample-s" title="Permalink to this headline">¶</a></h3>
<p>Depending on whether the samples you start seq2science with is your own data, public data, or a mix, the pipeline might start with downloading samples.
Take a look at the <a class="reference external" href="./download_fastq.html">downloading_fastq</a> workflow for extensive documentation about downloading of public samples.</p>
</div>
</div>
<div class="section" id="downloading-and-indexing-of-assembly-s">
<h2>Downloading and indexing of assembly(s)<a class="headerlink" href="#downloading-and-indexing-of-assembly-s" title="Permalink to this headline">¶</a></h2>
<p>Depending on whether the assembly and its index you align your samples against already exist seq2science will start with downloading of the assembly through genomepy.</p>
<div class="section" id="read-trimming">
<h3>Read trimming<a class="headerlink" href="#read-trimming" title="Permalink to this headline">¶</a></h3>
<p>The pipeline starts by trimming the reads with fastp.
Fastp automatically detects sequencing adapters, removes short reads and performs quality filtering.
The parameters of fastp for the pipeline can be set in the configuration by the variable fastp.</p>
<p><strong>Note</strong>: fastp currently runs in single-end mode even if you provide paired-end fastq files.
This is due to the fact that reads, barcode and umi sequences are stored in separate fastq files.
After trimming, the fastq containing the reads, usually R2, may contains less reads then R1.
Therefore, we perform an intermediate step by running <code class="docutils literal notranslate"><span class="pre">fastq-pair</span></code> to remove singleton reads before proceeding any further.</p>
</div>
<div class="section" id="quantification">
<h3>Quantification<a class="headerlink" href="#quantification" title="Permalink to this headline">¶</a></h3>
<p>Quantification is performed by running the kb-python wrapper for Kallisto bustools.
Kallisto bustools relies on pseudo-alignment of scRNA reads against a reference transcriptome index.
The resulting count matrices can be further processed with scRNA toolkits, such as Seurat or Scanpy.</p>
</div>
</div>
<div class="section" id="best-practices">
<h2>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="how-to-get-the-pipeline-started">
<h2>How to get the pipeline started?<a class="headerlink" href="#how-to-get-the-pipeline-started" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">¶</a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on.
Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs.
As an example, the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look something like this:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample  assembly        descriptive_name
pbmc    GRCh38.p13      pbmc
</pre></div>
</div>
<div class="section" id="sample-column">
<h3>Sample column<a class="headerlink" href="#sample-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows, not just the scRNA-seq workflow.
If you use the pipeline on public data this should be the name of the accession (e.g. GSM2837484).
If you use the pipeline on local data this should be the <em>basename</em> of the file without the <em>extension(s)</em>.
For example, for paired-ended data <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R1.fastq.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">/home/user/myfastqs/sample2_R2.fastq.gz</span></code> would be <code class="docutils literal notranslate"><span class="pre">sample2</span></code>.</p>
</div>
<div class="section" id="assembly-column">
<h3>Assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows, except the <em>downloading samples</em> workflow.
Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</div>
<div class="section" id="descriptive-name-column">
<h3>Descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Permalink to this headline">¶</a></h3>
<p>The descriptive_name column is used for the multiqc report.
In the multiqc report there will be a button to rename your samples after this column.</p>
</div>
<div class="section" id="filling-out-the-config-yaml">
<h3>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this headline">¶</a></h3>
<p>Every workflow has many configurable options, and can be set in the config.yaml file.
In each config.yaml we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) reasonable default values.</p>
<p>After initializing your working directory and editing the <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> file, you have to specify if you either want to perform quantification or RNA velocity analysis.
For velocity analysis, add the <code class="docutils literal notranslate"><span class="pre">--workflow</span> <span class="pre">lamanno</span></code> argument to the ref and count properties as shown below.</p>
<div class="section" id="quantifier">
<h4>Quantifier<a class="headerlink" href="#quantifier" title="Permalink to this headline">¶</a></h4>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>quantifier:
  kallistobus:
    count: &#39;-x 10xv3 --h5ad --verbose&#39;
</pre></div>
</div>
</div>
<div class="section" id="rna-velocity">
<h4>RNA velocity<a class="headerlink" href="#rna-velocity" title="Permalink to this headline">¶</a></h4>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>quantifier:
  kallistobus:
    ref: &#39;--workflow lamanno&#39;
    count: &#39;-x 10xv3 --h5ad --verbose --workflow lamanno&#39;
</pre></div>
</div>
<p>This will ensure that kallisto bustools generates the required files for RNA velocity estimation and produces count matrices for unspliced/spliced mRNA.</p>
</div>
<div class="section" id="bus-barcode-umi-set-format">
<h4>BUS (Barcode/UMI/Set) format<a class="headerlink" href="#bus-barcode-umi-set-format" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">-x</span></code> argument indicates the read and file positions of UMIs and barcodes in the supplied R1/R2 fastq files.
Kallisto bustools should auto-detect the correct settings if you use the short-hand syntax for your technology of choice, such as <code class="docutils literal notranslate"><span class="pre">-x</span> <span class="pre">10xv2</span></code>.
Internally, this is translated to the following group of <code class="docutils literal notranslate"><span class="pre">bc:umi:set</span></code> triplets:</p>
<p><code class="docutils literal notranslate"><span class="pre">0,0,16:0,16,26:1,0,0</span></code></p>
<p>The <code class="docutils literal notranslate"><span class="pre">bc:umi:set</span></code> format can be supplied as an alternative to the short-hand syntax.
For more information on the BUS format, consider the <a class="reference external" href="https://pachterlab.github.io/kallisto/manual">Kallisto</a> manual.</p>
</div>
</div>
<div class="section" id="custom-assembly-extensions">
<h3>Custom assembly extensions<a class="headerlink" href="#custom-assembly-extensions" title="Permalink to this headline">¶</a></h3>
<p>The genome and/or gene annotation can be extended with custom files, such as ERCC spike-ins for scRNA-seq.
To do so, add <code class="docutils literal notranslate"><span class="pre">custom_genome_extension:</span> <span class="pre">path/to/spike_in.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">custom_annotation_extension:</span> <span class="pre">path/to/spike_in.gtf</span></code> to the config.
Seq2science will place the customized assembly in a separate folder in the <code class="docutils literal notranslate"><span class="pre">genome_dir</span></code>.
You can control the name of the customized assembly by setting <code class="docutils literal notranslate"><span class="pre">custom_assembly_suffix</span></code> in the config.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../results.html" class="btn btn-neutral float-right" title="Using the results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="scatac_seq.html" class="btn btn-neutral float-left" title="scATAC-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, Rebecca Snabel, Tilman Schäfers, &amp; Simon van Heeringen..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>