# complement the configuration
configfile: "config.yaml"


# do onstart/onexit things
sample_schemas = ['sample', 'assembly', 'strandedness']
config_schemas = ['general', 'download', 'alignment_general', 'alignment_specific', 'gene_expression', 'rna_trackhub']
include: "../../rules/configuration.smk"


# load the remaining relevant rules
include: f"{config['rule_dir']}/alignment.smk"
include: f"{config['rule_dir']}/DE_analysis.smk"
include: f"{config['rule_dir']}/gene_counts.smk"
include: f"{config['rule_dir']}/get_genome.smk"
include: f"{config['rule_dir']}/get_fastq.smk"
include: f"{config['rule_dir']}/merge_replicates.smk"
include: f"{config['rule_dir']}/trackhub.smk"
include: f"{config['rule_dir']}/trimming.smk"
include: f"{config['rule_dir']}/qc.smk"


# load the quality control files
quality_control = [get_trimming_qc, get_alignment_qc]

# load DE analysis contrasts
contrasts = get_contrasts()

rule count_all:
    """
    count expressed genes per assembly and (optionally) perform differential expression analysis
    """
    input:
        expand("{result_dir}/gene_counts/{assemblies}-counts.tsv",              **{**config, **{'assemblies': set(samples['assembly'])}}),

        expand("{qc_dir}/multiqc_{assemblies}.html",                            **{**config, **{'assemblies': set(samples['assembly'])}}),

        [expand(f"{{result_dir}}/{{diffexp}}/{{assemblies}}-{contrast}.{data}", **{**config, **{'assemblies': set(samples['assembly'])}}) for data in ['diffexp.tsv', 'ma_plot.svg', 'pca_plot.svg'] for contrast in contrasts if config.get('contrasts', False)], #[expand(f"{{result_dir}}/{{diffexp}}/{{assemblies}}-{contrast}.diffexp.tsv",  **{**config, **{'assemblies': set(samples['assembly'])}}) for contrast in contrasts if config.get('contrasts', False)],
        expand("{result_dir}/deseq2/{assemblies}-clustering.svg",               **{**config, **{'assemblies': set(samples['assembly'])}}),

        expand("{result_dir}/trackhub", **config)
